---
title: "Annotate phosphorylation sites "
output: html_notebook
---
```{r}
#! /usr/bin/env Rscript

# Author(s): Ignatius Pang
# Email: ipang@cmri.org.au
# Childrenâ€™s Medical Research Institute, finding cures for childhood genetic diseases
```

```{r}

options(knitr.duplicate.label = "allow")


#Test if BioManager is installed 
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
   BiocManager::install(version = "3.12")
}

# load pacman package manager
if(!require(pacman)){
    install.packages("pacman")
    library(pacman)
}
p_load(optparse)
p_load(tictoc)
```
```{r}
tic()
```

```{r}

group_pattern <- "RPE"
## Directories management 
base_dir <- here::here()
data_dir <- file.path( base_dir, "Data")
results_dir <- file.path(base_dir, "Results",  paste0(group_pattern, "90"),  "Phosphopeptides", "DE_Analysis")
source_dir <- file.path(base_dir, "..")

phosphosite_db_dir <- file.path(   data_dir,  "PhosphositePlus", "20210730")

```
## Command to convert the R markdown file to a R file for use in command line. 
knitr::purl( file.path(  source_dir, "Shared",  "annot_phos_cmd.Rmd"), 
             output=file.path( source_dir,  "Shared",   "annot_phos_cmd.R") ) 

```{r}

input_long_file <- file.path( results_dir, "de_phos_long.tsv" ) 
input_wide_file  <- file.path( results_dir, "de_phos_wide.tsv" ) 
output_long_file <- file.path(results_dir, "de_phos_long_annot.tsv" ) 
output_wide_file <- file.path(results_dir, "de_phos_wide_annot.tsv" ) 
near_ptm_num_residues <- 5
taxonomy_id <- 10090
raw_counts_file <- file.path( base_dir, "Results",  paste0(group_pattern, "90"),  "Phosphopeptides", "Abundance_Tables", "sum_phoshpsites.tsv")

command_line_options <- commandArgs(trailingOnly = TRUE)

if( length(command_line_options ) > 0 ) {
  parser <- OptionParser(add_help_option =TRUE)

  parser <- add_option(parser, c( "--tax-id"), type="integer", default="", dest = "taxonomy_id",
                       help="The NCBI taxonomy ID of the organism being investigated (e.g. M. musculus=10090, H. sapien=9606).",
                       metavar="integer")     
  
  parser <- add_option(parser, c( "--output-dir"), type="character", default="", dest = "results_dir",
                       help="Directory path for all results files.",
                       metavar="string")   
  
  parser <- add_option(parser, c("--input-wide"), type="character", default="", dest = "input_wide_file",
                       help="Results table with values in wider format.",
                       metavar="string")   
  
  parser <- add_option(parser, c("--input-long"), type="character", default="", dest = "input_long_file",
                       help="Results table with values in longer format.",
                       metavar="string") 
  
  parser <- add_option(parser, c("--raw-counts"), type="character", default="", dest = "raw_counts_file",
                       help="Input file with the protein abundance data.",
                       metavar="string")   
  
  parser <- add_option(parser, c( "--output-wide"), type="character", default="", dest = "output_wide_file",
                       help="Results table with values in wider format.",
                       metavar="string")   
  
  parser <- add_option(parser, c( "--output-long"), type="character", default="", dest = "output_long_file",
                       help="Results table with values in longer format.",
                       metavar="string")   
  
  parser <- add_option(parser, c( "--psp-dir"), type="character", default="", dest = "phosphosite_db_dir",
                       help="File path for location of PhosphoSitePlus DB data files (https://www.phosphosite.org/staticDownloads).",
                       metavar="string")  
  
  
  parser <- add_option(parser, c( "--near-ptm"), type="integer", default=5, dest = "near_ptm_num_residues",
                       help="Find other PTM near within +/- this many number of residues from the phosphosites.",
                       metavar="integer") 
    
  print(command_line_options)
  
  cmd_arguments <- parse_args(parser)
  
  print(cmd_arguments)  

  input_wide_file <- cmd_arguments$input_wide_file
  input_long_file <- cmd_arguments$input_long_file
  output_wide_file <- cmd_arguments$output_wide_file
  output_long_file <- cmd_arguments$output_long_file
  results_dir <- cmd_arguments$results_dir
  taxonomy_id <- cmd_arguments$taxonomy_id
  raw_counts_file <- cmd_arguments$raw_counts_file
  phosphosite_db_dir <- cmd_arguments$phosphosite_db_dir
  near_ptm_num_residues <- cmd_arguments$near_ptm_num_residues
  
  if( results_dir == "" ) { stop("No value for --output-dir was supplied.") }
  if( group_pattern == "" ) { stop("No value for --group-pattern was supplied.") }
  if( input_wide_file == "" ) { stop("No value for --input-wide was supplied. ") }
  if( input_long_file == "" ) { stop("No value for --input-long was supplied. ") }
  if( output_wide_file == "" ) { stop("No value for --output-wide was supplied. ") }
  if( output_long_file == "" ) { stop("No value for --output-long was supplied. ") }
  if( taxonomy_id == "" ) { stop("No value for --tax_id was supplied.") }
  if( raw_counts_file == "" ) { stop("No value for --raw-counts was supplied.") }
  if( phosphosite_db_dir == "" ) { stop("No value for --psp-dir was supplied.") }

}

```

```{r}
p_load(tidyverse)
p_load(vroom)
p_load(magrittr)
p_load(rlang)
p_load(ProteomeRiver)

p_load(UniProt.ws)
p_load(biomaRt)
p_load(GO.db)
```

```{r}
createDirIfNotExists(results_dir)
ks_file <-  file.path( phosphosite_db_dir, "Kinase_Substrate_Dataset") 
reg_sites_file <- file.path( phosphosite_db_dir, "Regulatory_sites" ) 
disease_file <- file.path( phosphosite_db_dir, "Disease-associated_sites")
```

```{r}
print("Read abundance table.")
abundance_tbl <- vroom::vroom(raw_counts_file)

print("Read differentially abundant phosphopeptides table in long format")
de_phos_long <- vroom::vroom( input_long_file) %>%
  mutate( sites_id_copy = sites_id, .after="sites_id") %>%
  separate( sites_id_copy, sep="!", into=c("uniprot_acc", "gene_name", "position", "sequence")) %>%
  mutate( residue= purrr::map_chr( sequence, ~{ str_replace_all( ., "[A-Z_]{7}(.)[A-Z_]{7}([\\:;\\|]*)", "\\1\\2"    )  }  )) %>%
  dplyr::relocate(residue, .before="position") 

print("Read differentially abundant phosphopeptides table in wide format")
de_phos_wide <- vroom::vroom( input_wide_file) %>%
  mutate( sites_id_copy = sites_id, .after="sites_id") %>%
  separate( sites_id_copy, sep="!", into=c("uniprot_acc", "gene_name", "position", "sequence")) %>%
  mutate( residue= purrr::map_chr( sequence, ~{ str_replace_all( ., "[A-Z_]{7}(.)[A-Z_]{7}([\\:;\\|]*)", "\\1\\2"    )  }  )) %>%
  dplyr::relocate(residue, .before="position")

```

```{r}
print("Read PhosphoSitePlus (PSP) kinase-substrate table.")
ks_tbl <- vroom::vroom( ks_file, skip=3 ) %>%
  mutate( SUB_MOD_RSD_CLN = str_replace_all(SUB_MOD_RSD, "([A-Z])(\\d+)", "\\1 \\2")) %>%
  separate( SUB_MOD_RSD_CLN, into=c("residue", "position")) 
```


```{r}
print("Read PSP regulatory sites table.")
reg_sites_tbl <- vroom::vroom( reg_sites_file, skip=3)  %>%
      mutate( MOD_RSD_CLN=  str_replace_all(  MOD_RSD, "([A-Z])(\\d+)-(.*)", "\\1 \\2 \\3") ) %>%
      separate( MOD_RSD_CLN, sep=" ", into=c("residue", "position", "ptm_type")) %>%
      relocate (ACC_ID, residue, position, ptm_type, .before="GENE" ) %>%
      dplyr::select(-`...21`) %>%
      dplyr::rename( REG_SITES_PMIDs = "PMIDs") %>%
      dplyr::rename( REG_SITES_NOTES = "NOTES")
  
```

```{r}
print( "Read PSP disease table")
disease_tbl <- vroom::vroom( disease_file, skip=3 ) %>%
  dplyr::select( ACC_ID, MOD_RSD, DISEASE, ALTERATION, NOTES ) %>%
  mutate( MOD_RSD_CLN=  str_replace_all(  MOD_RSD, "([A-Z])(\\d+)-(.*)", "\\1 \\2 \\3") ) %>%
  separate( MOD_RSD_CLN, sep=" ", into=c("residue", "position", "ptm_type")) %>%
  dplyr::rename( DISEASE_NOTES = "NOTES") %>%
  dplyr::select(-MOD_RSD, -ptm_type) 
```


## Read PTM tables
```{r}

print("Read PSP post-translational modification (PTM) tables")
list_of_ptm_file_names <- c("Acetylation_site_dataset",
"Methylation_site_dataset",
"O-GalNAc_site_dataset",
"O-GlcNAc_site_dataset",
"Phosphorylation_site_dataset",
"Sumoylation_site_dataset",
"Ubiquitination_site_dataset")

list_of_ptm_files <- purrr::map_chr( list_of_ptm_file_names, ~file.path(phosphosite_db_dir, .))

ptm_tbl <- vroom::vroom( list_of_ptm_files, skip=3, id="ptm")  %>%
  dplyr::select( ACC_ID, MOD_RSD, ptm) %>%
  mutate( MOD_RSD_CLN=  str_replace_all(  MOD_RSD, "([A-Z])(\\d+)-(.*)", "\\1 \\2 \\3") ) %>%
  separate( MOD_RSD_CLN, sep=" ", into=c("residue", "position", "ptm_type")) %>%
  mutate( uniprot_acc = ACC_ID) %>%
  dplyr::mutate( position = as.integer(position)) %>%
  mutate( ptm= purrr::map_chr( ptm, ~{ temp_vec <- str_split(., "/")[[1]]
  temp_vec[length(temp_vec)] } )) %>%
  dplyr::mutate( ptm = str_replace_all( ptm, "_site_dataset", ""))

```

## Find nearby PTM + or - five residues apart 
```{r}
print( paste0("Find other PTM nearby +/- ", near_ptm_num_residues, " amino acid wihtin the phosphorylation sites.") )
get_nearby_ptm <- de_phos_long  %>% 
  dplyr::distinct( sites_id, uniprot_acc,  position) %>%
  separate_rows( uniprot_acc, position, sep="\\:") %>%
  separate_rows( uniprot_acc, position, sep="\\|") %>%  
  separate_rows( uniprot_acc, position, sep=";") %>%  
  mutate( position = str_replace_all( position, "\\(|\\)", "") %>% purrr::map_int(as.integer)) %>%
  mutate( residue_window =  purrr::map(position,  ~seq( from=as.integer( .)-near_ptm_num_residues, to= as.integer( .)+near_ptm_num_residues, by=1 )) ) %>%
  unnest( residue_window) %>%
  left_join(ptm_tbl %>% 
              mutate(ptm_count=1) %>%
              dplyr::select(-residue), by=c( "uniprot_acc" = "uniprot_acc",
                                 "residue_window" = "position")) %>%
  distinct %>% 
  dplyr::filter( !( residue_window == position & ptm_type == "p") ) %>% ## Avoid counting the same phosphorylation site as the query itself
  dplyr::select(-MOD_RSD, -uniprot_acc, -position, -ptm_type)

nearby_ptm_count <- get_nearby_ptm %>%
  dplyr::select(-ACC_ID) %>%
  dplyr::filter( !is.na(ptm_count)) %>%
  group_by( across(.cols=setdiff(colnames(get_nearby_ptm), c("ptm_count", "residue_window", "ACC_ID") ))) %>%
  distinct() %>%
  summarise( ptm_count = sum(ptm_count) ) %>%
  ungroup  %>%
  distinct %>%
  pivot_wider( id_cols = sites_id,
               values_from = ptm_count,
               names_from="ptm",
               names_prefix = paste0("nearby_+/-", near_ptm_num_residues, "_") )


```

## Get number of phosphosites on the peptide, is it multiply phosphorylated
```{r}
num_phos_sites <- de_phos_long  %>% 
  dplyr::distinct( sites_id, position) %>%
  dplyr::mutate( num_sites =   str_split(position, ":") %>% 
                   purrr::map_chr(1) %>% 
                   str_split( "\\|")  %>% 
                 purrr::map_chr(1) %>% 
                   str_split(";") %>% 
                   purrr::map_int(length) ) %>%
  dplyr::select(-position)


```


```{r}
# de_phos_long %>% dplyr::filter( str_detect( uniprot_acc, "[^[:alnum:]]+" ) )

phosphosite_plus_tbl <- de_phos_long  %>% 
  dplyr::distinct( sites_id, uniprot_acc, residue, position, sequence) %>%
  separate_rows( uniprot_acc, position, sequence, residue, sep="\\:") %>%
  separate_rows( uniprot_acc, position, sequence, residue, sep="\\|") %>%  
  separate_rows( uniprot_acc, position, sequence, residue, sep=";") %>%  
  mutate( position = str_replace_all( position, "\\(|\\)", "") %>% purrr::map_int(as.integer)) %>%
  left_join (reg_sites_tbl %>%
               dplyr::filter(ptm_type == "p") %>%
               dplyr::mutate( position = purrr::map_int(position, as.integer)), 
             by=c("uniprot_acc" = "ACC_ID",
                                 "residue" = "residue",
                                 "position" = "position"
                                 )) %>%
  left_join( ks_tbl %>%
               dplyr::rename(KINASE_GENE = "GENE") %>%
               dplyr::select(-DOMAIN, - `SITE_+/-7_AA`) %>%
               dplyr::mutate( position = purrr::map_int(position, as.integer)), 
             by=c("uniprot_acc" = "SUB_ACC_ID",
                          "residue"="residue",
                          "position" = "position", 
                          "SITE_GRP_ID" = "SITE_GRP_ID"))   %>%
  left_join( disease_tbl %>%
               dplyr::mutate( position = purrr::map_int(position, as.integer)), 
             by=c( "uniprot_acc" = "ACC_ID",
                                 "residue" = "residue",
                                 "position" = "position")) %>%
  group_by(sites_id, uniprot_acc ) %>%
  summarise( across( .cols=everything()  , ~paste(unique(.), collapse="//"))   ) %>%
  ungroup() %>%
  group_by(sites_id ) %>%
  summarise( across( .cols=everything()  , ~paste(unique(.), collapse=":"))   ) %>%
  ungroup() 
 
# de_phos_long_annot %>% dplyr::filter( str_detect( KINASE, "//" ) )
#   
# # colnames( de_phos_long_annot)[ which( !is.na( str_match( colnames( de_phos_long_annot), "\\.y" ) )  ) ]
# 
# de_phos_long_annot %>%
#   dplyr::filter( !is.na(ptm_type) | !is.na( KINASE))

```


## Download Reactome annotation file 
```{r}

print("Download Reactome UniProt to pathways file.")

temp_file <- tempfile()

reactome_file <- download.file(url="https://reactome.org/download/current/UniProt2Reactome.txt", destfile=temp_file)

reactome_map <- vroom::vroom( temp_file , 
                              col_names = c("uniprot_acc", "reactome_id", "url", "reactome_term", "evidence", "organism") )




```

```{r}

print("Get the best UniProt accession per row.")

uniprot_acc_tbl <- de_phos_long %>%
  mutate( uniprot_acc_copy = uniprot_acc ) %>%
  separate_rows(uniprot_acc_copy, sep=":" ) %>%
  mutate( join_uniprot_acc = cleanIsoformNumber(uniprot_acc_copy)) %>%
  dplyr::distinct( uniprot_acc, join_uniprot_acc) %>%
  group_by( uniprot_acc) %>%
  mutate( acc_order_id = row_number()) %>% 
  ungroup


```


```{r}
print("Download information from UniProt.")
uniprot_dat <- NA
up <- NA
if( ! file.exists( file.path(results_dir, "uniprot_data.RDS"))) {

  if( is.na(up)) {
    up <- UniProt.ws(taxId=taxonomy_id )
  } 

 # keytypes(up)
  list_of_sp_columns <- c("EXISTENCE", "SCORE", "REVIEWED", "GENENAME", "PROTEIN-NAMES", "LENGTH", "ENSEMBL", "GO-ID", "KEYWORDS")

  uniprot_dat <- batchQueryEvidence(uniprot_acc_tbl, join_uniprot_acc, uniprot_handle=up,
                                    uniprot_columns = list_of_sp_columns)
  
  saveRDS( uniprot_dat, file.path(results_dir, "uniprot_data.RDS"))
  
} else {
  uniprot_dat <- readRDS( file.path(results_dir, "uniprot_data.RDS"))
}

```


```{r eval=FALSE}
 # uniprot_dat <- batch_query_evidence(uniprot_acc_tbl %>% head(100), best_uniprot_acc, uniprot_handle=up, 
 #                                      uniprot_columns = list_of_sp_columns)
```

```{r}

print("Merge with Gene Ontology terms.")
goterms <- Term(GOTERM)
gotypes <- Ontology(GOTERM)

tic()
uniprot_dat_cln <- uniprotGoIdToTerm(uniprot_dat, sep="; ", goterms, gotypes  )
toc()

uniprot_dat_multiple_acc <- uniprot_acc_tbl %>%
  left_join( uniprot_dat_cln, by=c("join_uniprot_acc" = "UNIPROTKB") )   %>%
  arrange( uniprot_acc, acc_order_id) %>%
  group_by(uniprot_acc ) %>%
  summarise( across( .cols=setdiff( colnames( uniprot_dat_cln), "UNIPROTKB")   , ~paste(., collapse=":"))   ) %>%
  ungroup()   %>%
  dplyr::rename( UNIPROT_GENENAME = "GENENAME")

uniprot_dat_multiple_acc

```

```{r}
print("Add reactome pathways annotation.")
reactome_term_tbl <- uniprot_acc_tbl %>%
  left_join( reactome_map, by=c("join_uniprot_acc" = "uniprot_acc") )   %>%
  dplyr::filter(reactome_term != "NA" ) %>%
  group_by(uniprot_acc, join_uniprot_acc) %>%
  summarise( reactome_term = paste(reactome_term, collapse="; ") ) %>%
  ungroup()     %>%
  mutate(reactome_term = str_replace_all( reactome_term , ":", "-")) %>%
  group_by(uniprot_acc ) %>%
  summarise( reactome_term = paste(reactome_term, collapse=":") ) %>%
  ungroup()   

```


## Commented out as output file is too large
```{r}
# print("Output wider format results table with protein annotation.")
# de_phos_wide_annot <- de_phos_wide %>%
#   left_join( num_phos_sites, by =c("sites_id" = "sites_id")) %>%
#   left_join( uniprot_dat_multiple_acc, by = c("uniprot_acc" = "uniprot_acc") ) %>%
#   left_join( reactome_term_tbl, by = c("uniprot_acc" = "uniprot_acc"))  %>%
#   left_join( phosphosite_plus_tbl %>%
#                dplyr::select(-uniprot_acc, -position, -residue, -sequence), 
#              by=c("sites_id" = "sites_id")) %>%
#   left_join(  abundance_tbl %>%
#    dplyr::select( sites_id, maxquant_row_ids ), 
#    by=c("sites_id" = "sites_id")) %>%
#   relocate(maxquant_row_ids, .after="sites_id") %>%
#   left_join( get_nearby_ptm, 
#              by=c("sites_id" = "sites_id")) %>%
#   distinct()
# 
# head( de_phos_wide_annot ) 
# 
# vroom::vroom_write(de_phos_wide_annot, path=output_wide_file ) 

```


```{r}
print("Output longer format results table with protein annotation.")
de_phos_long_annot <- de_phos_long %>%
  left_join( num_phos_sites, by =c("sites_id" = "sites_id")) %>%
  left_join( uniprot_dat_multiple_acc, by = c("uniprot_acc" = "uniprot_acc") ) %>%
  left_join( reactome_term_tbl, by = c("uniprot_acc" = "uniprot_acc"))  %>%
  left_join( phosphosite_plus_tbl %>%
               dplyr::select(-uniprot_acc, -position, -residue, -sequence), 
             by=c("sites_id" = "sites_id"))  %>%
  left_join(  abundance_tbl %>%
   dplyr::select( sites_id, maxquant_row_ids ), 
   by=c("sites_id" = "sites_id")) %>%
  relocate(maxquant_row_ids, .after="sites_id")  %>%
  left_join( nearby_ptm_count, 
             by=c("sites_id" = "sites_id")) %>%
  distinct()

head( de_phos_long_annot )

vroom::vroom_write(de_phos_long_annot, path=output_long_file ) 

```


```{r}
toc()
```

