---
title: "ALPK1 RPE and NR analysis"
output:
  html_document:
    df_print: paged
---
```{r}
#! /usr/bin/env Rscript

# Author(s): Ignatius Pang
# Email: ipang@cmri.org.au
# Childrenâ€™s Medical Research Institute, finding cures for childhood genetic diseases
```

To investigate if RUV will overfit data.

## Packages installation and loading
```{r}

options(knitr.duplicate.label = "allow")


#Test if BioManager is installed 
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
   BiocManager::install(version = "3.12")
}

# load pacman package manager
if(!require(pacman)){
    install.packages("pacman")
    library(pacman)
}

p_load(optparse)
p_load(tictoc)
```


## Parameters
```{r}

tic()
group_pattern <- "RPE" 
analysis_type <-  group_pattern # "RPE_Test_RUV" #


## Directories management 
base_dir <- here::here()
data_dir <- file.path( base_dir, "Data", "Abundance_Data", "P90")
results_dir <- file.path(base_dir, "Results", paste0( analysis_type, "90") ,  "Proteins", "DE_Analysis")
source_dir <- file.path(base_dir, "..")
```


## Command to convert the R markdown file to a R file for use in command line. 
knitr::purl( file.path(  source_dir, "Shared",  "de_analysis_cmd.Rmd"), 
             output=file.path( source_dir,  "Shared",   "de_analysis_cmd.R") ) 
       


```{r}
# Minimum number of samples per experimental group for the protein to be accepted for analysis 
# min_num_samples_per_group <- 4
max_num_samples_miss_per_group <- NA
abundance_threshold <- 0
column_pattern <- "Reporter intensity corrected \\d+ RPE"
## q-value threshold (protein logFC's q-value must be below this value to be included in further analysis)
q_val_thresh <- 0.05
ruv_k <- 4
num_neg_ctrl <- 500
ruv_method <- "ruv3"
design_matrix_file <- file.path(results_dir, "design_matrix_cleaned.tab")
counts_table_file <- file.path(results_dir, "raw_counts_table.tab")
# X is the control for Y and A is the control for B because of the different mouse strains used as WTs, but it might be useful just to compare everything to everything.
#test_pairs_file <- file.path( data_dir, "test_pairs.tab")
contrasts_file <- file.path( data_dir, "contrast_strings_RPE.tab")
formula_string <- "~ 0 + group "
my_sample_id <- "Sample_ID"
my_group_id <- "group"
my_row_id <- "uniprot_acc"
#limma_method <- "contrasts" # pairs or contrasts
file_prefix <- "de_proteins"


# min_num_samples_per_group <- 4
# abundance_threshold <- 0
# column_pattern <- "Reporter intensity corrected \\d+"
# group_pattern <- ""
# q_val_thresh <- 0.05
# ruv_k <- 4
# num_neg_ctrl <- 500
# ruv_method <- "ruv3"
# counts_table_file <- "/home/ignatius/PostDoc/2021/Podocyte_2021/Results/Part_1/Proteins/DE_Analysis/raw_counts_table_cleaned.tab"
# test_pairs_file <- ""
# contrasts_file <- "/home/ignatius/PostDoc/2021/Podocyte_2021/Data/Abundance_Data/Part_1/contrast_strings.tab"
# formula_string <- "~ 0 + group "
# design_matrix_file <- "/home/ignatius/PostDoc/2021/Podocyte_2021/Results/Part_1/Proteins/DE_Analysis/design_matrix.tab"
# results_dir <- "/home/ignatius/PostDoc/2021/Podocyte_2021/Results/Part_1/Proteins/DE_Analysis"
# my_sample_id <- "Sample_ID"
# my_group_id <- "group"
# my_row_id <- "uniprot_acc"
# limma_method <- "contrasts"
# file_prefix <- "de_proteins"

command_line_options <- commandArgs(trailingOnly = TRUE)
if( length(command_line_options ) > 0 ) {
  
  parser <- OptionParser(add_help_option =TRUE)
  
  # parser <- add_option(parser, c("-s", "--min-samples"), type="integer", default=4, dest = "min_num_samples_per_group",
  #                      help="Minimum number of samples per experimental group for the protein to be accepted for analysis [default %default]",
  #                      metavar="integer")
  
    parser <- add_option(parser, c( "--max-missing"), type="integer", default=NA, dest = "max_num_samples_miss_per_group",
                       help="Remove protein if it exceeds this maximum number of samples with missing values per experimental group [default %default]",
                       metavar="integer")
  
  parser <- add_option(parser, c("-a", "--abundance-thresh"), type="integer", default=0, dest = "abundance_threshold",
                       help="Abundance threshold above which the protein in the sample is accepted for analysis [default %default]",
                       metavar="integer")
  
  parser <- add_option(parser, c("-p", "--column-pattern"), type="character", default="Reporter intensity corrected \\d+ ", dest = "column_pattern",
                       help="Regular expression pattern to identify column with abundance values [default %default]",
                       metavar="string")
  
  parser <- add_option(parser, c( "--group-pattern"), type="character", default="", dest = "group_pattern",
                       help="Regular expression pattern to identify columns with abundance values belonging to the experiment. [default %default]",
                       metavar="string")
  
  parser <- add_option(parser, c("-q", "--q-value-thresh"), type="double", default=0.05, dest = "q_val_thresh",
                       help="q-value threshold below which a protein has statistically significant differetial expression [default %default]",
                       metavar="double")
  
  parser <- add_option(parser, c( "-k", "--ruv-k"), type="integer", default=7, dest = "ruv_k",
                       help="The number of unwanted factors to use.  [default %default]",
                       metavar="integer")
  
  parser <- add_option(parser, c( "-n", "--num-neg-ctrl"), type="integer", default=500, dest = "num_neg_ctrl",
                       help="The number of negative control proteins to use.  [default %default]",
                       metavar="number")
  
  parser <- add_option(parser, c("-m", "--ruv-method"), type="character", default="ruv3", dest = "ruv_method",
                       help="A string representing the ruv3 method to use.  [default %default]",
                       metavar="character")
  
  parser <- add_option(parser, c("-c", "--counts"), type="character", default="", dest = "counts_table_file",
                       help="Input file with the protein abundance values",
                       metavar="string")
  
  parser <- add_option(parser, c("-t", "--test-pairs"), type="character", default="", dest = "test_pairs_file",
                       help="Input file with a table listing all the pairs of experimental groups to compare. First column represents group A and second column represents group B. Linear model comparisons (e.g. Contrasts) would be group B minus group A.",
                       metavar="string")
  
  parser <- add_option(parser, c("--contrasts"), type="character", default="", dest = "contrasts_file",
                       help="Input file with a table listing all comparisons to be made in string, one comparison per line (e.g. groupB.vs.group_A = groupB - groupA).",
                       metavar="string")  
  
  parser <- add_option(parser, c("--formula"), type="character", default="", dest = "formula_string",
                       help="A string representing the formula for input into the model.frame function. (e.g. ~ 0 + group).",
                       metavar="string")  
  
  parser <- add_option(parser, c("-d", "--design-matrix"), type="character", default="", dest = "design_matrix_file",
                       help="Input file with the design matrix",
                       metavar="string")
  
  parser <- add_option(parser, c("-o", "--output-dir"), type="character", default="", dest = "results_dir",
                       help="Directory path for all results files.",
                       metavar="string")
  
  parser <- add_option(parser, c("--sample-id"), type="character", default="Sample_ID", dest = "my_sample_id",
                       help="A string describing the sample ID. This must be a column that exists in the design matrix.",
                       metavar="string") 
  
  parser <- add_option(parser, c("-g", "--group-id"), type="character", default="group", dest = "my_group_id",
                       help="A string describing the experimental group ID. This must be a column that exists in the design matrix.",
                       metavar="string") 
  
  parser <- add_option(parser, c("-r", "--row-id"), type="character", default="uniprot_acc", dest = "my_row_id",
                       help="A string describing the row id.",
                       metavar="string") 
  
  # parser <- add_option( parser, c("--limma-method"), type="character", default="contrasts", dest = "limma_method",
  #                      help="A string to select which strategy is employed to use limma.",
  #                      metavar="string")

  parser <- add_option( parser, c("--prefix"), type="character", default="de_proteins", dest = "file_prefix",
                       help="A string to indicate the type of analysis and is used in the file name of the output results table.",
                       metavar="string")
  
  print(commandArgs(trailingOnly = TRUE))
  
  cmd_arguments <- parse_args(parser)
  
  print(cmd_arguments)
  
  min_num_samples_per_group <- cmd_arguments$min_num_samples_per_group
  max_num_samples_miss_per_group <- cmd_arguments$max_num_samples_miss_per_group
  abundance_threshold       <- cmd_arguments$abundance_threshold
  column_pattern            <- cmd_arguments$column_pattern
  group_pattern             <- cmd_arguments$group_pattern
  q_val_thresh              <- cmd_arguments$q_val_thresh
  ruv_k                     <- cmd_arguments$ruv_k
  num_neg_ctrl              <- cmd_arguments$num_neg_ctrl
  ruv_method                <- cmd_arguments$ruv_method
  design_matrix_file        <- cmd_arguments$design_matrix_file
  counts_table_file         <- cmd_arguments$counts_table_file
  test_pairs_file           <- cmd_arguments$test_pairs_file
  contrasts_file            <- cmd_arguments$contrasts_file
  formula_string            <- cmd_arguments$formula_string
  results_dir               <- cmd_arguments$results_dir
  my_sample_id              <- cmd_arguments$my_sample_id
  my_group_id               <- cmd_arguments$my_group_id
  my_row_id                 <- cmd_arguments$my_row_id
  # limma_method              <- cmd_arguments$limma_method
  file_prefix               <- cmd_arguments$file_prefix
  
  # if(limma_method != "pairs" & limma_method != "contrasts") { stop("The variable --limma-method must be 'pairs' or 'contrasts'.") }
  if( group_pattern == "") { warning("No --group-pattern variable supplied."); 
    group_pattern <- "\\d+"}
  if( column_pattern == "") { stop("No --column-pattern variable supplied.") }
  if( design_matrix_file == "" ) { stop("No file --design-matrix was supplied. It is a file describing the sample names and treatment groups.") }
  if( formula_string == "" ) { stop("No file --formula was supplied.") }
  if( counts_table_file == "" ) { stop("No value for --counts was supplied. It is a file with the abundance values for the proteins for each sample.") }
  if( results_dir == "" ) { stop("No value for --output-dir was supplied.") }
  # if( limma_method == "pairs" & test_pairs_file == "") { stop("No value for --test-pairs was supplied. It is a file listing all pairs of groups to compare.") }
  if( contrasts_file=="" ) { stop("No value for --contrasts was supplied. It is a file listing all contrasts for analysis.") }
  
}

if( group_pattern == "") {    group_pattern <- "\\d+"}

```

```{r}
p_load(tidyverse)
p_load(plotly)
p_load(vroom)
p_load(ggplot2)
p_load(ggpubr)
p_load(magrittr)
p_load(knitr)
p_load(rlang)


p_load(limma)
p_load(qvalue)
p_load(ruv)
p_load(mixOmics)

p_load(ProteomeRiver)

```

## Create directories
```{r}
print ("Create results directory if it does not yet exists.")
create_dir_if_not_exists( results_dir)

```

## Read in test pairs file 
```{r}

# test_pairs <- NA
# if( limma_method == "pairs" & test_pairs_file != "") {
#   print("Read file with pairs of experimental groups to test.")
#   test_pairs <- vroom::vroom( test_pairs_file)
# }



  print("Read file with lists of experimental contrasts to test.")
  contrasts_tbl <- vroom::vroom( contrasts_file, delim="\t")

```

```{r}

print("Read file with counts table.")
evidence_tbl_filt <- vroom::vroom( counts_table_file)  %>%
  dplyr::select( one_of(c(my_row_id)),   matches(group_pattern)  )


if( ! my_row_id %in% colnames( evidence_tbl_filt)) {
  stop( "The row ID specified in --row-id was not found in the input counts file.")
}

if( group_pattern != "" & 
    length(which( str_detect(  setdiff( colnames( evidence_tbl_filt), my_row_id), group_pattern))) == 0 )  {
  stop( "The input counts file did not contain any columns that matches the string provided in --group-pattern.")
}

```
## Remove peptides without abundance values at all 
```{r}
print("Remove empty proteins without abundance data.")
cln_dat_wide_unsorted <- ProteomeRiver::remove_empty_rows( evidence_tbl_filt, 
                                                  col_pattern=group_pattern, 
                                                  row_id=!!rlang::sym(my_row_id))

```

## Create the design matrix
```{r}

design_mat_cln <- vroom::vroom(design_matrix_file) %>% as.data.frame() %>%
  dplyr::mutate( !!rlang::sym( my_sample_id) := as.character( !!rlang::sym( my_sample_id) ))

rownames( design_mat_cln) <- design_mat_cln %>% pull ( as.name( my_sample_id))

## Check that the sample ID and group ID name is found in the design matrix 
if(   length( which( c(my_sample_id, my_group_id ) %in% colnames( design_mat_cln) )) != 2   ) {
    stop("Sample ID and group ID are not matching to the column names used in the design matrix.")
}

cols_for_analysis <- design_mat_cln %>% pull(as.name( my_sample_id))

```

## Remove row with any missing values
```{r}
cln_dat_wide_cleaned <- NA
if( !is.na(max_num_samples_miss_per_group) ) {
  print("Remove row with any missing values")
  
cln_dat_wide_cleaned <- remove_rows_with_missing_values( cln_dat_wide_unsorted, 
                                            matches(group_pattern), 
                                            design_mat_cln %>% 
                                              dplyr::mutate( Sample_ID = as.character(Sample_ID)), 
                                            !!rlang::sym(my_sample_id), 
                                            !!rlang::sym(my_row_id), 
                                            !!rlang::sym(my_group_id), 
                                            max_num_samples_miss_per_group, 
                                            abundance_threshold)

} else {
  cln_dat_wide_cleaned <- cln_dat_wide_unsorted
}

```

### Check that the column names in design matrix matches those in the abundancet able. Create a row ID column.
```{r}

if( length(cln_dat_wide_cleaned) - 1 != length(cols_for_analysis) |

    length(cols_for_analysis) != length(intersect(colnames( cln_dat_wide_cleaned )[-1], cols_for_analysis )) ) {
  stop("Column names in design matrix table not consistent with column names in abundance table.")
}

# Sort the columns in the abundance table according
if(  length(which(colnames( cln_dat_wide_cleaned )[-1] == cols_for_analysis )) != length(cols_for_analysis ) ) {
  warning("Sorting the order of the column of the abundance matrix according to the order in the design matrix.") 
}

cln_dat_wide <- cln_dat_wide_cleaned[, c(colnames(cln_dat_wide_cleaned)[1], cols_for_analysis)]

print("Assign the indexing row names to the data frame.")
counts_filt <- cln_dat_wide %>%
  column_to_rownames(my_row_id)

vroom::vroom_write( cln_dat_wide, file.path( results_dir, "raw_counts_after_removing_empty_rows.tsv"))


```

## Functions to create the replicate matrix for RUVs
```{r}

ruvIII_replicates_matrix <- get_ruvIII_replicate_matrix( design_mat_cln,  
                                                         !!rlang::sym(my_sample_id), 
                                                         !!rlang::sym(my_group_id))

ruvIII_replicates_matrix
```

## Get a list of rows (e.g. proteins) with too many missing values
Return a list, names are the sample groups,  each element is the list of proteins kept since it meets the minimum samples per group
```{r}

# sample_rows_lists <- get_rows_to_keep_list( cln_dat_wide, 
#                                             matches(group_pattern), 
#                                             design_mat_cln %>% 
#                                               dplyr::mutate( Sample_ID = as.character(Sample_ID)), 
#                                             !!rlang::sym(my_sample_id), 
#                                             !!rlang::sym(my_row_id), 
#                                             !!rlang::sym(my_group_id), 
#                                             min_num_samples_per_group, 
#                                             abundance_threshold)
# 
# saveRDS(sample_rows_lists, file.path(results_dir, "keep_sample_rows_lists.RDS" ))

```

## Count number of missing or zero values, plot graph
No missing values detected.
```{r}
print("Count the number of missing values for each sample.")
## Count the total number of missing values in total
table(is.infinite(data.matrix(log2(counts_filt))))

plot_num_missing_values <- get_plot_num_missing_vales(counts_filt[,cols_for_analysis])

plot_num_missing_values

ggsave(filename=file.path(results_dir,  "num_missing_values.png"), plot=plot_num_missing_values)
ggsave(filename=file.path(results_dir,  "num_missing_values.svg"), plot=plot_num_missing_values)

```

## Mark entries with zero values as NA
```{r}
print("Mark entries with zero values as NA, take log of values.")
na_values_marker <- (counts_filt == 0)

counts_na <- counts_filt

counts_na[na_values_marker ] <- NA

counts_na.log <- log2(counts_na)
```

## Median centering to enable normal distribution calculations for missing values imputation
```{r}
print("Median centering.")
counts_na.log.quant <- normalizeBetweenArrays(counts_na.log,  method="scale")


vroom::vroom_write( as.data.frame(counts_na.log.quant) %>% 
                      rownames_to_column(my_row_id), 
                    file.path( results_dir, "counts_after_median_scaling_before_imputation.tsv"))


```

## Missing values imputation approach 2: Fill with values drawn from normal distribution, mean and standard deviation down-sized
```{r}

print("Missing value imputation")

counts_rnorm.log.quant <- counts_na.log.quant %>%
  as.data.frame %>%
  mutate_all( ~{impute_per_col(., width = 0.3, downshift = 1.8)} ) %>%
  as.matrix

counts_rnorm.is_nan <- counts_rnorm.log.quant %>% 
  as.data.frame %>% 
  mutate_all(  ~{!is.finite(.)} ) %>%
  as.matrix


vroom::vroom_write( as.data.frame(counts_rnorm.log.quant) %>% 
                      rownames_to_column(my_row_id), 
                    file.path( results_dir, "counts_after_median_scaling_and_imputation.tsv"))


```

## Run statistical tests without RUV
```{r}
print("Run statistical tests without RUV.")

ID <- rownames(counts_rnorm.log.quant)

type_of_grouping <- get_type_of_grouping( design_matrix=design_mat_cln, group_id=my_group_id, sample_id=my_sample_id)

list_rnorm.log.quant.ruv.r0 <- NA
myRes_rnorm.log.quant <- NA

# if( limma_method == "pairs") {
#   
#   list_rnorm.log.quant.ruv.r0 <- runTests(ID, counts_rnorm.log.quant, test_pairs,  cols_for_analysis, sample_rows_lists, type_of_grouping, 
#                                   design_matrix = design_mat_cln, formula_string= paste0("~ 0 + ", my_group_id),
#                                   contrast_variable=my_group_id,
#                                   weights=NA) 
#   
#   myRes_rnorm.log.quant <- extract_results(list_rnorm.log.quant.ruv.r0)
# 
# } else if ( limma_method == "contrasts") {
  

  list_rnorm.log.quant.ruv.r0 <- runTests_contrasts( counts_rnorm.log.quant, 
                                                     contrast_strings = contrasts_tbl[,1][[1]], 
                                                     design_matrix = design_mat_cln, 
                                                     formula_string = formula_string,  
                                                     weights=NA )  
  
  myRes_rnorm.log.quant <- list_rnorm.log.quant.ruv.r0$results
# }



```
## save the lists of differentially abundant proteins, identified without using RUV, into output files 
```{r}
# print( "Save the list of differentially expressed proteins, without using RUV.")
# save_de_protein_list(myRes_rnorm.log.quant, 
#                      row_id="protein_id",
#                      sort_by_column =q.mod, 
#                      results_dir = results_dir, 
#                      file_suffix = "_without_ruv.tsv")

```


## Get first set of control genes for RUV analysis
This is the set of proteins without statistically significant differential expression in all pairs of comparisons.
```{r}
print( "Find the list of negative control genes using ANOVA.")
control_genes_index <- get_neg_ctrl_prot_anova( counts_rnorm.log.quant, 
                                                design_matrix = design_mat_cln, 
                                                group_column = "group", 
                                                num_neg_ctrl = num_neg_ctrl, 
                                                q_val_thresh = q_val_thresh)

vroom::vroom_write(  data.frame( temp_col = names(control_genes_index), 
                                is_control_genes=control_genes_index) %>%
                       set_colnames(c(my_row_id,  "is_control_genes")), 
                     file.path(results_dir, "ctrl_genes_list_ruv3.tsv"), 
                     delim="\t") 

print("Total Num. Genes")
length( control_genes_index) 

print("Num. Control Genes")
length( which(control_genes_index) ) 

```

## Canonical correlation plot round 1
```{r}
print("Draw canonical correlation plot.")
ruv_groups <- data.frame( temp_column = colnames( counts_rnorm.log.quant)) %>%
  dplyr::rename( !!rlang::sym( my_sample_id) := "temp_column")  %>%
  left_join( design_mat_cln %>%
  dplyr::mutate( !!rlang::sym( my_sample_id) := as.character( !!rlang::sym( my_sample_id) ) ), 
  by =  my_sample_id) 

cancorplot_r1 <- ruv_cancorplot(t(counts_rnorm.log.quant) , 
                                X = ruv_groups %>% pull( !!rlang::sym(my_group_id)), 
                                ctl = control_genes_index) 

cancorplot_r1

ggsave( plot=cancorplot_r1, filename=file.path(results_dir, "cancor_plot_round_1.pdf"))
ggsave( plot=cancorplot_r1, filename=file.path(results_dir, "cancor_plot_round_1.png"))


```

### Run RUVIII first on all samples
```{r}
print( "Run RUVIII on the input counts table.")
counts_rnorm.log.ruvIII_v1 <- my_RUVfit( counts_rnorm.log.quant, X=ruv_groups$group, control_genes_index, Z = 1, k = ruv_k, 
                             method = ruv_method, M=ruvIII_replicates_matrix )  %>% t()

vroom::vroom_write(counts_rnorm.log.ruvIII_v1 %>% 
                     as.data.frame %>% 
                     rownames_to_column(my_row_id), 
                   file.path(results_dir, "normalized_counts_after_ruv.tsv"  ) )

```

## Run statistical test after running RUV the first time (we will use this result to get a second set of control genes) 
```{r}
print("Compare the different experimental groups and obtain lists of differentially expressed proteins.")

list_rnorm.log.quant.ruv.r1 <- NA
myRes_rnorm.log.quant.ruv.r1 <- NA
# if( limma_method == "pairs") {
#   list_rnorm.log.quant.ruv.r1 <- runTests( ID, counts_rnorm.log.ruvIII_v1, test_pairs,  
#                                            cols_for_analysis, sample_rows_lists,
#                                            type_of_grouping, 
#                                            design_matrix = design_mat_cln, 
#                                            formula_string= "~ 0 + group ",
#                                            contrast_variable="group",
#                                            weights=NA)
#   
#   myRes_rnorm.log.quant.ruv.r1 <- extract_results(list_rnorm.log.quant.ruv.r1)
# 
# } else if ( limma_method == "contrasts") {

  list_rnorm.log.quant.ruv.r1 <- runTests_contrasts( counts_rnorm.log.ruvIII_v1, 
                                                     contrast_strings = contrasts_tbl[,1][[1]], 
                                                     design_matrix = design_mat_cln, 
                                                     formula_string = "~ 0 + group ",  
                                                     weights=NA )
  
 myRes_rnorm.log.quant.ruv.r1 <- list_rnorm.log.quant.ruv.r1$results

# }


```

## save the lists of differentially abundant proteins, identified after first round of RUV, into output files 
```{r}
# print("Save the lists of diffierentially expressed proteins (e.g. log fold-change, q-values).")
# save_de_protein_list(myRes_rnorm.log.quant.ruv.r1, 
#                      row_id="protein_id",
#                      sort_by_column =q.mod, 
#                      results_dir = results_dir, 
#                      file_suffix = "_round_1_ruv.tsv")


```



## PCA plots and RLE plots, compare different approaches for data analysis
```{r}
print("Draw the RLE and PCA plots.")
rle_pca_plots_arranged <- rle_pca_plot_list( list_of_data_matrix=list(counts_rnorm.log.quant, counts_rnorm.log.ruvIII_v1), 
                                             design_matrix = design_mat_cln, 
                               sample_id_column=!!rlang::sym(my_sample_id), 
                               group_column=!!rlang::sym(my_group_id),
                               list_of_descriptions=list( "Before RUVIII", "After RUVIII") )

rle_pca_plots_arranged

pdf(file.path(results_dir,  "rle_pca_plots.pdf") )

rle_pca_plots_arranged

dev.off()

```

## Select the data for volcano plot and p-values distribution plot 
```{r}
print("Prepare data for drawing the volcano plots")
selected_data <- get_significant_data(  list_of_de_tables = list( myRes_rnorm.log.quant,  
                                                                  myRes_rnorm.log.quant.ruv.r1),
                                        list_of_descriptions = list("No RUV",  
                                                                    "RUV applied" ), 
                                        row_id = !!sym(my_row_id), 
                                        p_value_column = p.mod, 
                                        q_value_column = q.mod, 
                                        log_q_value_column = lqm,
                                        log_fc_column = logFC, 
                                        comparison_column = comparison, 
                                        expression_column = expression,
                                        facet_column = analysis_type, 
                                        q_val_thresh = 0.05)  %>%
                  dplyr::rename( log2FC = "logFC") 

## Write all the results in one single table 
selected_data %>% 
  dplyr:::select(- colour, -lqm) %>%
  vroom::vroom_write(path=file.path(results_dir, "lfc_qval_long.tsv" ) ) 

```


## Volcano plots
```{r}

print("Print the volcano plots")

volplot_gg.all <- print_volcano_plot (  selected_data, 
                       log_q_value_column = lqm, 
                       log_fc_column = log2FC, 
                       q_val_thresh = 0.05, 
                       formula_string="analysis_type ~ comparison" )


volplot_gg.all

ggsave(  filename=file.path( results_dir, "volplot_gg.all.png"), plot=volplot_gg.all, width=7.29, height=6   ) 
ggsave(  filename=file.path( results_dir, "volplot_gg.all.svg"), plot=volplot_gg.all, width=7.29, height=6   ) 

```


## Count the number of Up or Down DE genes 
```{r}
print("Count the number of up or down significnat differentially expressed proteins.")

num_sig_de_genes<- print_count_de_genes_table(  list_of_de_tables = list( myRes_rnorm.log.quant,  
                                                                                   myRes_rnorm.log.quant.ruv.r1), 
                                                         list_of_descriptions = list("No RUV",  
                                                                                     "RUV applied" ), 
                                                         formula_string="analysis_type ~ comparison" )

num_sig_de_genes$plot

ggsave( filename = file.path( results_dir, "num_de_genes_barplot.png" ), 
        plot=num_sig_de_genes$plot,
        height = 10, 
        width = 7)

ggsave( filename = file.path( results_dir, "num_de_genes_barplot.svg" ), 
        plot=num_sig_de_genes$plot,
        height = 10, 
        width = 7)

num_sig_de_genes$table

vroom::vroom_write( num_sig_de_genes$table, 
                    path=file.path( results_dir,
                                    "num_significant_de_genes_all.tab") )

```



## P-values distribution
```{r}
print("Print p-values distribution figure")
pvalhist <- print_p_values_distribution(selected_data, 
                                        p_value_column = p.mod, 
                                        formula_string="analysis_type ~ comparison" ) 

pvalhist

ggsave( filename = file.path( results_dir, "p_values_distn.png" ), 
        plot=pvalhist,
        height = 10, 
        width = 7)

ggsave( filename = file.path( results_dir, "p_values_distn.svg" ), 
        plot=pvalhist,
        height = 10, 
        width = 7)

```

## all log fold-change,  p-values, normalized values,  in a pivot wider table 

```{r}

norm_counts <- counts_rnorm.log.ruvIII_v1 %>%
  as.data.frame  %>%
  set_colnames(  paste0( colnames( counts_rnorm.log.ruvIII_v1 ), ".log2norm") ) %>%
  rownames_to_column(my_row_id)


raw_counts <- counts_filt  %>%
  as.data.frame  %>%
  set_colnames(  paste0( colnames( counts_filt ), ".raw") ) %>%
  rownames_to_column(my_row_id)

de_proteins_wide <- selected_data %>%
  dplyr::filter( analysis_type == "RUV applied") %>%
  dplyr::select(-lqm, -colour, -analysis_type) %>%
  pivot_wider( id_cols = c( !!sym(my_row_id) ),
               names_from = c(comparison) ,
               names_sep = ":",
               values_from = c( log2FC, q.mod, p.mod))  %>%
  left_join( norm_counts, by=my_row_id) %>%
  left_join( raw_counts, by=my_row_id)

head(de_proteins_wide)

vroom::vroom_write(de_proteins_wide, path=file.path(results_dir, paste0(file_prefix, "_wide.tsv" ) ) )

```


## Result tables, different contracts represented by different set of rows. The normalized and raw counts represented as columns.
```{r}

norm_counts <- counts_rnorm.log.ruvIII_v1 %>%
  as.data.frame  %>%
  rownames_to_column(my_row_id) %>%
  pivot_longer( cols=matches(group_pattern),
                names_to="Sample_ID",
                values_to="log2norm") %>%
  left_join( design_mat_cln, by="Sample_ID" ) %>%
  mutate( temp_id = Sample_ID ) %>%
  separate( temp_id, sep="_", into=c("sample_number", "group_pattern")) %>%
  group_by(!!sym(my_row_id), group) %>%
  arrange(!!sym(my_row_id), group, sample_number) %>%
  mutate( replicate_number = paste0("log2norm.", row_number()) ) %>%
  ungroup %>%
  pivot_wider(id_cols=c(!!sym(my_row_id), group),
              names_from =replicate_number, 
              values_from= log2norm)


raw_counts <- counts_filt  %>%
  as.data.frame  %>%
  rownames_to_column(my_row_id) %>%
  pivot_longer( cols=matches(group_pattern),
                names_to="Sample_ID",
                values_to="raw") %>%
  left_join( design_mat_cln, by="Sample_ID" ) %>%
  mutate( temp_id = Sample_ID ) %>%
  separate( temp_id, sep="_", into=c("sample_number", "group_pattern")) %>%
  group_by(!!sym(my_row_id), group) %>%
  arrange(!!sym(my_row_id), group, sample_number) %>%
  mutate( replicate_number = paste0("raw.", row_number()) ) %>%
  ungroup %>%
  pivot_wider(id_cols=c(!!sym(my_row_id), group),
              names_from =replicate_number, 
              values_from= raw)

left_join_columns <- rlang::set_names( c(my_row_id, "group")  ,
                                       c(my_row_id,   "left_group" ))

right_join_columns <- rlang::set_names( c(my_row_id,  "group")  ,
                                       c(my_row_id, "right_group"))

de_proteins_long <-  selected_data %>%
  dplyr::filter( analysis_type == "RUV applied") %>%
  dplyr::select(-lqm, -colour, -analysis_type) %>%
  dplyr::mutate( expression = str_replace_all(expression, "group", "")) %>%
  separate( expression, sep="-", into=c("left_group", "right_group")) %>%
  left_join( norm_counts, by=left_join_columns) %>%
  left_join( norm_counts, by=right_join_columns,
               suffix=c(".left", ".right") ) %>%
  left_join( raw_counts, by=left_join_columns) %>%
  left_join( raw_counts, by=right_join_columns,
               suffix=c(".left", ".right") )


head( de_proteins_long ) 


vroom::vroom_write(de_proteins_long, path=file.path(results_dir, paste0(file_prefix, "_long.tsv" ) ) ) 

```



```{r}
sessionInfo()

stop_time <- toc()
```


```{r}
sink(file.path(results_dir, paste0(file_prefix, "_cmd.log")))
print(date())
print(stop_time)
print(cmd_arguments)
tools::md5sum( c(file.path(results_dir, paste0(file_prefix, "_wide.tsv" ) ) , 
          file.path(results_dir, paste0(file_prefix, "_long.tsv" ) ) ,
          file.path(results_dir, paste0("normalized_counts_after_ruv.tsv"))
          ))
sessionInfo()
sink()
```




