```{r}
#!/usr/bin/env Rscript

# Author(s): Ignatius Pang, Pablo Galaviz
# Email: cmri-bioinformatics@cmri.org.au
# Childrenâ€™s Medical Research Institute, finding cures for childhood genetic diseases

## ---------------------------------------------------------------------------------------------------------------------------------------------------

#Test if BioManager is installed
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
   BiocManager::install(version = "3.12")
}

# load pacman package manager
if(!require(pacman)){
    install.packages("pacman")
    library(pacman)
}

p_load(tidyverse)
p_load(vroom)
p_load(BiocParallel)
p_load(magrittr)
p_load(KinSwingR)
p_load(RColorBrewer)
p_load(ggrepel)

p_load(optparse)
p_load(configr)
p_load(logging)
p_load(svglite)
p_load(tictoc)
p_load(ProteomeRiver)
p_load(UniProt.ws)

tic()


## ---------------------------------------------------------------------------------------------------------------------------------------------------

command_line_options <- commandArgs(trailingOnly = TRUE)
#Note: options with default values are ignored in the configuration file parsing.

parser <- OptionParser(add_help_option = TRUE)

#Note: options with default values are ignored in the configuration file parsing.
parser <- add_option(parser, c("-d", "--debug"), action = "store_true", default = FALSE,
                     help = "Print debugging output")

parser <- add_option(parser, c("-s", "--silent"), action = "store_true", default = FALSE,
                     help = "Only print critical information to the console.")

parser <- add_option(parser, c("-c", "--config"), type = "character", default = "/home/ignatius/PostDoc/2021/Podocyte_2021/Source/Part_1/config_phos.ini",
                     help = "Configuration file.",
                     metavar = "string")

parser <- add_option(parser, c("-t","--tmp_dir"), type = "character", default = "cache", dest = "tmp_dir",
                     help = "Directory path for temporary files.",
                     metavar = "string")

parser <- add_option(parser, c( "--num_cores"), type = "integer", default = 1,
                     help = "The number of cores used for the computation.",
                     metavar = "integer")

parser <- add_option(parser, c( "--random_seed"), type = "integer", default = 123456,
                     help = "The number of cores used for the computation.",
                     metavar = "integer")

parser <- add_option(parser, c( "--motif_score_iteration"), type = "integer", default = 1000,
                     help = "The number iterations for scoring each substrate against kinase motifs.",
                     metavar = "integer")

parser <- add_option(parser, c( "--swing_iteration"), type = "integer", default = 1000,
                     help = "The number iterations for swing score calcualtions.",
                     metavar = "integer")

parser <- add_option(parser, c( "--min_num_sites_per_kinase"), type = "integer", default = 10,
                     help = "The minimum number of known substrates for each kinase",
                     metavar = "integer")

parser <- add_option(parser, c( "--p_value_cutoff"), type = "double", default = 0.05,
                     help = "The p-value cutoff for identifying a kinase as significant by KinSwingR",
                     metavar = "integer")


parser <- add_option(parser, c( "--p_value_cutoff"), type = "double", default = 0.05,
                     help = "The p-value cutoff for identifying a kinase as significant by KinSwingR",
                     metavar = "integer")

out_tab_lfc_p_value_cutoff <- 0.05
out_tab_motif_p_value_cutoff <- 0.2 
out_tab_swing_p_value_cutoff <- 0.2


parser <- add_option(parser, c( "--kinase_specificity"), type = "character",
                     help = "Specificity of the kinase. One of Ser/Thr kinase= ST, Tyr kinase = Y, or Ser/Thr/Tyr kinase = STY",
                     metavar = "integer")

parser <- add_option(parser, c( "--phosphosite_db_dir"), type = "character",
                     help = "The directory in which PhosphositePlus data is stored",
                     metavar = "character")

parser <- add_option(parser, c( "--uniprot_kinase_file"), type = "character",
                     help = "The path to the UniProt tab-separated data file.",
                     metavar = "character")

parser <- add_option(parser, "--norm_phos_logfc_file", type="character",
                     help="Results table in which the phosphorylation log fold-change is normalized by protein log fold-change.",
                     metavar="string")

parser <- add_option(parser, "--uniprot_other_kinase_file", type="character",
                     help="File listing the UniProt accession of atypical and other kinases and their UniProt Keywords.",
                     metavar="string")

parser <- add_option(parser, c("-o", "--output_dir"), type = "character", dest = "output_dir",
                     help = "Directory path for all results files.",
                     metavar = "string")

parser <- add_option(parser, c("-n", "--no_backup"), action = "store_true", default = FALSE,
                     help = "Deactivate backup of previous run.")

parser <- add_option(parser, c("-l", "--log_file"), type = "character", default = "output.log",
                     help = "Name of the logging file.",
                     metavar = "string")

#parse command line arguments first.
args <- parse_args(parser)


 # knitr::purl(input="/home/ignatius/PostDoc/2021/proteomeriver/Source/Rmd/run_kinswingr.Rmd",
 #             output="/home/ignatius/PostDoc/2021/proteomeriver/Source/R/run_kinswingr.R" )

```


```{r}
args$debug <- TRUE
if (args$debug == TRUE) {

  args$tmp_dir <- "/home/ignatius/PostDoc/2021/Podocyte_2021/Results/Part_1/cache"
  args <- config.list.merge(eval.config(file = args$config, config = "phos_kinase_substrate"), args)

}
```

```{r}
## ---------------------------------------------------------------------------------------------------------------------------------------------------

createOutputDir(args$output_dir, args$no_backup)
createDirectoryIfNotExists(args$tmp_dir)

## Logger configuration
logReset()
addHandler(writeToConsole , formatter = cmriFormatter)
addHandler(writeToFile, file = file.path(args$output_dir, args$log_file), formatter = cmriFormatter)

level <- ifelse(args$debug, loglevels["DEBUG"], loglevels["INFO"])
setLevel(level = ifelse(args$silent, loglevels["ERROR"], level))

#parse and merge the configuration file options.
if (args$config != "") {
  args <- config.list.merge(eval.config(file = args$config, config = "phos_kinase_substrate"), args)
}

cmriWelcome("ProteomeRiver", c("Ignatius Pang", "Pablo Galaviz"))
loginfo("Reading configuration file %s", args$config)
loginfo("Argument: Value")
loginfo("----------------------------------------------------")
for (v in names(args))
{
  loginfo("%s : %s", v, args[v])
}
loginfo("----------------------------------------------------")

testRequiredArguments(args, c(
  "kinase_specificity",
  "phosphosite_db_dir",
  "uniprot_kinase_file",
  "norm_phos_logfc_file",
  "uniprot_other_kinase_file",
  "output_dir"
))


ks_file <-  file.path( args$phosphosite_db_dir, "Kinase_Substrate_Dataset")

testRequiredFiles(c( args$norm_phos_logfc_file,
                     args$uniprot_kinase_file
))
```




```{r}
## ---------------------------------------------------------------------------------------------------------------------------------------------------

loginfo("Read phosphosite log fold-change normalized by protein log fold-change file")
captured_output<-capture.output(
  de_phos <- vroom::vroom( args$norm_phos_logfc_file )
  ,type = "message"
)
logdebug(captured_output)

## ---------------------------------------------------------------------------------------------------------------------------------------------------

# uniprot_kinase_file<-file.path(args$tmp_dir,args$uniprot_kinase_file)
# if(!file.exists(uniprot_kinase_file))
# {
#   logwarn("Download UniProt kinases list file.")
#   status <- download.file(url="https://www.uniprot.org/docs/pkinfam.txt", destfile=uniprot_kinase_file)
#   loginfo(status)
# }
loginfo("Reading UniProt kinases list file.")
captured_output<-capture.output(
  uniprot_kinase_tbl <- vroom::vroom(  args$uniprot_kinase_file   )
    , type = "message"
)
logdebug(captured_output)


## ---------------------------------------------------------------------------------------------------------------------------------------------------


loginfo("Download information from UniProt.")
uniprot_other_kinase_file<-file.path(args$tmp_dir,args$uniprot_other_kinase_file)
if( ! file.exists( uniprot_other_kinase_file )) {

  up <- UniProt.ws(taxId=9606 )
  list_of_sp_columns <- c("EXISTENCE"
                          , "SCORE"
                          , "REVIEWED"
                          , "GENENAME"
                          , "PROTEIN-NAMES"
                          , "LENGTH"
                          , "ENSEMBL"
                          , "GO-ID"
                          , "KEYWORDS"
  )
  up_cls<-unlist(columns(up))
  list_intersect<-intersect(list_of_sp_columns,up_cls)
  if(length(setdiff( list_of_sp_columns,list_intersect)) > 0)
  {
    logerror("UniProt fields not found: %s",paste(list_of_sp_columns[,list_intersect],sep=", "))
  }

  uniprot_acc_tbl <- uniprot_kinase_tbl %>%
      dplyr::filter( Family == "Other" |
                       str_detect(Family, "Atypical")) %>%
      dplyr::distinct(uniprot_acc_human) %>%
      dplyr::rename(uniprot_acc = uniprot_acc_human)

  atypical_and_other <- batchQueryEvidence(uniprot_acc_tbl, uniprot_acc_column="uniprot_acc", uniprot_handle=up,
                                uniprot_columns = list_of_sp_columns)
  saveRDS( atypical_and_other, file.path(args$tmp_dir, uniprot_other_kinase_file))

}

loginfo("Reading UniProt data from %s.",uniprot_other_kinase_file)
atypical_and_other <- readRDS( uniprot_other_kinase_file )
```


```{r}
## ---------------------------------------------------------------------------------------------------------------------------------------------------
loginfo("Read PhosphoSitePlus (PSP) kinase-substrate table.")
captured_output<-capture.output(
  ks_tbl <- vroom::vroom( ks_file, skip=3 ) %>%
  mutate( SUB_MOD_RSD_CLN = str_replace_all(SUB_MOD_RSD, "([A-Z])(\\d+)", "\\1 \\2")) %>%
  separate( SUB_MOD_RSD_CLN, into=c("residue", "position"))
  ,type = "message"
)
logdebug(captured_output)



## ---------------------------------------------------------------------------------------------------------------------------------------------------
# As an example of control over multi-core processing
# load BiocParallel library
# finally set/register the number of cores to use

loginfo("Set number of cores for parallel computation.")
register(SnowParam(workers = args$num_cores))


## ---------------------------------------------------------------------------------------------------------------------------------------------------
loginfo("Filter the correct subset of kinases for the analysis.")

uniprot_kinases <- NA

if( args$kinase_specificity == "ST") {
   uniprot_kinases <- uniprot_kinase_tbl %>%
     dplyr::filter( str_detect( Family, "Ser/Thr" )  | str_detect( Family, "Atypical") | str_detect( Family, "Other" ) ) %>%
     left_join( atypical_and_other %>%
                  dplyr::select(UNIPROTKB, KEYWORDS),
                by=c("uniprot_acc_human" = "UNIPROTKB")) %>%
      dplyr::filter( str_detect( Family, "Ser/Thr" ) |
                       (   str_detect(KEYWORDS, "Serine/threonine-protein kinase") &
                          (!str_detect(KEYWORDS, "Tyrosine-protein kinase") ) ) )

} else if ( args$kinase_specificity == "Y") {
  uniprot_kinases <- uniprot_kinase_tbl %>%
    dplyr::filter( str_detect( Family, "Tyr" )  | str_detect( Family, "Atypical") | str_detect( Family, "Other" ) ) %>%
    left_join( atypical_and_other %>%
                  dplyr::select(UNIPROTKB, KEYWORDS),
               by=c("uniprot_acc_human" = "UNIPROTKB")) %>%
    dplyr::filter( str_detect( Family, "Tyr" ) |
                     ( (!str_detect(KEYWORDS, "Serine/threonine-protein kinase")) &
                         str_detect(KEYWORDS, "Tyrosine-protein kinase") ) )


} else if ( args$kinase_specificity == "STY") {
  uniprot_kinases <- uniprot_kinase_tbl %>%
    dplyr::filter(  str_detect( Family, "Atypical") | str_detect( Family, "Other" ) ) %>%
    left_join( atypical_and_other %>%
                  dplyr::select(UNIPROTKB, KEYWORDS),
               by=c("uniprot_acc_human" = "UNIPROTKB")) %>%
    dplyr::filter( str_detect(KEYWORDS, "Serine/threonine-protein kinase") &
                         str_detect(KEYWORDS, "Tyrosine-protein kinase")  )
}


## ---------------------------------------------------------------------------------------------------------------------------------------------------


phosphositeplus <- ks_tbl %>%
  dplyr::select( GENE, KINASE, `SITE_+/-7_AA`, SUB_MOD_RSD) %>%
  distinct() %>%
  dplyr::rename( kinase = "KINASE",
                 substrate = "SITE_+/-7_AA") %>%
  dplyr::mutate( substrate = toupper(substrate)) %>%
  dplyr::mutate( kinase = str_replace( kinase, " ", "_")) %>%
  arrange( GENE, kinase, substrate)  %>%
  mutate( GENE = toupper(GENE),
          kinase = toupper(kinase),
          residue = purrr::map_chr( SUB_MOD_RSD , ~str_sub(., 1,1 )   ) ) %>%
  dplyr::select( - SUB_MOD_RSD )

kinases_to_include <-  phosphositeplus %>%
  left_join( uniprot_kinases %>%
               dplyr::select(gene_name) %>%
               dplyr::mutate( is_uniprot_a= 1),
             by=c( "GENE" = "gene_name") ) %>%
  left_join( uniprot_kinases %>%
               dplyr::select(gene_name) %>%
               dplyr::mutate( is_uniprot_b= 1),
             by=c( "kinase" = "gene_name") ) %>%
  dplyr::filter( is_uniprot_a == 1 | is_uniprot_b == 1) %>%
  dplyr::select( -is_uniprot_a, - is_uniprot_b ) %>%
  dplyr::filter(  str_detect(  args$kinase_specificity, residue )  ) %>%
  group_by(kinase )  %>%
  summarise( counts =n()) %>%
  ungroup() %>%
  dplyr::filter( counts >= args$min_num_sites_per_kinase)

loginfo("Filter the correct subset of substrates for the analysis.")
phosphositeplus_filt <- phosphositeplus %>%
  mutate( kinase = toupper(kinase) ) %>%
  inner_join( kinases_to_include, by="kinase") %>%
  dplyr::filter(  str_detect(  args$kinase_specificity, residue )  ) %>%
  dplyr::select(-counts, - GENE, - residue) %>%
  distinct() %>%
  as.matrix()




## ---------------------------------------------------------------------------------------------------------------------------------------------------

loginfo("Build the position-specific scoring matrices (PSSM).")

pwms <- buildPWM(as.matrix(phosphositeplus_filt))

# pwms$kinase


## ---------------------------------------------------------------------------------------------------------------------------------------------------
loginfo("Clean up phoshopsite log fold-change table.")

annotated_data_pre_residue_filter <- de_phos %>%
  mutate( peptide = sequence) %>%
  mutate( positions_peptide = position   ) %>%
  mutate(uniprot_acc = str_split( uniprot_acc, ":") %>% purrr::map_chr(1) ) %>%
  mutate(gene_name = str_split( gene_name, ":") %>% purrr::map_chr(1) )  %>%
  mutate(position = str_split( position, ":") %>% purrr::map_chr(1) ) %>%
  mutate( peptide =  str_split(peptide, ":"  )%>% purrr::map_chr(1) ) %>%
  mutate( residue =  str_split(residue, ":"  )%>% purrr::map_chr(1) ) %>%
  mutate( position = str_split( position, "\\|") %>% purrr::map_chr(1) %>% str_replace_all( "\\(|\\)", "") ) %>%
  mutate ( is_multisite = case_when ( str_detect( position, ";") ~ TRUE,
                                   TRUE ~ FALSE)) %>%
  separate_rows(peptide, position, residue, sep=";") %>%  
  dplyr::mutate( peptide_copy = peptide) %>%
  dplyr::filter(  !str_detect( peptide, "X"))   %>%
  dplyr::select(sites_id, comparison, uniprot_acc, gene_name, peptide, peptide_copy, position, residue, 
                norm_phos_logFC, combined_q_mod, is_multisite) %>%
  unite( annotation,  uniprot_acc, gene_name, position, peptide_copy , sep="|"  ) 

annotated_data <- annotated_data_pre_residue_filter %>%
  dplyr::filter(  str_detect(  args$kinase_specificity, residue )  )  %>%
  dplyr::filter( str_sub( peptide, 8, 8) == residue) %>%
  dplyr::rename( fc = norm_phos_logFC,
                 pval = combined_q_mod)

## ---------------------------------------------------------------------------------------------------------------------------------------------------
loginfo("Perform analysis of data from each contrast")

grouped_annotated_data <- annotated_data %>%
  dplyr::select( -is_multisite, -sites_id, -residue ) %>%
  distinct() %>%
  group_by( comparison) %>%
  nest() %>%
  ungroup



## ---------------------------------------------------------------------------------------------------------------------------------------------------
# grouped_annotated_data$data[[1]]
```


```{r}
## ---------------------------------------------------------------------------------------------------------------------------------------------------
loginfo("Score each phosphorylation site for similarity to kinase-specific motif.")

rds_dir <- args$output_dir

if (!args$no_backup) {
  rds_dir <- paste(args$output_dir, "_prev", sep = "")
}

if(file.exists(file.path(rds_dir,  paste0("kinswingr_scores_list_", args$kinase_specificity, ".RDS")))) {

  scores_list <- readRDS( file.path(rds_dir, paste0("kinswingr_scores_list_", args$kinase_specificity, ".RDS")))

  captured_output<-capture.output(
    copy_attempt <- file.copy( from=  file.path(rds_dir, paste0("kinswingr_scores_list_", args$kinase_specificity, ".RDS")),
                               to =  file.path(args$output_dir, paste0("kinswingr_scores_list_", args$kinase_specificity, ".RDS")) ),
    type = "message"
  )
  logdebug(captured_output)


} else {
  # set seed for reproducible results
  set.seed(args$random_seed )
  tic()

  print( purrr::map_int(grouped_annotated_data$data, ~nrow(as.data.frame(.) )))
  scores_list <-  grouped_annotated_data %>%
    dplyr::mutate( pwms_scores = purrr::map( data, ~scoreSequences(input_data = as.data.frame(.),
                                                                   pwm_in = pwms,
                                                                   n = args$motif_score_iteration) ))
  toc()


    saveRDS( scores_list, file.path(args$output_dir,  paste0("kinswingr_scores_list_", args$kinase_specificity, ".RDS")))

}
```


## ---------------------------------------------------------------------------------------------------------------------------------------------------

```{r}
purrr::walk2( scores_list$data,
              scores_list$comparison,
              ~vroom::vroom_write( .x,
                                   file.path( args$output_dir,
                                              paste0( "input_data_", args$kinase_specificity, "_",  .y , ".tsv" )) ) )




## ---------------------------------------------------------------------------------------------------------------------------------------------------


purrr::walk2( scores_list$pwms_scores,
              scores_list$comparison,
              ~vroom::vroom_write( .x$peptide_scores,
                                   file.path( args$output_dir,
                                              paste0( "peptide_scores_", args$kinase_specificity, "_",  .y , ".tsv" )) ) )




## ---------------------------------------------------------------------------------------------------------------------------------------------------


purrr::walk2( scores_list$pwms_scores,
              scores_list$comparison,
              ~vroom::vroom_write( .x$peptide_p,
                                   file.path( args$output_dir,
                                              paste0( "peptide_p_", args$kinase_specificity, "_",  .y , ".tsv" )) ) )




## ---------------------------------------------------------------------------------------------------------------------------------------------------


purrr::walk2( scores_list$pwms_scores,
              scores_list$comparison,
              ~vroom::vroom_write( .x$background,
                                   file.path( args$output_dir,
                                              paste0( "peptide_background_", args$kinase_specificity, "_",  .y , ".tsv" )) ) )

```



```{r}
loginfo("Perform randomization analysis with Swing.")

# set seed for reproducible results
set.seed(args$random_seed)


if( file.exists(file.path(rds_dir,  paste0("kinswingr_swing_out_list_", args$kinase_specificity, ".RDS")))) {

  swing_out_list <- readRDS( file.path(rds_dir,
                                       paste0("kinswingr_swing_out_list_", args$kinase_specificity, ".RDS")))

  captured_output<-capture.output(
    copy_attempt <- file.copy( from=  file.path(rds_dir, paste0("kinswingr_swing_out_list_", args$kinase_specificity, ".RDS")),
                               to =  file.path(args$output_dir, paste0("kinswingr_swing_out_list_", args$kinase_specificity, ".RDS")) ),
    type = "message"
  )
  logdebug(captured_output)

} else {

  tic()

  swing_out_list  <-  scores_list  %>%
    dplyr::mutate( swing_result = purrr::map2( data, pwms_scores, ~swing(input_data = as.data.frame( .x),
                                                                        pwm_in = pwms,
                                                                        pwm_scores = .y,
                                                                        permutations = args$swing_iteration,
                                                                        return_network= TRUE)))

  toc()

  # This will produce two tables, one is a network for use with e.g. Cytoscape and the other is the scores. To access the scores:
  head(swing_out_list$swing_result[[1]]$scores)


  saveRDS( swing_out_list, file.path( args$output_dir,
                                      paste0("kinswingr_swing_out_list_", args$kinase_specificity, ".RDS")))
}




## ---------------------------------------------------------------------------------------------------------------------------------------------------


purrr::walk2( swing_out_list$swing_result,
              swing_out_list$comparison,
              ~vroom::vroom_write( .x$scores,
                                   file.path( args$output_dir,
                                              paste0( "KinSwingR_", args$kinase_specificity, "_",  .y , ".tsv" )) ) )




## ---------------------------------------------------------------------------------------------------------------------------------------------------


purrr::walk2( swing_out_list$swing_result,
              swing_out_list$comparison,
              ~vroom::vroom_write( .x$network,
                                   file.path( args$output_dir,
                                              paste0( "network_", args$kinase_specificity, "_",  .y , ".tsv" )) ) )

```

```{r}
## ---------------------------------------------------------------------------------------------------------------------------------------------------
loginfo("Compile KingSwinger results together.")


compileKinswingerResults <- function( list_position ) {

  my_comparison <- swing_out_list$comparison[[list_position]]

  fc_pval_tab <- scores_list$data[[list_position]]

  up_or_down_kinases <- swing_out_list$swing_result[[list_position]]$scores %>%
    dplyr::filter( p_greater < 0.2 | p_less < 0.2 )

  motif_score_table <- scores_list$pwms_scores[[list_position]]$peptide_scores %>%
    pivot_longer( cols= !(contains("annotation") | contains("peptide")),
                  names_to="kinase",
                  values_to = "motif.score")

  is_phosphoylated_tbl <- annotated_data %>%
    dplyr::filter( pval < 0.05) %>%
    dplyr::mutate( uniprot_acc = str_split(annotation, "\\|") %>% purrr::map_chr(1) ) %>%
    distinct( uniprot_acc) %>%
    mutate(is_kinase_phosphorylated = 1)

  selected_scores_list_help <- scores_list$pwms_scores[[list_position]]$peptide_p %>%
    pivot_longer( cols= !(contains("annotation") | contains("peptide")),
                  names_to="kinase",
                  values_to = "motif.p.value") %>%
    left_join( motif_score_table, by=c("annotation" = "annotation",
                                       "peptide" = "peptide",
                                       "kinase" = "kinase")) %>%
    inner_join( up_or_down_kinases, by=c( "kinase" = "kinase")) %>%
    dplyr::filter( motif.p.value < 0.2 ) %>%
    inner_join( fc_pval_tab, by=c("annotation" = "annotation",
                                  "peptide" = "peptide")) %>%
    left_join ( annotated_data %>%
                  dplyr::select( annotation, sites_id),
                by=c("annotation" = "annotation")) %>%
    left_join( de_phos %>%
                 dplyr::select( sites_id, `PROTEIN-NAMES`, reactome_term, 
                                KINASE,
                                ON_FUNCTION,
                                ON_PROCESS,
                                ON_PROT_INTERACT,
                                ON_OTHER_INTERACT,
                                REG_SITES_NOTES ),
               by=c("sites_id" = "sites_id")) %>%
    dplyr::mutate( kinase_copy = KINASE ) %>%
    dplyr::rename( known_upstream_kinase = "kinase_copy") %>%
    dplyr::rename( one_known_kinase = "KINASE") %>%
    separate_rows( one_known_kinase , sep= "//")  %>%
    dplyr::mutate( one_known_kinase = toupper(one_known_kinase) )  %>%
    left_join( phosphositeplus %>%
                 distinct( GENE, kinase), by=c("kinase" = "kinase") ) %>%
    dplyr::mutate( substrate_gene_name  = str_split(annotation, "\\|") %>% purrr::map_chr(2)) %>%
    dplyr::mutate( prediction_match_known_kinase = case_when( kinase == one_known_kinase ~ TRUE,
                                                              TRUE ~ FALSE) )  %>%
    dplyr::select(-one_known_kinase) %>%
    left_join( uniprot_kinases %>%
                 dplyr::select(-KEYWORDS),
               by= c("GENE" = "gene_name")) %>%
    dplyr::rename( kinase_gene_name = "GENE") %>%
    distinct()
  
  ## Use mouse or human uniprot accession if it makes sense to do so.
  selected_scores_list <- selected_scores_list_help
  if( length( intersect(   is_phosphoylated_tbl$uniprot_acc, uniprot_kinases$uniprot_acc_human  ) ) >0 )  {
    selected_scores_list <- selected_scores_list_help  %>%
      left_join( is_phosphoylated_tbl, by=c("uniprot_acc_human" = "uniprot_acc"))  %>%
      distinct()
  } else if ( length( intersect(   is_phosphoylated_tbl$uniprot_acc, uniprot_kinases$uniprot_acc_mouse  ) > 0 ))  {
    selected_scores_list <- selected_scores_list_help  %>%
      left_join( is_phosphoylated_tbl, by=c("uniprot_acc_mouse" = "uniprot_acc"))  %>%
      distinct()
  }

  
  
  return( selected_scores_list)
}




selected_scores_list <- purrr::map( seq_along(swing_out_list$comparison),
             ~compileKinswingerResults(.))

names( selected_scores_list) <- swing_out_list$comparison



purrr::walk2( selected_scores_list, 
              swing_out_list$comparison,
              ~vroom::vroom_write( .x,
                                   file.path( args$output_dir,
                                              paste0( "selected_kinase_substrate_", 
                                                      args$kinase_specificity, "_",  
                                                      .y , 
                                                      ".tsv" )) ) )

```


```{r}

```


## ---------------------------------------------------------------------------------------------------------------------------------------------------
```{r}
loginfo("Plot kinase-level volcano plot.")

kinase_volcano_plot <- function(
  x_label_name, # <- c("Patient - Healthy", "Patient - Untreated", "Unreated - Healthy")
  kinase_type = "ST",
  list_position,
  brewer_pal_set = "Set1",
  p_value_threshold = 0.05,
  n_cutoff = 10 ) {

  swing_scores_tbl <- swing_out_list$swing_result[[list_position]]$scores

  kinase_type_string <- case_when( kinase_type == "ST" ~ "predicted Ser/Thr PK activity",
                            kinase_type == "Y" ~ "predicted Tyr PK activity",
                            kinase_type == "STY" ~ "predicted Ser/Thr/Tyr PK activity")

  my_colours <- brewer.pal(length(swing_out_list$comparison), brewer_pal_set)

  swing_scores_tbl %>%
    mutate( p_value = case_when( swing >=0 ~ p_greater,
                                 swing < 0 ~ p_less) ) %>%
    mutate( colour = case_when( p_value < p_value_threshold ~ "Significant",
                                TRUE ~ "Not significant")) %>%
    mutate(kinase_label = case_when( p_value < p_value_threshold  &
                                       n > n_cutoff ~ kinase,
                                     TRUE ~ "")) %>%
    ggplot( aes( x=swing, y = -log10(p_value + 10^-100), size=n , col=colour, label=kinase_label), alpha=0.5)  +
    theme_bw () +
    geom_point( ) +
    scale_color_manual( values = c("grey", my_colours[list_position]), name="Is significant" ) +
    scale_size_continuous(name = "Num. substrates") +
    ylab( expression("Significance, -log"[10]*"(P)") ) +
    xlab( paste0( x_label_name[list_position], ", ", kinase_type_string)) +
    geom_hline(aes(yintercept = -log10(p_value_threshold)), linetype=3)  +
    geom_text_repel( show.legend = FALSE, size = 3 )

}

partial_kinase_volcano_plot <-  partial( kinase_volcano_plot,
           x_label_name = c("Patient - Healthy", "Patient - Untreated", "Unreated - Healthy"),
           kinase_type = "ST",
           p_value_threshold = args$p_value_cutoff, 
           n_cutoff = args$min_num_sites_per_kinase)

kinase_volcano_plot <- purrr::map( seq_along(swing_out_list$comparison),
             ~partial_kinase_volcano_plot(list_position =.))

purrr::walk2( kinase_volcano_plot, swing_out_list$comparison,
              ~ggsave( filename=file.path(args$output_dir, paste0(  "kinase_volcano_plot_", args$kinase_specificity, "_",  .y, ".pdf" )),
                       plot=.x) )
```


```{r}
#kinase_volcano_plot
```


```{r}

plotAvgLogfcKnownSites <- function( list_position, 
                                    y_label_name,
                                    brewer_pal_set= "Set1",
                                    kinase_type,
                                    p_value_cutoff = 0.05) {
  
  kinase_type_string <- case_when( kinase_type == "ST" ~ "predicted Ser/Thr PK activity",
                            kinase_type == "Y" ~ "predicted Tyr PK activity",
                            kinase_type == "STY" ~ "predicted Ser/Thr/Tyr PK activity")
  
  my_colours <- brewer.pal(length(swing_out_list$comparison), brewer_pal_set)
  
  my_comparison <- swing_out_list$comparison[[list_position]]
  
  fc_pval_tab <- scores_list$data[[list_position]]
  
  up_or_down_kinases <- swing_out_list$swing_result[[list_position]]$scores 
  
  motif_score_table <- scores_list$pwms_scores[[list_position]]$peptide_scores %>%
    pivot_longer( cols= !(contains("annotation") | contains("peptide")),
                  names_to="kinase",
                  values_to = "motif.score")
  
  selected_scores_list <- scores_list$pwms_scores[[list_position]]$peptide_p %>%
    pivot_longer( cols= !(contains("annotation") | contains("peptide")),
                  names_to="kinase",
                  values_to = "motif.p.value") %>%
    left_join( motif_score_table, by=c("annotation" = "annotation",
                                       "peptide" = "peptide",
                                       "kinase" = "kinase")) %>%
    inner_join( up_or_down_kinases, by=c( "kinase" = "kinase")) %>%
    inner_join( fc_pval_tab, by=c("annotation" = "annotation",
                                  "peptide" = "peptide")) %>%
    left_join ( annotated_data %>%
                  dplyr::select( annotation, sites_id),
                by=c("annotation" = "annotation")) %>%
    left_join( de_phos %>%
                 dplyr::select(sites_id, KINASE), by = c("sites_id" = "sites_id")) %>%
    distinct() %>%
    separate_rows( KINASE , sep= "//")  %>%
    dplyr::mutate( KINASE = toupper(KINASE) ) %>%
    dplyr::filter( kinase == KINASE) 
  
  avg_logfc_vs_swing_score <- selected_scores_list %>%
    group_by( kinase, swing, p_greater, p_less ) %>%
    summarise( avg_fc = mean(fc), 
               median_fc = median(fc),
               total_fc = sum(fc),
               counts = n()) %>%
    ungroup 
  
  avg_logfc_vs_swing_score_plot <- avg_logfc_vs_swing_score %>%
    dplyr::filter( counts >= 5 | p_greater < p_value_cutoff | p_less < p_value_cutoff )  %>%
    mutate( colour = case_when( ( p_greater < p_value_cutoff | 
                                  p_less < p_value_cutoff ) ~ "Significant",
                                TRUE ~ "Not significant" )) %>%    
    dplyr::mutate( my_kinase = case_when( counts >= 10 | 
                                            ( p_greater < p_value_cutoff | 
                                              p_less < p_value_cutoff ) ~ kinase,
                                          TRUE ~ "") ) %>%
    ggplot(aes( y = avg_fc, x = swing, label = my_kinase, size = counts,  color=colour ), alpha=0.5) +
    geom_point() +
    geom_text_repel(show.legend = FALSE, size = 3) +
    scale_color_manual( values = c("grey", my_colours[list_position]), name="Is significant" ) +
    ylab( paste0(y_label_name[list_position], ", ", expression(  "av. Substrate log[2](intensity) difference") ) ) +
    xlab( paste0( y_label_name[list_position], ", ", kinase_type_string)) 
  
  return( avg_logfc_vs_swing_score_plot)
}


partial_plotAvgLogfcKnownSites <-  partial( plotAvgLogfcKnownSites,
           y_label_name = c("Patient - Healthy", "Patient - Untreated", "Unreated - Healthy"),
            brewer_pal_set= "Set1", 
           kinase_type = args$kinase_specificity,
           p_value_cutoff = args$p_value_cutoff)


kinase_avg_logfc <- purrr::map( seq_along(swing_out_list$comparison),  
                                ~partial_plotAvgLogfcKnownSites(.) ) 


purrr::walk2( kinase_avg_logfc, swing_out_list$comparison,
              ~ggsave( filename=file.path(args$output_dir, paste0(  "kinase_avg_logfc_of_known_sites_", args$kinase_specificity, "_",  .y, ".pdf" )),
                       plot=.x) )

```


```{r}
  colnames(de_phos )
  
```


```{r}
## ---------------------------------------------------------------------------------------------------------------------------------------------------
#  RColorBrewer::display.brewer.all()

## ---------------------------------------------------------------------------------------------------------------------------------------------------
te<-toc(quiet = TRUE)
loginfo("%f sec elapsed",te$toc-te$tic)
writeLines(capture.output(sessionInfo()), file.path(args$output_dir,"sessionInfo.txt"))


```

