---
title: "QC Analysis for xyz"
format:
    html:
        embed-resources: true
        self-contained: true
        code-fold: true
        code-summary: "Show the code"
        #toc: true
        #toc-depth: 3
        #number-sections: true
        html-math-method: katex
        css: styles.css
editor: source
execute:
    echo: false
    warning: false
    message: false
author: Your fancy self
output: html_document
---

## Dependency, Directory, Hardware Management & Orgnamism Annotation
#### Set and create directories for outputs and project management, set number of cores for threaded tasks

```{r Environment Management}
devtools::install_github("APAF-BIOINFORMATICS/ProteomeScholaR", ref = "dev-jr")
library(ProteomeScholaR)
# Dependency Management
## Load required packages
## Checks your R environment for the required packages to run ProteomeScholaR, and installs them if they are not.
## Please note that you will require a Github account in order to install the ProteomeScholaR package.
## Github account creation is free, and is a very useful tool for learning to code and also to engage with the wider coding community.
loadDependencies()

# Hardware Management
## Change this if you have large experiments and need more compute, or are running a toaster and don't have many cores to work with :)
#core_utilisation <- new_cluster(4)
#cluster_library(core_utilisation, c("tidyverse", "glue", "rlang", "lazyeval"))


# Directory Management
## Set up the project directory structure
## This section sets up the project directory structure for ProteomeScholaR
## Directory management can be challenging, particularly when managing objects across multiple chunks within a single R Markdown document.
setupDirectories()
showDirectories()

# Annotation Management
## Please download the organism fasta file from UniProt. If UniProt is not available, the program will extract the 
## relevant identifiers from the fasta provided and attempt to match them to user supplied UniProt / UniParc conversions
## Please set the name of your fasta file here in the root directory if you already have it
data_tbl <- vroom::vroom( file.path(data_dir, "proteomics", "KV_DIANN_report.tsv"))
fasta_file_path <- file.path( data_dir, "UniProt", "04153260899A.fasta")
#uniprot_search_results <- NULL
uniprot_search_results <- vroom::vroom(file.path(data_dir, "UniProt", "idmapping_2024_10_07.tsv"))
#uniparc_search_results <- NULL
uniparc_search_results <- vroom::vroom(file.path(data_dir, "UniProt", "kv_uniparc.csv"))

## Please supply your organism's taxon ID here
taxon_id <- 244366
```


## Input Parameters for Quality Control
### Parameters in this section are experiment-specific. Their default parameters are intended as a guide only - every source of variance is different just as every set of proteins ### going through a mass spectrometer is different! One size does not fit all and you *will* most likely need to fine tune these to get the most out of your data.
```{r}
# Read the entire config file 
config_list <- readConfigFile( file = file.path(source_dir, "config.ini") )

```


```{r}
RunApplet("designMatrix")
```


```{r}
fasta_meta_file <- "parsed_fasta_data.rds"
aa_seq_tbl_final <- processFastaFile(fasta_file_path, uniprot_search_results, uniparc_search_results, fasta_meta_file)
data_cln <- updateProteinIDs(data_cln, aa_seq_tbl_final)
```

## Create the PeptideQuantitativeData object
### This section initializes a PeptideQuantitativeData object with peptide-level quantitative data and experimental design information.
### It specifies the column names for various data attributes and sets up the design matrix for the experiment.
```{r}
peptide_data <- new( "PeptideQuantitativeData"
  
  # Protein vs Sample quantitative data
  , peptide_data = data_cln
  , protein_id_column = "Protein.Ids"
  , peptide_sequence_column = "Stripped.Sequence"
  , q_value_column = "Q.Value"
  , global_q_value_column = "Global.Q.Value"
  , proteotypic_peptide_sequence_column = "Proteotypic"
  , raw_quantity_column = "Precursor.Quantity"
  , norm_quantity_column = "Precursor.Normalised"
  , is_logged_data = FALSE
  
  # Design Matrix Information
  , design_matrix = design_matrix
  , sample_id="Run"
  , group_id="group"
  , technical_replicate_id="replicates"
  
  , args = config_list)  
```

# Raw Data QC
```{r}

updateProteinFiltering(
  data = data_cln,
  step_name = "Raw Data",
  publication_graphs_dir = publication_graphs_dir,
  return_grid = TRUE,
  overwrite = TRUE
)
```

## Filter peptides based on q-value and proteotypic peptide match
### NB should be left to default unless you have specific experimental needs
```{r}

search_srl_quant_cln <-  srlQvalueProteotypicPeptideClean( theObject = peptide_data  )

updateProteinFiltering(
  data = search_srl_quant_cln@peptide_data,
  step_name = "qval Filtered",
  publication_graphs_dir = publication_graphs_dir,
  return_grid = TRUE,
  overwrite = TRUE
)
```


## Roll-up of precursor ions to peptide level intensity value quantitation
```{r}
peptide_normalized_tbl <- rollUpPrecursorToPeptide( search_srl_quant_cln  )

peptide_normalized_tbl@peptide_data |> 
  dplyr::filter( peptidoform_count > 1)

peptide_normalized_tbl@peptide_data |> distinct(Protein.Ids) |> nrow()

updateProteinFiltering(
    data = peptide_normalized_tbl@peptide_data,
  step_name = "peptidoform count",
  publication_graphs_dir = publication_graphs_dir,
  return_grid = TRUE,
overwrite = TRUE
)

```


## Remove peptide based on the intensity threshold and the proportion of samples below the threshold
### NB should be left to default unless you have specific experimental needs
```{r}
peptide_normalized_pif_cln <- peptideIntensityFiltering( theObject = peptide_normalized_tbl )

peptide_normalized_pif_cln@peptide_data |> distinct(Protein.Ids) |> nrow()

updateProteinFiltering(
  data = peptide_normalized_pif_cln@peptide_data,
  step_name = "peptideIntensityFiltering",
  publication_graphs_dir = publication_graphs_dir,
  return_grid = TRUE,
overwrite = TRUE
)
```


## Keep the proteins only if they have two or more peptides mapping
### NB should be left to default unless you have specific experimental needs 
#### CHANGE IF YOU ARE INTERESTED IN QUANTIFYING SINGLE PEPTIDE PROTEIN MATCHES
```{r}
removed_peptides_with_less_than_two_peptides <- filterMinNumPeptidesPerProtein( theObject =peptide_normalized_pif_cln ) 


removed_peptides_with_less_than_two_peptides@peptide_data |> distinct(Protein.Ids) |> nrow()

updateProteinFiltering(
  data = removed_peptides_with_less_than_two_peptides@peptide_data,
  step_name = "<2pep",
  publication_graphs_dir = publication_graphs_dir,
  return_grid = TRUE,
overwrite = TRUE
)
```


## Remove samples with insufficient peptide counts
#### This section filters out samples that have fewer than a specified minimum number of peptides - ie poor sample performance
```{r}
peptide_keep_samples_with_min_num_peptides <- filterMinNumPeptidesPerSample(theObject = removed_peptides_with_less_than_two_peptides )

(removed_peptides_with_less_than_two_peptides@args)$filterMinNumPeptidesPerSample

peptide_keep_samples_with_min_num_peptides@peptide_data |> distinct(Protein.Ids) |> nrow()

updateProteinFiltering(
  data = peptide_keep_samples_with_min_num_peptides@peptide_data,
  step_name = "minpeppersample",
  publication_graphs_dir = publication_graphs_dir,
  return_grid = TRUE,
overwrite = TRUE
)


removed_peptides_with_less_than_two_peptides@peptide_data |> distinct(Run,Protein.Ids, Stripped.Sequence
, peptidoform_count) |> group_by( Run) |> summarise( n = sum(peptidoform_count) ) |> arrange( desc(n))

## Show the distinct number of Samples to manually check how many samples are removed from the previous step
peptide_keep_samples_with_min_num_peptides@peptide_data |> distinct(Run) |> nrow()
```


## Remove peptides with only one replicate in the data set
### This section filters out peptides that appear in only one replicate across all groups.
### Ensures that the analysis is based on peptides with consistent detection across multiple replicates.
```{r}

removed_peptides_with_only_one_replicate <- removePeptidesWithOnlyOneReplicate(peptide_keep_samples_with_min_num_peptides  )


removed_peptides_with_only_one_replicate@peptide_data |> distinct(Protein.Ids) |> nrow()

updateProteinFiltering(
  data = removed_peptides_with_only_one_replicate@peptide_data,
  step_name = "peptidesonerep",
  publication_graphs_dir = publication_graphs_dir,
  return_grid = TRUE,
overwrite = TRUE
)

peptide_keep_samples_with_min_num_peptides@peptide_data |> distinct(Protein.Ids, Stripped.Sequence) |> nrow()
removed_peptides_with_only_one_replicate@peptide_data |> distinct(Protein.Ids, Stripped.Sequence) |> nrow()

```


## Missing values "imputation" using technical replicates
### This section imputes missing values using the average of technical replicate samples.
### The imputation is performed if there's a high proportion of technical replicate samples with non-missing values.
### By default this section is commented out. Comment back in if technical replicates are included in the experimental data.

```{r}

peptide_values_imputed <- peptideMissingValueImputation( theObject = removed_peptides_with_only_one_replicate )

# colnames( removed_peptides_with_only_one_replicate@peptide_data  )

(removed_peptides_with_only_one_replicate@args)$peptideMissingValueImputation

peptide_values_imputed@peptide_data |> distinct(Protein.Ids) |> nrow()

#head(peptide_values_imputed)

peptide_values_imputed_file <- file.path( results_dir
        , "peptide_qc"
        , "peptide_values_imputed.tsv")

vroom::vroom_write( peptide_values_imputed@peptide_data |>
                      mutate( Q.Value = 0.0009
                              , PG.Q.Value = 0.009 ) |>
                      mutate( Peptide.Imputed = ifelse( is.na(Peptide.Imputed), 0, Peptide.Imputed))
                    , peptide_values_imputed_file )
```


## Read in the fasta organism specific fasta file to extract details on the protein sequences
### There should be no need to change this chunk, ever

## This section aggregates the long-format peptide-level intensity values into protein-level quantification:
### It uses the IQ tool (https://github.com/tvpham/iq), which implements the same algorithm as DIA-NN's maxLFQ but runs faster (written in C++) 
### Unless your experiment has specific requirements or you wish to use IQ differently, it is recommended to leave default settings here
```{r}
process_long_format(input_filename = peptide_values_imputed_file
        , output_filename = file.path(results_dir, "protein_qc", "iq_output_file.txt")
        , sample_id = "Run"
        , primary_id = "Protein.Ids"
        , secondary_id = "Stripped.Sequence"
        , intensity_col = "Peptide.Imputed"
        , filter_double_less = c("Q.Value" = "0.01", "PG.Q.Value" = "0.01")
        ## very important for this workflow that you do NOT perform normalization here
        , normalization = "none")

## Read in the IQ output file (which outputs a file, not an object)
dir.create( file.path(results_dir, "proteomics", "protein_qc"), recursive=TRUE  )
protein_log2_quant <- vroom::vroom( file.path(results_dir, "protein_qc", "iq_output_file.txt"))

## Check the number of proteins here to ensure no data lost
nrow( protein_log2_quant ) 
```


## Create Protein Quantitative Data Object
### Unless you have changed the column identifiers or the object names leave defaults
```{r}
protein_obj <- ProteinQuantitativeData( 
  # Protein Data Matrix Information
    protein_quant_table = protein_log2_quant
  , protein_id_column= "Protein.Ids"
  , protein_id_table = protein_log2_quant |> distinct(Protein.Ids)
  # Design Matrix Information
  , design_matrix = peptide_values_imputed@design_matrix
  , sample_id="Run"
  , group_id="group"
  , technical_replicate_id="replicates"
  , args = peptide_values_imputed@args
)
```


## Arrange the protein ID's list to opt for the best accession in the list to be placed first
### It requires the information from the fasta file to choose the best accession for the protein ID.
### Please skip this section if you have uploaded your own annotation from Uniprot/uniparc
```{r}
#please run this if you have a uniprot-sourced fasta
#protein_log2_quant_cln <- chooseBestProteinAccession( theObject = protein_obj
#                                                      , delim = ";"
#                                                      , seqinr_obj = aa_seq_tbl
#                                                      , seqinr_accession_column = "uniprot_acc"
#                                                      , replace_zero_with_na = TRUE)

#please run this if you have supplied your own uniprot/uniparc searches
protein_log2_quant_cln <- protein_obj
```



## Remove protein based on the intensity threshold and the proportion of samples below the threshold
### The threshold is determined by the 1% quantile of the protein intensity values
### This helps to ensure that only reliably quantified proteins are retained for further analysis
```{r}
protein_normalized_pif_cln <-  removeRowsWithMissingValuesPercent(protein_log2_quant_cln )

updateProteinFiltering(
  data = protein_normalized_pif_cln@protein_quant_table,
  step_name = "protein_missingvals_percent",
  publication_graphs_dir = publication_graphs_dir,
  return_grid = TRUE,
overwrite = TRUE
)

protein_normalized_pif_cln@protein_quant_table |> distinct(Protein.Ids) |> nrow()

```


## Summarize data from duplicate proteins
### Calculate mean across matching duplicate proteins and populate the new identifier with a single value
### Leave as default unless you wish to perform another form of duplication handling
### Or if you have an exotic experiment that looks at very identical proteins, modifications etc you may wish to skip this.
```{r}
# Identify duplicates
duplicates <- protein_normalized_pif_cln@protein_quant_table |>
  group_by(Protein.Ids) |>
  filter(n() > 1) |>
  pull(Protein.Ids) |>
  unique()

duplicates

# Clean duplicates
protein_normalized_pif_cln@protein_quant_table <- protein_normalized_pif_cln@protein_quant_table |>
  group_by(Protein.Ids) |>
  summarise(across(matches("\\d+"), ~ mean(.x, na.rm = TRUE))) |>
  ungroup()

protein_normalized_pif_cln@protein_quant_table |> distinct(Protein.Ids) |> nrow()
```

## Remove proteins with only one replicate in the data set
### No need to change this unless you wish to include single replicate proteins
```{r}
core_utilisation <- new_cluster(4) #note to change this in the below function to cores from config_list
remove_proteins_with_only_one_rep <- removeProteinsWithOnlyOneReplicate (  protein_normalized_pif_cln, core_utilisation, grouping_variable = "group" )
vroom::vroom_write(remove_proteins_with_only_one_rep@protein_quant_table, file.path( results_dir, "protein_qc", "remove_proteins_with_only_one_rep.tsv"))

updateProteinFiltering(
  data = remove_proteins_with_only_one_rep@protein_quant_table,
  step_name = "proteins_onerep_filter",
  publication_graphs_dir = publication_graphs_dir,
  return_grid = TRUE,
overwrite = TRUE
)

remove_proteins_with_only_one_rep@protein_quant_table |> distinct(Protein.Ids) |> nrow()
```


## Pre-normalisation data QC
### RLE plot 
### PCA plot
### Pearson correlation
### Spearman correlation
```{r}
QC_composite_figure <- InitialiseGrid()

QC_composite_figure@rle_plots$rle_plot_before_cyclic_loess <- plotRle(remove_proteins_with_only_one_rep, "group"
            ,  yaxis_limit=c(-4, 4))

QC_composite_figure@pca_plots$pca_plot_before_cyclic_loess_group <-  plotPca( remove_proteins_with_only_one_rep
            , grouping_variable = "group"
            , label_column = ""
            , title = ""
            , font_size = 8 )
pca_mixomics_before_cyclic_loess <- getPcaMatrix(remove_proteins_with_only_one_rep)

QC_composite_figure@pearson_plots$pearson_correlation_pair_before_cyclic_loess <- plotPearson(remove_proteins_with_only_one_rep
            , tech_rep_remove_regex = "pool")

summarizeQCPlot(QC_composite_figure)

save_plot(QC_composite_figure@rle_plots$rle_plot_before_cyclic_loess, results_dir, "rle_plot_before_cyclic_loess")
save_plot(QC_composite_figure@pca_plots$pca_plot_before_cyclic_loess_group, results_dir, "pca_plot_before_cyclic_loess")
save_plot(QC_composite_figure@pearson_plots$pearson_correlation_pair_before_cyclic_loess, results_dir, "pearson_correlation_pair_before_cyclic_loess")

frozen_protein_matrix_tech_rep <- proteinTechRepCorrelation (remove_proteins_with_only_one_rep
                                                                           , tech_rep_num_column =  "group"
                                                                           , tech_rep_remove_regex = "pool")
## change if you wish to filter on pearson or spearman
frozen_protein_matrix_tech_rep |> dplyr::filter( pearson > 0.8) |> nrow()
frozen_protein_matrix_tech_rep |> dplyr::filter( spearman > 0.8) |> nrow()
```


## Cyclic loess normalisation and QC
### RLE plot 
### PCA plot
### Pearson correlation
```{r}
normalised_frozen_protein_matrix_obj <- normaliseBetweenSamples( remove_proteins_with_only_one_rep
         , normalisation_method = "cyclicloess" )

QC_composite_figure@rle_plots$rle_plot_before_ruvIIIc_group <- plotRle(normalised_frozen_protein_matrix_obj, "group"
                                                                       , yaxis_limit=c(-4, 4))

QC_composite_figure@pca_plots$pca_plot_before_ruvIIIc_group <- plotPca(normalised_frozen_protein_matrix_obj
            , grouping_variable = "group"
            , label_column = ""
            , title = ""
            , font_size = 8 )
pca_mixomics_before_ruvIIIc <- getPcaMatrix(normalised_frozen_protein_matrix_obj)

QC_composite_figure@pearson_plots$pca_plot_before_ruvIIIc_group <- plotPearson(normalised_frozen_protein_matrix_obj
            , tech_rep_remove_regex = "pool")
  
summarizeQCPlot(QC_composite_figure)

save_plot(QC_composite_figure@rle_plots$rle_plot_before_ruvIIIc_group, results_dir, "rle_plot_before_ruvIIIc_by_group")
save_plot(QC_composite_figure@pca_plots$pca_plot_before_ruvIIIc_group, results_dir, "pca_plot_before_ruvIIIc_by_group")
save_plot(QC_composite_figure@pearson_plots$pearson_correlation_pair_before_ruvIIIc, results_dir, "pearson_correlation_pair_before_ruvIIIc")

```
## Use ANOVA to define subset (negative controls) of proteins that do not change in the dataset
### The k value with the highest separation between All and 'Control' group is selected as the best k.
#### It is a heuristic and the best_k value can be adjusted manually, if required.
```{r}

#REDRAW RUV GRAPH with line for best_k????? as a vertical line
control_genes_index <-  getNegCtrlProtAnova(normalised_frozen_protein_matrix_obj
                                            , ruv_grouping_variable = "group"
                                            # Percentage of dataset to use as a negative control.
                                            # Best left to default.
                                            , percentage_as_neg_ctrl = 10
                                            , ruv_qval_cutoff = 0.05
                                            , ruv_fdr_method = "BH" )
cancorplot_r1 <- ruvCancor(normalised_frozen_protein_matrix_obj
                              , ctrl = control_genes_index
                              , num_components_to_impute= 5
                              , ruv_grouping_variable = "group")
cancorplot_r1 + xlim(1, ncol(normalised_frozen_protein_matrix_obj@protein_quant_table)-1)

save_plot(cancorplot_r1, results_dir, "canonical_correlation_plot")
# Display the plot
cancorplot_r1
# Find the best k
best_k <- findBestK(cancorplot_r1)
# Display the best k
best_k
```
Use ANOVA to define subset (negative controls) of proteins that do not change in the dataset


## Run RUVIII-C and QC
### RLE plot 
### PCA plot
### Pearson correlation
```{r}
ruv_normalised_results_temp_obj <- ruvIII_C_Varying( normalised_frozen_protein_matrix_obj
                     , ruv_grouping_variable = "group"
                     , ruv_number_k = 2
                     , ctrl = control_genes_index) 

## Sometimes RUV will blank out some of the values, so we need to remove proteins if too many values are blanked out 

ruv_normalised_results_cln_obj <-  removeRowsWithMissingValuesPercent(theObject = ruv_normalised_results_temp_obj
                                                                      , ruv_grouping_variable =  'group'
                                                                      , groupwise_percentage_cutoff = 50
                                                                      , max_groups_percentage_cutoff = 60
                                                                      , proteins_intensity_cutoff_percentile = 1 )

QC_composite_figure@rle_plots$rle_plot_after_ruvIIIc_group <- plotRle( ruv_normalised_results_cln_obj
                                            , group="group"
                                            , yaxis_limit =c(-4, 4) )

QC_composite_figure@pca_plots$pca_plot_after_ruvIIIc_group <- plotPca( ruv_normalised_results_cln_obj
            , grouping_variable = "group"
            , label_column = ""
            , title = ""
            , font_size = 8 )
pca_mixomics_after_ruvIIIc <- getPcaMatrix(ruv_normalised_results_cln_obj)
QC_composite_figure@pearson_plots$pearson_correlation_pair_after_ruvIIIc_group <- plotPearson(ruv_normalised_results_cln_obj
            , tech_rep_remove_regex = "pool")


save_plot(QC_composite_figure@rle_plots$rle_plot_after_ruvIIIc_group, results_dir, "pca_plot_after_ruvIIIc")
save_plot(QC_composite_figure@pca_plots$pca_plot_after_ruvIIIc_group, results_dir, "rle_plot_after_ruvIIIc_by_group")
save_plot(QC_composite_figure@pearson_plots$pearson_correlation_pair_after_ruvIIIc_group, results_dir, "pearson_correlation_pair_after_ruvIIIc_group")     
ruv_normalised_results_cln_obj
summarizeQCPlot(QC_composite_figure)
```


## RUV normalized data for de analysis
### After RUVIII-C is performed, some proteins will have their values blanked out as NA by RUV (most likely assoicated with their batch effects).
### Therefore, we need to remove proteins with high proportion of missing values in each group before procceding to DE analysis.
```{r}
##CHUNK TO FUNCTION

ruv_correlation_vec <-  pearsonCorForSamplePairs(ruv_normalised_results_cln_obj
                            , tech_rep_remove_regex = "pool" ) 
ruv_normalised_filtered_results_obj <- filterSamplesByProteinCorrelationThreshold(ruv_normalised_results_cln_obj
                                           , pearson_correlation_per_pair  = ruv_correlation_vec
                                           , min_pearson_correlation_threshold = 0.5 )

vroom::vroom_write( ruv_normalised_filtered_results_obj@protein_quant_table
              , file.path(results_dir, "protein_qc",  "ruv_normalized_results_cln_with_replicates.tsv"))


saveRDS( ruv_normalised_filtered_results_obj
        , file.path( results_dir, "protein_qc", "ruv_normalized_results_cln_with_replicates.RDS"))


ruv_normalised_for_de_analysis_obj <-  ruv_normalised_filtered_results_obj

##CHUNK TO FUNCTION
##THIS NEEDS TO GO INTO THE OUTPUT FUNCTION?????

ruv_normalised_for_de_analysis <- ruv_normalised_for_de_analysis_obj@protein_quant_table |>
  pivot_longer( cols= !matches("Protein.Ids")
                , names_to = "replicates"
                , values_to = "Log2.Protein.Imputed") |>
  dplyr::select( Protein.Ids, replicates, Log2.Protein.Imputed) |>
  mutate( Protein.Imputed = 2^Log2.Protein.Imputed) |>
  mutate( Protein.Imputed = ifelse( is.na(Protein.Imputed), NA, Protein.Imputed)) |>
  pivot_wider( id_cols = Protein.Ids
               , names_from = replicates
               , values_from = Protein.Imputed) |>
  dplyr::rename(uniprot_acc = "Protein.Ids")

vroom::vroom_write(ruv_normalised_for_de_analysis
        , file.path(results_dir, "protein_qc", "ruv_normalised_results.tsv"))

vroom::vroom_write(ruv_normalised_for_de_analysis |>
                     dplyr::mutate(across(!matches("uniprot_acc"), log2))
        , file.path(results_dir, "protein_qc", "ruv_normalised_results_log.tsv"))

vroom::vroom_write( design_matrix |>
                      distinct( replicates, group) |>
                      dplyr::rename( Run = replicates)
                    , file.path(results_dir, "protein_qc", "design_matrix_avrg.tsv") )

ruv_normalised_for_de_analysis_mat <- ruv_normalised_for_de_analysis |>
  column_to_rownames("uniprot_acc") |>
  as.matrix()

  vroom::vroom_write( ruv_normalised_filtered_results_obj@protein_quant_table
              , file.path(results_dir, "protein_qc",  "ruv_normalised_results_cln_with_replicates.tsv"))


saveRDS( ruv_normalised_filtered_results_obj
        , file.path( results_dir, "protein_qc", "ruv_normalised_results_cln_with_replicates.RDS"))

####CHUNK END
ruv_normalised_for_de_analysis_obj <-  ruv_normalised_filtered_results_obj

ruv_normalised_for_de_analysis_obj
```


# maxquant_row_id	num_gene_names	gene_names	uniprot_acc	is_unique	protein_ids
## ONLY RUN THIS if you have imported a UniProt tfasta
* I need to import the UniProt tab file
* Then I need to recreate the cleaned_accession_to_protein_group.tab file
```{r}
#  ruv_normalized_results <- saveRDS( file.path( results_dir, "protein_qc", "ruv_normalized_results.RDS"))
cleaned_accession_table <- data.frame( protein_ids = setdiff(colnames(ruv_normalized_results_cln_obj@protein_data), "Protein")) |>
        left_join( uniprot_tbl
        , by = join_by( protein_ids == Entry)) |>
        mutate( maxquant_row_id = row_number() -1 ) |>
        mutate( num_gene_names =   1) |>
        mutate( gene_names = purrr::map_chr( `Gene Names`, \(x){ str_split(x, " ")[[1]][1]} ) ) |>
        mutate( is_unique = "Unique") |>
        mutate ( uniprot_acc = protein_ids) |>
        dplyr::select(maxquant_row_id,	num_gene_names,	gene_names,	uniprot_acc,	is_unique,	protein_ids )
cleaned_accession_table

vroom::vroom_write(cleaned_accession_table
        , file.path(results_dir, "clean_proteins", "cleaned_accession_to_protein_group.tab"))
```

# Add up to date annotation data
```{r}
up <- UniProt.ws(taxId=taxon_id)
tmp_dir <- file.path(getwd(), "tmp")
if (!dir.exists(tmp_dir)) {
  dir.create(tmp_dir)
}
annotations <- NULL
if (!file.exists(file.path(tmp_dir, "uniprot_annotations.RDS"))) {
  
  annotations <- batchQueryEvidenceGeneId(
    ruv_normalised_for_de_analysis_obj@protein_quant_table,
    gene_id_column = "Protein.Ids",
    uniprot_handle = up,
    uniprot_columns = c(
      "protein_existence",
      "annotation_score",
      "reviewed",
      "gene_names",
      "protein_name",
      "length",
      "xref_ensembl",
      "go_id",
      "keyword"
    )
  ) 
   # |> dplyr::rename(uniprot_acc = From)  # Debug to check if the From column needed renaming
  
  uniprot_dat_cln <- annotations |>
    uniprotGoIdToTerm(
      uniprot_id_column = Entry,
      go_id_column = Gene.Ontology.IDs,
      sep = "; "
    ) |>
    dplyr::rename(
      Protein_existence = "Protein.existence",
      Protein_names = "Protein.names"
    )
  
  saveRDS(uniprot_dat_cln, file.path(tmp_dir, "uniprot_annotations.RDS"))
  
} else {
  uniprot_dat_cln <- readRDS(file.path(tmp_dir, "uniprot_annotations.RDS"))
}
```


## Create the composite QC figure
### Change the titles here if you so wish

```{r}

pca_ruv_rle_correlation_merged <- createGridQC(QC_composite_figure
     , pca_titles = c("a)", "b)", "c)")
     , rle_titles = c("d)", "e)", "f)")
     , pearson_titles = c("g)", "h)", "i)")
     , save_path = file.path(results_dir, "protein_qc")
     , file_name = "composite_QC_figure"
   )
pca_ruv_rle_correlation_merged
```



## Run Differential Abundance Analysis
```{r}

#RunApplet2("contrastMatrix")
# change linear model string for more complicated experiments
# output is formula_string
# OUTPUT FILE CREATION - CONFIG FILE PARAMS + PARAMS IN WORKBOOK + CONTRAST STRING + LIMMA MODEL

#contrasts_tbl <- read.table(file.path(source_dir, "contrast_strings.tab"), header = TRUE, sep = "\t")
contrasts_tbl <- as.data.frame(contrasts_tbl)[-1, , drop = FALSE] #need to fix this in the shiny output
config_list$deAnalysisParameters$args_group_pattern <-"(\\d+)" #need to fix this in the config list readin
de_analysis_results_list <- deAnalysisWrapperFunction ( ruv_normalised_for_de_analysis_obj
                                       , contrasts_tbl
                                       , formula_string = config_list$deAnalysisParameters$formula_string
                                       , de_q_val_thresh = config_list$deAnalysisParameters$de_q_val_thresh
                                       , treat_lfc_cutoff = config_list$deAnalysisParameters$treat_lfc_cutoff
                                       , eBayes_trend = config_list$deAnalysisParameters$eBayes_trend
                                       , eBayes_robust = config_list$deAnalysisParameters$eBayes_robust
                                       , args_group_pattern = config_list$deAnalysisParameters$args_group_pattern
                                       , args_row_id = config_list$deAnalysisParameters$args_row_id ) 

```


## Output the results of the DE analysis
```{r}

outputDeAnalysisResults (de_analysis_results_list
                         , ruv_normalised_for_de_analysis_obj
                         , uniprot_dat_cln
                         , de_output_dir
                         , publication_graphs_dir
                         , file_prefix="de_proteins"
                         , plots_format = config_list$deAnalysisParameters$plots_format
                         , args_row_id = "uniprot_acc"
                         , de_q_val_thresh=0.05
                         , gene_names_column = "Gene.Names")
```


```{r}
# Create workflow args with config and git info
workflow_args <- createWorkflowArgsFromConfig(
    workflow_name = "DIA Proteomics Workflow",
    description = "Full protein analysis workflow with config parameters" #ADD A MEANINGFUL LABEL HERE FOR TRACKING WHAT YOU DID
)

# Show the workflow arguments
workflow_args

### add a second output for params used as config_timestamp.ini
```

# Copy all the relevant files to a publication directory
```{r}
## Need to write down where these all are systematically
## copy to results_summary_dir/proteomics
## report doc will pull everything necessary from this folder
```