---
output: reprex::reprex_document
knit: reprex::reprex_render
---


## Packages Management
```{r}

#Test if BioManager is installed
if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

# load pacman package manager
if(!require(pacman)){
  install.packages("pacman")
  library(pacman)
}

p_load(tidyverse)
p_load(seqinr)
p_load(lazyeval)
p_load(rlang)
p_load(glue)
p_load(GGally)
p_load(here)
p_load(tibble)
p_load(mixOmics)
p_load(limma)
p_load(magrittr)

p_load(future.apply)
p_load(tictoc)
p_load(beepr)

p_load(furrr)
p_load(readxl)
p_load(writexl)
p_load(RColorBrewer)
p_load(multidplyr)

# require(devtools)
# install_github('CMRI-ProCan/RUV-III-C')
p_load(RSpectra)
p_load(progress)
p_load(Rcpp)
p_load(RcppEigen)

p_load(qvalue)
p_load(Glimma)
p_load(UniProt.ws)
p_load(GO.db)

if(!require(RUVIIIC)) {
  devtools::install_github("cran/RUVIIIC")
  library(RUVIIIC)
}

p_load(RcppEigen)

# https://github.com/cran/RUVIIIC
p_load(RUVIIIC)
p_load(ruv)

# Sys.setenv( R_LIBS_USER="/Users/ignatiuspang/R/x86_64-pc-linux-gnu-library/4.4")
# .libPaths( c( Sys.getenv("R_LIBS_USER"), Sys.getenv("R_LIBS") ) )
devtools::install_github("APAF-Bioinformatics/ProteomeScholaR")
p_load(ProteomeScholaR)
p_load(iq)
p_load(ggrepel)
p_load(patchwork)
```

## Directories Management
```{r}
base_dir <- here::here()
results_dir <- file.path(base_dir, "results", "proteomics")
data_dir <- file.path(base_dir, "data")
data_proteomics_dir <- 
source_dir <- file.path(base_dir, "scripts", "proteomics")
de_output_dir <- file.path( results_dir, "de_proteins")
tmp_dir <- file.path(results_dir, "cache")
publication_graphs_dir <- file.path( results_dir, "publication_graphs")

createDirectoryIfNotExists(file.path( results_dir))
createDirectoryIfNotExists(tmp_dir)
createOutputDir(de_output_dir, FALSE)
createOutputDir(file.path( results_dir, "clean_proteins"), FALSE)
createOutputDir(file.path(results_dir, "protein_qc"), FALSE)
createOutputDir(file.path(results_dir, "peptide_qc"), FALSE)
createOutputDir(publication_graphs_dir, FALSE)

```


## Input Parameters
```{r}
choose_only_proteotypic_peptide <- 1
global_qvalue_threshold <- 0.01
qvalue_threshold <- 0.01

core_utilisation <- new_cluster(4)
cluster_library(core_utilisation, c("tidyverse", "glue", "rlang", "lazyeval"))


peptides_intensity_cutoff_percentile <- 0.01
peptides_proportion_of_samples_below_cutoff <- 0.5
peptides_per_protein_cutoff <- 1
peptidoforms_per_protein_cutoff <- 2
peptides_per_sample_cutoff <- 5000
num_of_cores <- 8
min_pearson_correlation_threshold <- 0.70

proportion_missing_for_imputation <- 0.50

## DE analysis parameters
experimental_contrast_file <- file.path( source_dir, "contrast_strings.tab")
plots_format <- c("pdf", "png")

formula_string <- "~ 0 + group"
treat_lfc_cutoff <- log2(1.5)
eBayes_trend <- TRUE
eBayes_robust <- TRUE
de_q_val_thresh <- 0.05
args_row_id <- "uniprot_acc"
args_group_pattern="(\\d+)"


```


## Read peptides data from DIA-NN
```{r, message=FALSE, warning=FALSE}
data_tbl <- vroom::vroom( file.path(data_dir
                                    , "proteomics"
                                    , "241002_34513_Quant_Results.tsv.zip"))
```


```{r}
colnames( data_tbl)
data_tbl |> distinct(Run)
```

## Custom cleaning of DIA-NN peptide data
```{r}
data_cln <- data_tbl |>
        mutate(Run = str_remove(Run
            , "20240926_P34513_OE_")) |>
        mutate(Run = str_remove(Run, "_DIA")) 

## View cleaned peptides table from DIA-NN

#data_cln
data_cln |> distinct(Run)

```


## Tidy-up the design matrix
```{r}
design_matrix <- vroom::vroom( file.path( source_dir
                                          , "design_matrix.tab")) 
design_matrix <- as.data.frame(design_matrix)
rownames( design_matrix) <- design_matrix[,"Run"]

```

## Read in the fasta file to extract details on the protein sequences
```{r}

fasta_meta_file<-file.path(tmp_dir,"aa_seq_tbl.RDS")

fasta_file <- file.path( data_dir, "UniProt", "uniprotkb_proteome_UP000000589_2024_09_30.fasta")

if (file.exists(fasta_meta_file))
{
  aa_seq_tbl <- readRDS(fasta_meta_file)
}else{
  aa_seq_tbl <- parseFastaFile(fasta_file)
  saveRDS(aa_seq_tbl, fasta_meta_file)

}


```

```{r}
peptide_data <- new( "PeptideQuantitativeData"
  
  # Protein vs Sample quantitative data
  , peptide_data = data_cln
  , protein_id_column = "Protein.Ids"
  , peptide_sequence_column = "Stripped.Sequence"
  , q_value_column = "Q.Value"
  , global_q_value_column = "Global.Q.Value"
  , proteotypic_peptide_sequence_column = "Proteotypic"
  , raw_quantity_column = "Precursor.Quantity"
  , norm_quantity_column = "Precursor.Normalised"
  , is_logged_data = FALSE
  
  # Design Matrix Information
  , design_matrix = design_matrix
  , sample_id="Run"
  , group_id="group"
  , technical_replicate_id="replicates"
  
  , args = list( srlQvalueProteotypicPeptideClean = list( qvalue_threshold = qvalue_threshold
                                                          , global_qvalue_threshold = global_qvalue_threshold
                                                          , choose_only_proteotypic_peptide = choose_only_proteotypic_peptide
                                                          , input_matrix_column_ids = c("Run"
                                                                                 , "Precursor.Id"
                                                                                 , "Protein.Ids"
                                                                                 , "Stripped.Sequence"
                                                                                 , "Modified.Sequence"
                                                                                 , "Precursor.Charge"
                                                                                 , "Precursor.Quantity"
                                                                                 , "Precursor.Normalised") )
                 , rollUpPrecursorToPeptide = list( core_utilisation=core_utilisation )
                 , peptideIntensityFiltering = list( peptides_intensity_cutoff_percentile = peptides_intensity_cutoff_percentile
                                                     , peptides_proportion_of_samples_below_cutoff = peptides_proportion_of_samples_below_cutoff
                                                     , core_utilisation = core_utilisation )
                 , filterMinNumPeptidesPerProtein = list( peptides_per_protein_cutoff = peptides_per_protein_cutoff
                                                         , peptidoforms_per_protein_cutoff = peptidoforms_per_protein_cutoff
                                                         , core_utilisation = core_utilisation )
                 , filterMinNumPeptidesPerSample = list( peptides_per_sample_cutoff = peptides_per_sample_cutoff
                                                        , core_utilisation = core_utilisation
                                                        , inclusion_list = c(  ))
                 , removePeptidesWithOnlyOneReplicate = list( grouping_variable  = "group"
                                                             , core_utilisation = core_utilisation )
                 , peptideMissingValueImputation = list( imputed_value_column = "Peptide.Imputed"
                                                         , proportion_missing_values = proportion_missing_for_imputation
                                                         , core_utilisation = core_utilisation )
                 , chooseBestProteinAccession = list( delim = ";"
                                                     , seqinr_obj = aa_seq_tbl
                                                     , seqinr_accession_column = "uniprot_acc")
                 , removeRowsWithMissingValuesPercent = list( ruv_grouping_variable =  'group'
                                                             , groupwise_percentage_cutoff = 50
                                                             , max_groups_percentage_cutoff = 50
                                                             , min_protein_intensity_percentile = 1)
                 , proteinTechRepCorrelation = list( tech_rep_num_column =  "Run"
                                                    , tech_rep_remove_regex = "pool")
                 , removeProteinsWithOnlyOneReplicate = list( core_utilisation = core_utilisation
                                                              , grouping_variable = "group" )
                 , normalizeBetweenSamples = list( method = "cyclicloess" )
                 , pearsonCorForSamplePairs = list( tech_rep_remove_regex = "pool")
                 , plotRle = list( ylim=c(-0.75, 0.75)) )) 
 
```


## Filter by peptide q-value and number of proteotypic peptides
```{r}
 search_srl_quant_cln <-  srlQvalueProteotypicPeptideClean( theObject = peptide_data  )

search_srl_quant_cln

```
##----------------------------------------------------------------------------------------------------------------------------------------------------------------------


## Roll-up of precursor to peptide level intensity value quantitation
```{r}
peptide_normalized_tbl <- rollUpPrecursorToPeptide( search_srl_quant_cln  )


peptide_normalized_tbl

peptide_normalized_tbl@peptide_data |> 
  dplyr::filter( peptidoform_count > 1)
```

## Remove peptide based on the intensity threshold and the proportion of samples below the threshold
```{r}
peptide_normalized_pif_cln <- peptideIntensityFiltering( theObject = peptide_normalized_tbl )

peptide_normalized_pif_cln@peptide_data |> distinct(Protein.Ids) |> nrow()
```

## Keep the proteins only if they have two or more peptides.  
```{r}

removed_peptides_with_less_than_two_peptides <- filterMinNumPeptidesPerProtein( theObject =peptide_normalized_pif_cln ) 


removed_peptides_with_less_than_two_peptides@peptide_data |> distinct(Protein.Ids) |> nrow()

```


## Remove samples below the minimum number of peptides after passing intensity threshold
```{r}
peptide_keep_samples_with_min_num_peptides <- filterMinNumPeptidesPerSample(theObject = removed_peptides_with_less_than_two_peptides )

(removed_peptides_with_less_than_two_peptides@args)$filterMinNumPeptidesPerSample

peptide_keep_samples_with_min_num_peptides@peptide_data |> distinct(Protein.Ids) |> nrow()


removed_peptides_with_less_than_two_peptides@peptide_data |> distinct(Run,Protein.Ids, Stripped.Sequence
, peptidoform_count) |> group_by( Run) |> summarise( n = sum(peptidoform_count) ) |> arrange( desc(n))

```

## Show the distinct number of Samples to manually check how many samples are removed from the previous step
```{r}
peptide_keep_samples_with_min_num_peptides@peptide_data |>
  distinct(Run) |>
  nrow()
```


## Remove peptides with only one replicate in the data set
```{r}


removed_peptides_with_only_one_replicate <- removePeptidesWithOnlyOneReplicate(peptide_keep_samples_with_min_num_peptides  )


removed_peptides_with_only_one_replicate@peptide_data |> distinct(Protein.Ids) |> nrow()

peptide_keep_samples_with_min_num_peptides@peptide_data |> distinct(Protein.Ids, Stripped.Sequence) |> nrow()
removed_peptides_with_only_one_replicate@peptide_data |> distinct(Protein.Ids, Stripped.Sequence) |> nrow()

```



```{r}
plotPeptidesProteinsCountsPerSample(removed_peptides_with_only_one_replicate )

```


```{r}
## Missing values imputation using technical replicates
## * Among technical replicate samples, impute the missing value using the average of the technical replicate samples, if there are high proportion of technical replicate samples with values (and not NA). 

peptide_values_imputed <- peptideMissingValueImputation( theObject = removed_peptides_with_only_one_replicate )

# colnames( removed_peptides_with_only_one_replicate@peptide_data  )

(removed_peptides_with_only_one_replicate@args)$peptideMissingValueImputation

peptide_values_imputed@peptide_data |> distinct(Protein.Ids) |> nrow()

#head(peptide_values_imputed)
```



## Save a copy of the data matrix with the imputated values from techical replicates 
```{r}
peptide_values_imputed_file <- file.path( results_dir
        , "peptide_qc"
        , "peptide_values_imputed.tsv")

vroom::vroom_write( peptide_values_imputed@peptide_data |>
                      mutate( Q.Value = 0.0009
                              , PG.Q.Value = 0.009 ) |>
                      mutate( Peptide.Imputed = ifelse( is.na(Peptide.Imputed), 0, Peptide.Imputed))
                    , peptide_values_imputed_file )
```




## Perform peptide-to-protein roll-up 
* Intensity values from peptides are aggregated into a protein level using a specific algorithm, implemented by the IQ tool (which is the same algorithm as DIA-NN's maxlfq code). IQ just runs faster as it was written in C++.
* Use IQ instead of DIA-NN's maxlfq (https://github.com/tvpham/iq)

```{r}


process_long_format(input_filename = peptide_values_imputed_file
        , output_filename = file.path(results_dir, "protein_qc", "iq_output_file.txt")
        , sample_id = "Run"
        , primary_id = "Protein.Ids"
        , secondary_id = "Stripped.Sequence"
        , intensity_col = "Peptide.Imputed"
        , filter_double_less = c("Q.Value" = "0.01", "PG.Q.Value" = "0.01")
        , normalization = "none")


```


## Read in the IQ output file 
That is because the results are saved in an Excel file (and no objects are returned by the IQ function)
```{r}

protein_log2_quant <- vroom::vroom( file.path(results_dir, "protein_qc", "iq_output_file.txt"))


nrow( protein_log2_quant ) 

```

## Create Protein Quantitative Data Object
```{r}
protein_obj <- ProteinQuantitativeData( 
  # Protein Data Matrix Information
  protein_data=protein_log2_quant
  , protein_id_column= "Protein.Ids"
  
  , protein_id_table = data.frame( )
  # Design Matrix Information
  , design_matrix = peptide_values_imputed@design_matrix
  , sample_id="Run"
  , group_id="group"
  , technical_replicate_id="replicates"
  , args = peptide_values_imputed@args
)
```

## Arrange the protein ID's list to opt for the best accession in the list to be placed first
It requires the information from the fasta file to choose the best accession for the protein ID.
```{r}

protein_log2_quant_cln <- chooseBestProteinAccession( theObject = protein_obj
                                                      , delim = ";"
                                                      , seqinr_obj = aa_seq_tbl
                                                      , seqinr_accession_column = "uniprot_acc")


# summed_data <- protein_log2_quant_cln@protein_data |>
#   mutate( Protein.Ids = purrr::map_chr( Protein.Ids, \(x){ str_split(x, ":")[[1]][1] } ) )  |> 
#   group_by(Protein.Ids) |> 
#   summarise( across( !matches("Protein.Ids"), \(x) { sum(x, na.rm=TRUE) } )) |>
#   ungroup()

# protein_log2_quant_summed <- protein_log2_quant_cln
# protein_log2_quant_summed@protein_data <- summed_data
```


## Remove protein based on the intensity threshold and the proportion of samples below the threshold
The threshold is determined by the 1% quantile of the protein intensity values.
```{r}
protein_normalized_pif_cln <-  removeRowsWithMissingValuesPercent(protein_log2_quant_cln
                                      , ruv_grouping_variable =  'group'
                           , groupwise_percentage_cutoff = 60
                           , max_groups_percentage_cutoff = 60
                           , min_protein_intensity_percentile = 1)



protein_normalized_pif_cln@protein_data |> distinct(Protein.Ids) |> nrow()

```

## Remove proteins with only one replicate in the data set
```{r}
remove_proteins_with_only_one_rep <- removeProteinsWithOnlyOneReplicate (  protein_normalized_pif_cln, core_utilisation, grouping_variable = "group" )

vroom::vroom_write(remove_proteins_with_only_one_rep@protein_data, file.path( results_dir, "protein_qc", "remove_proteins_with_only_one_rep.tsv"))

remove_proteins_with_only_one_rep@protein_data |> distinct(Protein.Ids) |> nrow()
```


## RLE plot before Cyclic loess normalization
```{r}

rle_plot_before_cyclic_loess <-   plotRle(remove_proteins_with_only_one_rep, "replicates", ylim=c(-0.75, 0.75))

pdf(file.path( results_dir, "protein_qc", "rle_plot_before_cyclic_loess.pdf" ))
rle_plot_before_cyclic_loess
dev.off()

png(file.path( results_dir, "protein_qc", "rle_plot_before_cyclic_loess.png"))
rle_plot_before_cyclic_loess
dev.off()

rle_plot_before_cyclic_loess

```



## Plot PCA plot before Cyclic loess normalization
```{r}
pca_plot_before_cyclic_loess_group <-  plotPca( remove_proteins_with_only_one_rep
            , group_column = "group"
            , label_column = ""
            , title = ""
            , geom_text_size = 8 )

pca_mixomics_before_cyclic_loess <- getPcaMatrix( remove_proteins_with_only_one_rep)


pdf(file.path( results_dir, "protein_qc", "pca_plot_before_cyclic_loess.pdf" ))
pca_plot_before_cyclic_loess_group
dev.off()

png(file.path( results_dir, "protein_qc", "pca_plot_before_cyclic_loess.png"))
pca_plot_before_cyclic_loess_group
dev.off()

pca_plot_before_cyclic_loess_group

```


# Calculate Pearson correlation between Tech replicates

## Calculate Pearson correlation between each pair of technical replicate samples, before cyclic loess and before RUVIII-C
```{r}
correlation_vec <- pearsonCorForSamplePairs( remove_proteins_with_only_one_rep
                                                , tech_rep_remove_regex = "pool")

pearson_correlation_pair_before_cyclic_loess <- correlation_vec |>
 ggplot( aes( pearson_correlation)) +
  geom_histogram( breaks = seq(min(round(correlation_vec$pearson_correlation-0.5,2), na.rm=TRUE), 1, 0.001)) +
   scale_y_continuous(breaks = seq(0, 4, 1), limits = c(0, 4)) +
  xlab("Pearson Correlation") +
  ylab("Counts") +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank())

pearson_correlation_pair_before_cyclic_loess

ggsave( plot=pearson_correlation_pair_before_cyclic_loess
        ,filename= file.path( results_dir, "protein_qc", "pearson_correlation_pair_before_cyclic_loess.pdf") )

ggsave( plot=pearson_correlation_pair_before_cyclic_loess
        ,filename= file.path( results_dir, "protein_qc", "pearson_correlation_pair_before_cyclic_loess.png") )

```




### Spearman correlation between Tech reps, unnormalized but log transformed data matrix 

```{r}

frozen_protein_matrix_tech_rep <- proteinTechRepCorrelation (remove_proteins_with_only_one_rep
                                                                           , tech_rep_num_column =  "group"
                                                                           , tech_rep_remove_regex = "pool")

frozen_protein_matrix_tech_rep |> dplyr::filter( pearson > 0.8) |> nrow()
frozen_protein_matrix_tech_rep |> dplyr::filter( spearman > 0.8) |> nrow()

```



## Perform cyclic loess normalization
```{r}


normalized_frozen_protein_matrix_obj <- normalizeBetweenSamples( remove_proteins_with_only_one_rep
         , method = "cyclicloess" )



# normalized_frozen_protein_matrix

```

## Normalized Frozen Protein Matrix, technical replicates
```{r}

normalized_frozen_protein_matrix_tech_rep <- proteinTechRepCorrelation (normalized_frozen_protein_matrix_obj
                                                                           , tech_rep_num_column =  "group"
                                                                           , tech_rep_remove_regex = "pool")


normalized_frozen_protein_matrix_tech_rep |> nrow()
normalized_frozen_protein_matrix_tech_rep |> dplyr::filter( pearson > 0.8) |> nrow()
normalized_frozen_protein_matrix_tech_rep |> dplyr::filter( spearman > 0.8) |> nrow()

```

## Calculate Pearson correlation between each pair of technical replicate samples, after cyclic loess and before RUVIII-C
```{r}

correlation_vec <- pearsonCorForSamplePairs( normalized_frozen_protein_matrix_obj
                                             , tech_rep_remove_regex = "pool")

pearson_correlation_pair_before_ruvIIIc <- correlation_vec |>
  ggplot( aes( pearson_correlation)) +
  geom_histogram( breaks = seq(min(round(correlation_vec$pearson_correlation-0.5,2), na.rm=TRUE), 1, 0.001)) +
  scale_y_continuous(breaks = seq(0, 4, 1), limits = c(0, 4)) +
  xlab("Pearson Correlation") +
  ylab("Counts") +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank())

pearson_correlation_pair_before_ruvIIIc

ggsave( plot=pearson_correlation_pair_before_ruvIIIc
        ,filename= file.path( results_dir, "protein_qc", "pearson_correlation_pair_before_ruvIIIc.pdf") )

ggsave( plot=pearson_correlation_pair_before_ruvIIIc
        ,filename= file.path( results_dir, "protein_qc", "pearson_correlation_pair_before_ruvIIIc.png") )

```



## Plot RLE Plot before RUVIII-c
```{r}
rle_plot_before_ruvIIIc_group <-   plotRle(normalized_frozen_protein_matrix_obj, "group", ylim=c(-3, 3))

pdf(file.path( results_dir, "protein_qc", "rle_plot_before_ruvIIIc_by_group.pdf" ))
rle_plot_before_ruvIIIc_group
dev.off()

png(file.path( results_dir, "protein_qc", "rle_plot_before_ruvIIIc_by_group.png"))
rle_plot_before_ruvIIIc_group
dev.off()

rle_plot_before_ruvIIIc_group
```

## Plot PCA plot before RUVIII-C
```{r}

pca_plot_before_ruvIIIc_group <- plotPca( normalized_frozen_protein_matrix_obj
            , group_column = "group"
            , label_column = ""
            , title = ""
            , geom_text_size = 8 )

pca_mixomics_before_ruvIIIc <- getPcaMatrix( normalized_frozen_protein_matrix_obj)

pdf(file.path( results_dir, "protein_qc", "pca_plot_before_ruvIIIc_by_group.pdf" ))
pca_plot_before_ruvIIIc_group
dev.off()

png(file.path( results_dir, "protein_qc", "pca_plot_before_ruvIIIc_by_group.png"))
pca_plot_before_ruvIIIc_group
dev.off()

pca_plot_before_ruvIIIc_group

```

## Compare each patient using ANOVA to identify negative controls
```{r}
control_genes_index <-  getNegCtrlProtAnova(normalized_frozen_protein_matrix_obj
                                               , ruv_grouping_variable = "replicates"
                                            , percentage_as_neg_ctrl = 5
                                            #, num_neg_ctrl = 100
                                            , q_val_thresh = 0.05
                                            , fdr_method = "BH" )
# 
# which(control_genes_index)


```

## Draw canonical correlation plot
The k value with the highest separation between All and 'Control' group is selected as the best k. It is a heuristic and the best_k value can be adjusted manually, if required.

```{r}

cancorplot_r1 <- ruvCancor(normalized_frozen_protein_matrix_obj
                              ,ctl = control_genes_index
                              ,ncomp= 5
                              ,ruv_grouping_variable = "replicates")

cancorplot_r1 +
        xlim(1, ncol(normalized_frozen_protein_matrix_obj@protein_data)-1)

pdf(file.path( results_dir, "protein_qc", "canonical_correlation_plot.pdf" ))
cancorplot_r1 +
        xlim(1, ncol(normalized_frozen_protein_matrix_obj@protein_data)-1)
dev.off()

png(file.path( results_dir, "protein_qc", "canonical_correlation_plot.png"))
cancorplot_r1 +
        xlim(1, ncol(normalized_frozen_protein_matrix_obj@protein_data)-1)
dev.off()

best_k <- findBestK( cancorplot_r1)

best_k

```

## Run RUVIII-C
```{r}
ruv_normalized_results_temp_obj <- ruvIII_C_Varying( normalized_frozen_protein_matrix_obj
                     , ruv_grouping_variable = "group"
                     , k = best_k
                     , ctl = control_genes_index) 
```

```{r}
## Sometimes RUV will blank out some of the values, so we need to remove proteins if too many values are blanked out 

ruv_normalized_results_cln_obj <-  removeRowsWithMissingValuesPercent(theObject = ruv_normalized_results_temp_obj
                                                                      , ruv_grouping_variable =  'group'
                                                                      , groupwise_percentage_cutoff = 50
                                                                      , max_groups_percentage_cutoff = 60
                                                                      , min_protein_intensity_percentile = 1 )

```


## Normalized Frozen Protein Matrix, correlation at protein level comparing first and second technical replicates
```{r}
ruv_frozen_protein_matrix_tech_rep <- proteinTechRepCorrelation ( ruv_normalized_results_cln_obj
                                                                     , tech_rep_num_column = "group"
                                                                     , tech_rep_remove_regex = "pool")

ruv_frozen_protein_matrix_tech_rep |> nrow()
ruv_frozen_protein_matrix_tech_rep |> dplyr::filter( pearson > 0.8) |> nrow()
ruv_frozen_protein_matrix_tech_rep |> dplyr::filter( spearman > 0.8) |> nrow()

ruv_frozen_protein_matrix_tech_rep_plot <- ruv_frozen_protein_matrix_tech_rep |>
  mutate( plot = purrr::pmap( list( Protein.Ids, spearman, data)
                              , \(x, y, z){ ggplot(z, aes(`1`, `2`)) + 
                                  geom_point() + 
                                  geom_smooth(method="lm") + 
                                  ggtitle( paste(x, round(y, 2))) }) ) 

saveRDS( ruv_frozen_protein_matrix_tech_rep_plot
        , file.path( results_dir, "protein_qc", "ruv_protein_level_correlation.RDS"))

ruv_frozen_protein_matrix_tech_rep |>
  dplyr::select( Protein.Ids, pearson) |>
  vroom::vroom_write( file.path( results_dir
                                 , "protein_qc"
                                 , "ruv_protein_level_correlation.tab") )

```

## Plot PCA plot after RUVIII-C
```{r}
pca_plot_after_ruvIIIc_group <- plotPca( ruv_normalized_results_cln_obj
        , group_column = "group"
        , label_column = ""
        , title = ""
        , geom_text_size = 8 )

pdf(file.path( results_dir, "protein_qc", "pca_plot_after_ruvIIIc.pdf" ))
pca_plot_after_ruvIIIc_group
dev.off()

png(file.path( results_dir, "protein_qc", "pca_plot_after_ruvIIIc.png"))
pca_plot_after_ruvIIIc_group
dev.off()


pca_plot_after_ruvIIIc_group


pca_mixomics_after_ruvIIIc <- getPcaMatrix ( ruv_normalized_results_cln_obj)

```


## Plot RLE Plot after RUVIII-C
```{r}

rle_plot_after_ruvIIIc_group <- plotRle( ruv_normalized_results_cln_obj
                                            , group="group"
                                            , ylim=c(-3, 3) )

pdf(file.path( results_dir, "protein_qc", "rle_plot_after_ruvIII_by_groupc.pdf" ))
rle_plot_after_ruvIIIc_group
dev.off()

png(file.path( results_dir, "protein_qc", "rle_plot_after_ruvIIIc_by_group.png"))
rle_plot_after_ruvIIIc_group
dev.off()

rle_plot_after_ruvIIIc_group
```

## Calculate Pearson correlation between each pair of technical replicate samples after RUVIII-C

```{r}
ruv_correlation_vec <-  pearsonCorForSamplePairs(ruv_normalized_results_cln_obj
                            , tech_rep_remove_regex = "pool" ) 

ruv_pearson_correlation_pair <- ruv_correlation_vec |>
 ggplot( aes( pearson_correlation)) +
  geom_histogram( breaks = seq(min(round(ruv_correlation_vec$pearson_correlation-0.5,2), na.rm=TRUE), 1, 0.001)) +
   scale_y_continuous(breaks = seq(0, 15, 1), limits = c(0, 15)) +
  xlab("Pearson Correlation") +
  ylab("Counts") +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank())

ruv_pearson_correlation_pair

ggsave( plot=ruv_pearson_correlation_pair
        ,filename= file.path( results_dir, "protein_qc", "ruv_pearson_correlation_pair.pdf") )

```




```{r}
ruv_normalized_filtered_results_obj <- filterSamplesByProteinCorrelationThreshold(ruv_normalized_results_cln_obj
                                           , ruv_correlation_vec          
                                           , min_pearson_correlation_threshold = 0.5 )


```



## RUV normalized data for de analysis

After RUVIII-C is performed, some proteins will have their values blanked out as NA by RUV (most likely assoicated with their batch effects).
Therefore, we need to remove proteins with high proportion of missing values in each group before procceding to DE analysis.
```{r}

vroom::vroom_write( ruv_normalized_filtered_results_obj@protein_data
              , file.path(results_dir, "protein_qc",  "ruv_normalized_results_cln_with_replicates.tsv"))


saveRDS( ruv_normalized_filtered_results_obj
        , file.path( results_dir, "protein_qc", "ruv_normalized_results_cln_with_replicates.RDS"))


ruv_normalized_for_de_analysis_obj <-  ruv_normalized_filtered_results_obj

ruv_normalized_for_de_analysis <- ruv_normalized_for_de_analysis_obj@protein_data |>
  pivot_longer( cols= !matches("Protein.Ids")
                , names_to = "replicates"
                , values_to = "Log2.Protein.Imputed") |>
  dplyr::select( Protein.Ids, replicates, Log2.Protein.Imputed) |>
  mutate( Protein.Imputed = 2^Log2.Protein.Imputed) |>
  mutate( Protein.Imputed = ifelse( is.na(Protein.Imputed), NA, Protein.Imputed)) |>
  pivot_wider( id_cols = Protein.Ids
               , names_from = replicates
               , values_from = Protein.Imputed) |>
  dplyr::rename(uniprot_acc = "Protein.Ids")

vroom::vroom_write(ruv_normalized_for_de_analysis
        , file.path(results_dir, "protein_qc", "ruv_normalized_results.tsv"))

vroom::vroom_write(ruv_normalized_for_de_analysis |>
                     dplyr::mutate(across(!matches("uniprot_acc"), log2))
        , file.path(results_dir, "protein_qc", "ruv_normalized_results_log.tsv"))

vroom::vroom_write( design_matrix |>
                      distinct( replicates, group) |>
                      dplyr::rename( Run = replicates)
                    , file.path(results_dir, "protein_qc", "design_matrix_avrg.tsv") )



ruv_normalized_for_de_analysis_mat <- ruv_normalized_for_de_analysis |>
  column_to_rownames("uniprot_acc") |>
  as.matrix()


```


## Create input files for ProteomeScholaR
```{r}

vroom::vroom_write(ruv_normalized_for_de_analysis
        , file.path(results_dir, "protein_qc", "ruv_normalized_results.tsv"))


ruv_normalized_for_de_analysis <- vroom::vroom(file.path(results_dir
        , "protein_qc"
        , "ruv_normalized_results.tsv"))

uniprot_tbl <- vroom::vroom(file.path( data_dir, "UniProt", "uniprotkb_proteome_UP000000589_2024_10_03.tsv.gz" ))


mock_protein_groups <- ruv_normalized_for_de_analysis |>
        left_join(  uniprot_tbl
                , by = join_by( uniprot_acc == Entry)) |>
        dplyr::rename(gene_names = `Gene Names`) |>
        dplyr::mutate( gene_names = purrr::map_chr( gene_names, \(x){ str_split(x, " ")[[1]][1]} ) ) |> # Only take the first gene name
        dplyr::mutate( id = row_number() -1 ) |>
        dplyr::mutate( protein_ids = uniprot_acc) |>
        dplyr::select( id, protein_ids, gene_names, matches("(\\d+|Ref)"))

vroom::vroom_write( mock_protein_groups
        , file.path(results_dir, "protein_qc", "mock_proteinGroups.tsv"))
```
# maxquant_row_id	num_gene_names	gene_names	uniprot_acc	is_unique	protein_ids

* I need to import the UniProt tab file
* Then I need to recreate the cleaned_accession_to_protein_group.tab file
```{r}
#  ruv_normalized_results <- saveRDS( file.path( results_dir, "protein_qc", "ruv_normalized_results.RDS"))


cleaned_accession_table <- data.frame( protein_ids = setdiff(colnames(ruv_normalized_filtered_results_obj@protein_data), "Protein")) |>
        left_join( uniprot_tbl
        , by = join_by( protein_ids == Entry)) |>
        mutate( maxquant_row_id = row_number() -1 ) |>
        mutate( num_gene_names =   1) |>
        mutate( gene_names = purrr::map_chr( `Gene Names`, \(x){ str_split(x, " ")[[1]][1]} ) ) |>
        mutate( is_unique = "Unique") |>
        mutate ( uniprot_acc = protein_ids) |>
        dplyr::select(maxquant_row_id,	num_gene_names,	gene_names,	uniprot_acc,	is_unique,	protein_ids )

cleaned_accession_table


vroom::vroom_write(cleaned_accession_table
        , file.path(results_dir, "clean_proteins", "cleaned_accession_to_protein_group.tab"))


```

## PCA plot after the data from technical replicates have been averaged
```{r}
pca_plot_after_ruvIIIc_avg_group <-  plotPca( ruv_normalized_for_de_analysis_obj
            , group_column = "group"
            , label_column = ""
            , title = ""
            , geom_text_size = 8) 

pdf(file.path( results_dir, "protein_qc", "pca_plot_after_ruvIIIc_by_avg_group.pdf" ))
pca_plot_after_ruvIIIc_avg_group
dev.off()

png(file.path( results_dir, "protein_qc", "pca_plot_after_ruvIIIc_by_avg_group.png"))
pca_plot_after_ruvIIIc_avg_group
dev.off()


pca_plot_after_ruvIIIc_avg_group
```




## RLE plot after data from technical replicates have been averaged
```{r}

rle_plot_after_ruvIIIc_avg_group <- plotRle( ruv_normalized_for_de_analysis_obj
        , group="group"
        , ylim=c(-3, 3) )

pdf(file.path( results_dir, "protein_qc", "rle_plot_after_ruvIII_by_avg_groupc.pdf" ))
rle_plot_after_ruvIIIc_avg_group
dev.off()

png(file.path( results_dir, "protein_qc", "rle_plot_after_ruvIIIc_by_avg_group.png"))
rle_plot_after_ruvIIIc_avg_group
dev.off()

rle_plot_after_ruvIIIc_avg_group

```

## Quick and simple heatmap to visualize the data
```{r}
heatmap(log2(ruv_normalized_for_de_analysis_mat[,setdiff(colnames(ruv_normalized_for_de_analysis_mat),"pool")]))
```


## Join PCA scatter plots and PCA density plots
```{r}
pca_variates_list <- list( pca_mixomics_before_cyclic_loess
                           ,pca_mixomics_before_ruvIIIc
                           ,pca_mixomics_after_ruvIIIc )


plotPcaDensity <- function( input_table, variable) {
  input_table |>
  ggplot( aes( {{variable}}, fill = replicates , colour = replicates, alpha=0.4 )) +
  geom_density( )  +    
  scale_alpha_continuous( guide = "none" ) +
  scale_fill_discrete(name = "Batch") +
  scale_colour_discrete(name = "Batch") +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank()) 
  
}

pca_dencity_pc1_list <- purrr::map( pca_variates_list, \(x) { plotPcaDensity(x, PC1) }) 
pca_dencity_pc2_list <- purrr::map( pca_variates_list, \(x) { plotPcaDensity(x, PC2) }) 

```




# RUV, RLE, Pearson correlation
```{r}
pca_ruv_rle_correlation_merged <- pca_plot_before_cyclic_loess_group + #  xlim(-40, 45) + ylim(-30, 25) + 
  ggtitle("a)") +
  theme(text=element_text(size=15)) +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank()) + 
  pca_plot_before_ruvIIIc_group + # xlim(-40, 45) + ylim(-30, 25) + 
  ggtitle("b)") +
  theme(text=element_text(size=15)) +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank()) +
  pca_plot_after_ruvIIIc_group + # xlim(-40, 45) +  ylim(-30, 25) + 
  ggtitle("c)") +
  theme(text=element_text(size=15)) +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank()) +
  plot_layout(  guides = 'collect') +
  
  
  # pca_dencity_pc1_list[[1]] + ggtitle("d)") + ylab("Density") +   
  # theme(text=element_text(size=15)) +
  # 
  # pca_dencity_pc1_list[[2]] + ggtitle("e)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  # 
  # pca_dencity_pc1_list[[3]] + ggtitle("f)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  # 
  # pca_dencity_pc2_list[[1]] + ggtitle("g)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  # 
  # pca_dencity_pc2_list[[2]] + ggtitle("h)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  # 
  # pca_dencity_pc2_list[[3]] + ggtitle("i)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  # 
  # plot_layout(ncol = 3
  #             , guides = 'collect') +
  # 
# RLE
rle_plot_before_cyclic_loess + ggtitle("d)") +
  theme(text=element_text(size=15)) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  rle_plot_before_ruvIIIc_group  + ggtitle("e)") +
  theme(text=element_text(size=15)) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  rle_plot_after_ruvIIIc_group  + ggtitle("f)") +
  theme(text=element_text(size=15)) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  plot_layout(guides = 'collect')  +
  
  # Pearson correlation
pearson_correlation_pair_before_cyclic_loess + ggtitle("g)") +
    theme(text=element_text(size=15)) +
pearson_correlation_pair_before_ruvIIIc + ggtitle("h)") +
    theme(text=element_text(size=15)) +
ruv_pearson_correlation_pair + ggtitle("i)") +
    theme(text=element_text(size=15)) +
  plot_layout(   ncol = 3
                , guides = 'collect')

pca_ruv_rle_correlation_merged

ggsave(plot=pca_ruv_rle_correlation_merged, filename =  file.path( results_dir, "protein_qc", "pca_rle_pearson_corr_plots_merged.png"), width = 14, height=14 )
ggsave(plot=pca_ruv_rle_correlation_merged, filename =  file.path( results_dir, "protein_qc", "pca_rle_pearson_corr_plots_merged.pdf"), width = 14, height=14 )
ggsave(plot=pca_ruv_rle_correlation_merged, filename =  file.path( results_dir, "protein_qc", "pca_rle_pearson_corr_plots_merged.svg"), width = 14, height=14 )
```



```{r}

## Add PCA plots
pca_ruv_rle_correlation_merged <- pca_plot_before_cyclic_loess_group + # xlim(-40, 45) + ylim(-30, 25) + 
  ggtitle("a)") +
  theme(text=element_text(size=15)) +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank()) + 
  pca_plot_before_ruvIIIc_group + # xlim(-40, 45) + ylim(-30, 25) + 
  ggtitle("b)") +
  theme(text=element_text(size=15)) +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank()) +
  pca_plot_after_ruvIIIc_group + # xlim(-40, 45) + ylim(-30, 25) + 
  ggtitle("c)") +
  theme(text=element_text(size=15)) +
  theme( panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         , panel.background = element_blank()) +
  plot_layout(  guides = 'collect') +
  
  ## Add PC1 Density Plot
  # pca_density_pc1_list[[1]] + ggtitle("d)") + ylab("Density") +   
  # theme(text=element_text(size=15)) +
  # 
  # pca_density_pc1_list[[2]] + ggtitle("e)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  # 
  # pca_density_pc1_list[[3]] + ggtitle("f)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  
  ## Add PC2 density plots
  # pca_density_pc2_list[[1]] + ggtitle("g)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  # 
  # pca_density_pc2_list[[2]] + ggtitle("h)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  # 
  # pca_density_pc2_list[[3]] + ggtitle("i)") + ylab("Density") +  
  # theme(text=element_text(size=15)) +
  # 
  # plot_layout(ncol = 3
  #             , guides = 'collect') +
  
# Add RLE plots
rle_plot_before_cyclic_loess + ggtitle("d)") +
  theme(text=element_text(size=15)) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  rle_plot_before_ruvIIIc_group  + ggtitle("e)") +
  theme(text=element_text(size=15)) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  rle_plot_after_ruvIIIc_group  + ggtitle("f)") +
  theme(text=element_text(size=15)) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  plot_layout(guides = 'collect')  +
  
# Add Pearson correlation plots
pearson_correlation_pair_before_cyclic_loess + ggtitle("g)") +
    theme(text=element_text(size=15)) +
pearson_correlation_pair_before_ruvIIIc + ggtitle("h)") +
    theme(text=element_text(size=15)) +
ruv_pearson_correlation_pair + ggtitle("i)") +
    theme(text=element_text(size=15)) +
  plot_layout(   ncol = 3
                , guides = 'collect')

pca_ruv_rle_correlation_merged 

ggsave(plot=pca_ruv_rle_correlation_merged, filename =  file.path( results_dir, "protein_qc", "pca_rle_pearson_corr_plots_merged.png"), width = 14, height=14 )
ggsave(plot=pca_ruv_rle_correlation_merged, filename =  file.path( results_dir, "protein_qc", "pca_rle_pearson_corr_plots_merged.pdf"), width = 14, height=14 )
ggsave(plot=pca_ruv_rle_correlation_merged, filename =  file.path( results_dir, "protein_qc", "pca_rle_pearson_corr_plots_merged.svg"), width = 14, height=14 )
```



## Read experimental contrasts file 
```{r}

contrasts_tbl <- vroom::vroom(experimental_contrast_file, delim = "\t")
```

## Run Differential Abundance Analysis
```{r}


de_analysis_results_list <- deAnalysisWrapperFunction ( ruv_normalized_for_de_analysis_obj
                                       , contrasts_tbl
                                       , formula_string = formula_string
                                       , de_q_val_thresh = de_q_val_thresh
                                       , treat_lfc_cutoff = treat_lfc_cutoff
                                       , eBayes_trend = eBayes_trend
                                       , eBayes_robust = eBayes_robust
                                       , args_group_pattern = args_group_pattern
                                       , args_row_id = args_row_id ) 
```



```{r}
up <- UniProt.ws(taxId=10090)

annotations <-  NULL
if( !file.exists( file.path( tmp_dir, "uniprot_annotations.RDS"))) {


  annotations <- batchQueryEvidenceGeneId( ruv_normalized_for_de_analysis_obj@protein_data
                                           , gene_id_column = "Protein.Ids"
                                           , uniprot_handle = up
                                           , uniprot_columns = c( "protein_existence"
                                                                  , "annotation_score"#?
                                                                  , "reviewed"
                                                                  , "gene_names"
                                                                  , "protein_name"
                                                                  , "length"
                                                                  , "xref_ensembl"
                                                                  , "go_id"
                                                                  , "keyword" ))  
  
  uniprot_dat_cln <- uniprotGoIdToTerm(annotations, uniprot_id_column = Entry
                                       , go_id_column = Gene.Ontology.IDs
                                       , sep="; " )  |>
    dplyr::rename( Protein_existence = "Protein.existence"
                   , Protein_names = "Protein.names")
  
  saveRDS( uniprot_dat_cln, file.path( tmp_dir, "uniprot_annotations.RDS") )
  
} else {
  uniprot_dat_cln <- readRDS( file.path( tmp_dir, "uniprot_annotations.RDS") )
}

```

## This is the bunch of code that is specific to this project 
* I need to add noAHA abundance in the table
* I need to add noAHA percentile in the table 
* I need to add WT versus noAHA log2 fold-change
* I need to add TG versus noAHA log2 fold-change
```{r}
# get noAHA abundance 
ruv_normalized_for_de_analysis_obj@protein_data

noAHA_log2_intensity <- ruv_normalized_for_de_analysis_obj@protein_data |>
  mutate( NoAHA = round(NoAHA,2)) |> 
  dplyr::rename( noAHA_log2_intensity = NoAHA) |>
  dplyr::select( Protein.Ids, noAHA_log2_intensity) 
  
# Calculate noAHA percentile in the table  from smallest to largest
noAHA_percentile <- noAHA_log2_intensity |>
  dplyr::mutate( noAHA_percentile = purrr::map_dbl( noAHA_log2_intensity, \(x){ round(sum( x > noAHA_log2_intensity)/length(noAHA_log2_intensity)*100,0) } )) |>
  dplyr::select( Protein.Ids, noAHA_percentile)

# Calculate WT versus noAHA log2 fold-change
wt_vs_no_aha_log2_fc <- ruv_normalized_for_de_analysis_obj@protein_data |>
  mutate( WT = round((WT1 + WT2 + WT3  + WT5 + WT6)/5,2)) |>
  dplyr::select( Protein.Ids, WT, NoAHA) |>
  dplyr::mutate( WT_vs_noAHA_log2_fc = WT - NoAHA ) |>
  dplyr::select( Protein.Ids, WT_vs_noAHA_log2_fc)

tg_vs_no_aha_log2_fc <- ruv_normalized_for_de_analysis_obj@protein_data |>
  mutate( TG = round((TG1 + TG2 + TG3 + TG4 + TG5 )/5, 2) ) |>
  dplyr::select( Protein.Ids, TG, NoAHA) |>
  dplyr::mutate( TG_vs_noAHA_log2_fc = TG - NoAHA ) |>
  dplyr::select( Protein.Ids, TG_vs_noAHA_log2_fc)

uniprot_dat_edited <- uniprot_dat_cln |>
  left_join( noAHA_log2_intensity, by = join_by(Entry == Protein.Ids)) |>
  left_join( noAHA_percentile, by = join_by(Entry == Protein.Ids)) |>
  left_join( wt_vs_no_aha_log2_fc, by = join_by(Entry == Protein.Ids)) |>
  left_join( tg_vs_no_aha_log2_fc, by = join_by(Entry == Protein.Ids)) |>
  relocate( noAHA_log2_intensity, noAHA_percentile, WT_vs_noAHA_log2_fc, TG_vs_noAHA_log2_fc, .before = Protein_existence) 

  saveRDS( uniprot_dat_edited, file.path( tmp_dir, "uniprot_dat_edited.RDS") )
```


## Output the results of the DE analysis
```{r}

outputDeAnalysisResults (de_analysis_results_list
                         , uniprot_dat_edited # normally replace uniprot_dat_edited with uniprot_dat_cln
                         , de_output_dir
                         , publication_graphs_dir
                         , file_prefix="de_proteins"
                         , plots_format = plots_format
                         , args_row_id = "uniprot_acc"
                         , de_q_val_thresh=0.05
                         , gene_names_column = "Gene.Names")
```


```{r}
  ## Write interactive volcano plot
  counts_mat <- (de_analysis_results_list$theObject)@protein_data |>
    column_to_rownames((de_analysis_results_list$theObject)@protein_id_column  ) |>
    as.matrix()
  
  this_design_matrix <- de_analysis_results_list$theObject@design_matrix
  
   rownames( this_design_matrix ) <- this_design_matrix$Run

   this_groups <- this_design_matrix[colnames( counts_mat), "group"]
  
    
  writeInteractiveVolcanoPlotProteomics(  de_analysis_results_list$de_proteins_long
                                         , groups = this_groups
                                         , uniprot_dat_edited 
                                         , de_analysis_results_list$contrasts_results$fit.eb
                                         , args_row_id = "uniprot_acc"
                                         , publication_graphs_dir
                                         , de_q_val_thresh = 0.05
                                         , counts_tbl = counts_mat
                                         , gene_names_column = "Gene.Names"
                                         , display_columns= c(  "noAHA_log2_intensity"
                                                                , "noAHA_percentile"
                                                                , "WT_vs_noAHA_log2_fc"
                                                                , "TG_vs_noAHA_log2_fc"
                                                                , "Protein_existence"
                                                               , "Annotation"
                                                               , "Reviewed"
                                                               , "Protein_names" 
                                                               , "Length"
                                                               , "Ensembl"
                                                               , "Keywords"
                                                               , "go_term_go_biological_process" 
                                                               , "go_term_go_cellular_compartment"
                                                               , "go_term_go_molecular_function"  ))
  

```


```{r}
print(head(counts_mat))
print(head(de_analysis_results_list$de_proteins_long))
print(head(uniprot_dat_cln))
```
## Check missing values at different stages of the analysis 
```{r}
which( is.na( ruv_normalized_results_cln_obj@protein_data ) )

which( is.na(normalized_frozen_protein_matrix_obj@protein_data ) )

which( is.na(remove_proteins_with_only_one_rep@protein_data ) )


```

