---
title: "R Notebook"
output: html_notebook
---



## Packages installation and loading
```{r}

#Test if BioManager is installed 
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
   BiocManager::install(version = "3.12")
}

# load pacman package manager
if(!require(pacman)){
    install.packages("pacman")
    library(pacman)
}
p_load(optparse)
p_load(tictoc)
```


## Parameters
```{r}
group_pattern <- "RPE" 

## Directories management 
base_dir <- here::here()
data_dir <- file.path( base_dir, "Data", "ALPK1")
results_dir <- file.path(base_dir, "Results", "ALPK1", group_pattern, "Proteins")
source_dir <- file.path(base_dir, "Source")

```


```{r}

de_proteins_file <- file.path(results_dir, "de_proteins_longer.tsv")
contrasts_file <- file.path( data_dir, "contrast_strings.tab")  
design_matrix_file <- file.path(results_dir, "design_matrix_cleaned.tab")
counts_table_file <- file.path(results_dir, "ruv_cleaned_counts_round1.tsv")
contrasts_file <- file.path( data_dir, "contrast_strings.tab")
formula_string <- "~ 0 + group "
my_sample_id <- "Sample_ID"
my_group_id <- "group"
my_row_id <- "uniprot_acc"
taxonomy_id <- 10090




command_line_options <- commandArgs(trailingOnly = TRUE)
if( length(command_line_options ) > 0 ) {
  
  parser <- OptionParser(add_help_option =TRUE)
  

  parser <- add_option(parser, c( "--tax-id"), type="integer", default="", dest = "taxonomy_id",
                       help="The NCBI taxonomy ID of the organism being investigated (e.g. M. musculus=10090, H. sapien=9606).",
                       metavar="integer")     

  parser <- add_option(parser, c( "--group-pattern"), type="character", default="", dest = "group_pattern",
                       help="Regular expression pattern to identify columns with abundance values belonging to the experiment. [default %default]",
                       metavar="string")
  
  parser <- add_option(parser, c("-c", "--counts"), type="character", default="", dest = "counts_table_file",
                       help="Input file with the protein abundance values",
                       metavar="string")
  
  parser <- add_option(parser, c("-c", "--de-proteins"), type="character", default="", dest = "de_proteins_file",
                       help="Input file with the list of diffierentiall expressed protein log fold-change and q-values for every contrasts.",
                       metavar="string")
  
  parser <- add_option(parser, c("--contrasts"), type="character", default="", dest = "contrasts_file",
                       help="Input file with a table listing all comparisons to be made in string, one comparison per line (e.g. groupB.vs.group_A = groupB - groupA).",
                       metavar="string")  
  
  parser <- add_option(parser, c("--formula"), type="character", default="", dest = "formula_string",
                       help="A string representing the formula for input into the model.frame function. (e.g. ~ 0 + group).",
                       metavar="string")    
  
  parser <- add_option(parser, c("-d", "--design-matrix"), type="character", default="", dest = "design_matrix_file",
                       help="Input file with the design matrix",
                       metavar="string")
  
  parser <- add_option(parser, c("-o", "--output-dir"), type="character", default="", dest = "results_dir",
                       help="Directory path for all results files.",
                       metavar="string")
  
  parser <- add_option(parser, c("--sample-id"), type="character", default="Sample_ID", dest = "my_sample_id",
                       help="A string describing the sample ID. This must be a column that exists in the design matrix.",
                       metavar="string") 
  
  parser <- add_option(parser, c("-g", "--group-id"), type="character", default="group", dest = "my_group_id",
                       help="A string describing the experimental group ID. This must be a column that exists in the design matrix.",
                       metavar="string") 
  
  parser <- add_option(parser, c("-r", "--row-id"), type="character", default="uniprot_acc", dest = "my_row_id",
                       help="A string describing the row id.",
                       metavar="string") 

  
  print(commandArgs(trailingOnly = TRUE))
  
  cmd_arguments <- parse_args(parser)
  
  print(cmd_arguments)

  group_pattern <- cmd_arguments$group_pattern
  design_matrix_file <- cmd_arguments$design_matrix_file
  counts_table_file <- cmd_arguments$counts_table_file
  de_proteins_file <- cmd_arguments$de_proteins_file
  contrasts_file <- cmd_arguments$contrasts_file
  results_dir <- cmd_arguments$results_dir
  my_sample_id <- cmd_arguments$my_sample_id
  my_group_id <-  cmd_arguments$my_group_id
  my_row_id <- cmd_arguments$my_row_id
  taxonomy_id <- cmd_arguments$taxonomy_id
  
  if( group_pattern == "") { stop("No --group-pattern variable supplied.") }
  if( column_pattern == "") { stop("No --column-pattern variable supplied.") }
  if( design_matrix_file == "" ) { stop("No file --design-matrix was supplied. It is a file describing the sample names and treatment groups.") }
  if( formula_string == "" ) { stop("No file --formula was supplied.") }
  if( counts_table_file == "" ) { stop("No value for --counts was supplied. It is a file with the abundance values for the proteins for each sample.") }
  if( de_proteins_file == "" ) { stop("No value for --de-proteins was supplied.") }
  if( results_dir == "" ) { stop("No value for --output-dir was supplied.") }
  if( taxonomy_id == "" ) { stop("No value for --tax_id was supplied. ") }

    
}
```

```{r}
p_load(tidyverse)
p_load(plotly)
p_load(vroom)
p_load(ggplot2)
p_load(ggpubr)
p_load(openxlsx)
p_load(magrittr)
p_load(knitr)
p_load(rlang)


p_load(limma)
p_load(qvalue)

p_load(UniProt.ws)
p_load(msigdb)
p_load(ExperimentHub)
p_load(GSEABase)
p_load(ProteomeRiver)
```



## Create directories
```{r}
create_dir_if_not_exists( results_dir)

```

```{r}
contrasts_tbl <- NA
if( contrasts_file != "") {
  print("Read file with lists of experimental contrasts to test.")
  contrasts_tbl <- vroom::vroom( contrasts_file, delim="\t")
}

```


## Create the design matrix
```{r}

design_mat_cln <- vroom::vroom(design_matrix_file) %>% as.data.frame()

rownames( design_mat_cln) <- design_mat_cln %>% pull ( as.name( my_sample_id))

## Check that the sample ID and group ID name is found in the design matrix 
if(   length( which( c(my_sample_id, my_group_id ) %in% colnames( design_mat_cln) )) != 2   ) {
    stop("Sample ID and group ID are not matching to the column names used in the design matrix.")
}

cols_for_analysis <- design_mat_cln %>% pull(as.name( my_sample_id))

```



## List of read abundance tables after normalization with RUVIII
```{r}

norm_abundance <- vroom::vroom( file.path( results_dir, "ruv_cleaned_counts_round1.tsv" ) )  %>%
  mutate( best_uniprot_acc = purrr::map_chr( uniprot_acc, ~str_split(. , ":"   )[[1]][1] ))

norm_abundance_mat <- norm_abundance %>%
  dplyr::select(best_uniprot_acc, contains( group_pattern ) ) %>%
  column_to_rownames("best_uniprot_acc")

```

## Functions to create the replicate matrix for RUVs
```{r}

ruvIII_replicates_matrix <- get_ruvIII_replicate_matrix( design_mat_cln,  
                                                         !!rlang::sym(my_sample_id), 
                                                         !!rlang::sym(my_group_id))

ruvIII_replicates_matrix
```



## Set up list of contrasts
```{r}

ff <- as.formula(  formula_string)
mod_frame <- model.frame( ff, design_mat_cln)
design_m <- model.matrix( ff, mod_frame)

# lists_of_contrasts

  contr.matrix <- makeContrasts( contrasts = contrasts_tbl %>% pull(contrasts),
                                 levels = colnames(design_m))
  
  
  contrasts_tbl[,1][[1]]
  
  
  

```



 EH5421 | msigdb.v7.2.hs.SYM 
  EH5422 | msigdb.v7.2.hs.EZID
  EH5423 | msigdb.v7.2.mm.SYM 
  EH5424 | msigdb.EZID
```{r}

eh <- ExperimentHub()
query(eh , 'msigdb')

```


```{r}

org_abbrev <- case_when(taxonomy_id == 10090  ~ "mm", 
          taxonomy_id == 9606 ~ "hs",
          TRUE ~ NA_character_)

if( is.na( org_abbrev ) ) {
  stop("Organism not yet supported by this code. Only human and mouse are supported.")
}

msigdb.EZID <- getMsigdb(org_abbrev, 'EZID')
msigdb.EZID <- appendKEGG(msigdb.EZID)

```



## List of all proteins
```{r}

#proteins_cln <- norm_abundance %>%  arrange( best_uniprot_acc) %>% distinct( best_uniprot_acc) %>% dplyr::rename( uniprot_acc = "best_uniprot_acc")


proteins_cln <- vroom::vroom(de_proteins_file) %>%
  mutate( best_uniprot_acc = purrr::map_chr( uniprot_acc, ~str_split(. , ":"   )[[1]][1] )) %>%
  dplyr::select( -uniprot_acc ) %>%
  dplyr::rename( uniprot_acc = "best_uniprot_acc") 

```


## Convert Uniprot accession to Entrez gene ID
```{r}

if( file.exists(file.path( results_dir, "uniprot_acc_to_entrez_id.RDS") )) {
    
    uniprot_acc_to_entrez_id <- readRDS( file.path( results_dir, "uniprot_acc_to_entrez_id.RDS"))
} else {
    up <- UniProt.ws(taxId=taxonomy_id )
    
    #keytypes(up)
    
    uniprot_acc_to_entrez_id <- batch_query_evidence( proteins_cln %>% distinct(uniprot_acc), uniprot_acc, up, 
                                                      uniprot_columns = c("ENTREZ_GENE"))

    saveRDS( uniprot_acc_to_entrez_id, file.path( results_dir, "uniprot_acc_to_entrez_id.RDS") )    
}

```
 [1] "AARHUS/GHENT-2DPAGE"        "AGD"                        "ALLERGOME"                  "ARACHNOSERVER"              "BIOCYC"                     "CGD"                        "CLEANEX"                    "CONOSERVER"                 "CYGD"                       "DICTYBASE"                 
[11] "DIP"                        "DISPROT"                    "DMDM"                       "DNASU"                      "DRUGBANK"                   "ECHOBASE"                   "ECO2DBASE"                  "ECOGENE"                    "EGGNOG"                     "EMBL/GENBANK/DDBJ"         
[21] "EMBL/GENBANK/DDBJ_CDS"      "ENSEMBL"                    "ENSEMBL_GENOMES"            "ENSEMBL_GENOMES PROTEIN"    "ENSEMBL_GENOMES TRANSCRIPT" "ENSEMBL_PROTEIN"            "ENSEMBL_TRANSCRIPT"         "ENTREZ_GENE"                "EUHCVDB"                    "EUPATHDB"                  
[31] "FLYBASE"                    "GENECARDS"                  "GENEFARM"                   "GENEID"                     "GENENAME"                   "GENETREE"                   "GENOLIST"                   "GENOMEREVIEWS"              "GENOMERNAI"                 "GERMONLINE"                
[41] "GI_NUMBER*"                 "H-INVDB"                    "HGNC"                       "HOGENOM"                    "HPA"                        "HSSP"                       "KEGG"                       "KO"                         "LEGIOLIST"                  "LEPROMA"                   
[51] "MAIZEGDB"                   "MEROPS"                     "MGI"                        "MIM"                        "MINT"                       "NEXTBIO"                    "NEXTPROT"                   "OMA"                        "ORPHANET"                   "ORTHODB"                   
[61] "PATRIC"                     "PDB"                        "PEROXIBASE"                 "PHARMGKB"                   "PHOSSITE"                   "PIR"                        "POMBASE"                    "PPTASEDB"                   "PROTCLUSTDB"                "PSEUDOCAP"                 
[71] "REACTOME"                   "REBASE"                     "REFSEQ_NUCLEOTIDE"          "REFSEQ_PROTEIN"             "RGD"                        "SGD"                        "TAIR"                       "TCDB"                       "TIGR"                       "TUBERCULIST"               
[81] "UCSC"                       "UNIPARC"                    "UNIPATHWAY"                 "UNIPROTKB"                  "UNIPROTKB_ID"               "UNIREF100"                  "UNIREF50"                   "UNIREF90"                   "VECTORBASE"                 "WORLD-2DPAGE"              
[91] "WORMBASE"                   "WORMBASE_PROTEIN"           "WORMBASE_TRANSCRIPT"        "XENBASE"                    "ZFIN"                      




```{r}


de_prot_for_camera_output <- merge_with_entrez_id( proteins_cln, 
                                                   uniprot_acc_to_entrez_id  )


de_prot_for_camera_tab <-  de_prot_for_camera_output$de_proteins %>%
    group_by( comparison) %>%
    nest() %>%
    ungroup()

de_prot_for_camera_mapped <- de_prot_for_camera_tab %>%
     mutate(  data_edited = purrr::map( data, ~{ as.matrix(column_to_rownames(.,  "ENTREZ_GENE" ))   }) )


de_prot_for_camera_list <- de_prot_for_camera_tab$data
names( de_prot_for_camera_list) <- de_prot_for_camera_tab$comparison


```



```{r}
norm_abundance_helper <- norm_abundance_mat %>%
                         rownames_to_column("uniprot_acc") %>%
                           as.data.frame %>% 
                          mutate ( temp =1) %>% 
                          inner_join( data.frame( comparison=unique(names(de_prot_for_camera_list)), temp=1 ), by="temp"        )  %>%
                          dplyr::select( -temp) %>%
                         anti_join(de_prot_for_camera_output$without_entrez_id, by=c("uniprot_acc" = "uniprot_acc", 
                                                                                     "comparison" = "comparison") ) %>%
                        anti_join(de_prot_for_camera_output$excluded_duplicates, by=c("uniprot_acc" = "uniprot_acc", 
                                                                                      "comparison" = "comparison") )  %>%
                        left_join( uniprot_acc_to_entrez_id %>% 
                                   dplyr::filter(!is.na(ENTREZ_GENE)), 
                                    by=c("uniprot_acc" = "UNIPROTKB") ) %>%
                        dplyr::filter( !is.na(ENTREZ_GENE))  %>%
                        dplyr::relocate(ENTREZ_GENE, "uniprot_acc" ) %>%
                        dplyr::select(-uniprot_acc) %>%
                        mutate( ENTREZ_GENE  =trimws( ENTREZ_GENE))


norm_abundance_mat_list <- norm_abundance_helper %>%
   group_by(comparison) %>%
   nest() %>%
  ungroup %>%
  mutate( data =  purrr::map(data,  function(x) {  as.matrix(column_to_rownames(x, "ENTREZ_GENE")) })) 

## norm_abundance_mat_list$data[[1]]

```

## Prepare gene ID to gene set dictionary 

```{r}
list_of_collections <- purrr::map(msigdb.EZID, ~bcCategory(collectionType(.)) )

list_of_sub_collections <- purrr::map(msigdb.EZID, ~bcSubCategory(collectionType(.)) )


collections_tab <- data.frame( collection =unlist(list_of_collections),
                               sub_collection = unlist( list_of_sub_collections ) ) %>%
  distinct() %>%
  dplyr::filter( collection !="archived") 

gene_set_idx_list_file <- file.path(results_dir, "Functional_Enrichment", "gene_set_idx_list_file.RDS")

if(file.exists(gene_set_idx_list_file)) {
  gene_set_idx_list <- readRDS(gene_set_idx_list_file)
  
} else {
  

  gene_set_idx_list <- purrr::pmap(collections_tab, function( collection, sub_collection) { 
    
    if( is.na(sub_collection )) {
      subsetCollection(msigdb.EZID, 
                       collection=collection )        
      
    } else {
      subsetCollection(msigdb.EZID, 
                       collection=collection, 
                       subcollection=sub_collection)   
      
    }
    
    
  } )    
  
  
  names( gene_set_idx_list) <- collections_tab %>%
    mutate( all_collection = paste( collection, sub_collection, sep=", ")) %>% 
    pull(all_collection)
  
  saveRDS(gene_set_idx_list, gene_set_idx_list_file)
  
}

```


## Example usage
```{r}
abundance_mat_trimmed <- norm_abundance_mat_list[1,"data"][[1]][[1]]

msigdb_ids <- geneIds(gene_set_idx_list[["h, NA"]])

#convert gene sets into a list of gene indices
camera_indices <- ids2indices(msigdb_ids,
                              rownames(abundance_mat_trimmed))


camera( y=abundance_mat_trimmed, 
        index=camera_indices, 
        design =ruvIII_replicates_matrix, 
        contrast=c(1, -1, 0, 0 ))
```

## List of contrasts
```{r}

# colnames(ruvIII_replicates_matrix)
# 
# "RPE_B" "RPE_A" "RPE_X" "RPE_Y"
# 
# contrasts_tbl %>%
#   separate( contrasts, sep="=", into=c("contrast", "calculation")) %>%
#   pull(contrast)

lists_of_contrasts <- list( RPE_Y.vs.RPE_X=c(0, 0, -1, 1 ) ,
                            RPE_B.vs.RPE_A=c(1, -1, 0, 0 ) ,
                            RPE_B.vs.RPE_X=c(1, 0, -1, 0 ) ,
                            RPE_Y.vs.RPE_A=c(0, -1, 0, 1 ) ,
                            RPE_A.vs.RPE_X=c(0, 1, -1, 0 ) ,
                            RPE_B.vs.RPE_Y=c(1, 0, 0, -1 )   )



```


## Run the camera test 
```{r}

    create_dir_if_not_exists(file.path(results_dir, "Functional_Enrichment"))

    camera_results_file <- file.path(results_dir, "Functional_Enrichment", "camera_results.RDS")

if( file.exists(camera_results_file) ) {
    camera_results <- readRDS( camera_results_file )
} else {
    
    index_tab <- data.frame( index_name = names(gene_set_idx_list) ) %>%
        mutate( temp = 1 )
    
    
    contrast_tab <-    data.frame( contrast_name = names(lists_of_contrasts)) %>%
        mutate( temp = 1 )
    
    combination_tab <- contrast_tab %>% 
        full_join( index_tab, by="temp") %>% 
        dplyr::select(-temp)
    
    # Pre-fill the data, other data to be filled with the pmap function
    my_partial_camera <- partial( my_camera, 
                                  abundance_mat = norm_abundance_mat_list , 
                                  replicates_mat = ruvIII_replicates_matrix, 
                                  lists_of_contrasts = lists_of_contrasts, 
                                  list_of_gene_sets = gene_set_idx_list, 
                                  min_set_size=4 )
    
    camera_results <- purrr::pmap (combination_tab , 
                                   my_partial_camera)
    
    
    names(norm_abundance_mat_list)

    
    
    saveRDS( camera_results, camera_results_file)
}

```


## Covert the camera results that are stored in list structures into a table 
```{r}

camera_results_filt <- camera_results [purrr::map_lgl( camera_results, ~{!is.na(.[["camera"]][[1]][1])})]


camera_results_cln <-  purrr::map( camera_results_filt, 
  ~ { rownames_to_column(.[["camera"]], "pathway")   }  )

camera_results_tbl <- tibble( temp = camera_results_cln)  %>% 
    bind_cols(  data.frame( comparison =  purrr::map_chr( camera_results_filt, 
   ~ {  .[["contrast_name"]]  } ) )  ) %>% 
    bind_cols(  data.frame( gene_set =  purrr::map_chr( camera_results_filt, 
   ~ {  .[["index_name"]]  } ) )  ) %>%
    unnest(temp)

camera_results_filt <- camera_results_tbl %>%
    dplyr::filter( FDR <0.05)


camera_results_tbl %>%
    arrange(FDR)


vroom::vroom_write( camera_results_filt, 
                    file.path( results_dir, "Functional_Enrichment", "filtered_camera_results.tab"))

```

