---
title: "ALPK1 RPE and NR analysis"
output:
  html_document:
    df_print: paged
---

To investigate if RUV will overfit data.

## Packages installation and loading
```{r}

options(knitr.duplicate.label = "allow")

#Test if BioManager is installed 
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
   BiocManager::install(version = "3.12")
}

# load pacman package manager
if(!require(pacman)){
    install.packages("pacman")
    library(pacman)
}

p_load(tidyverse)
p_load(vroom)
p_load(magrittr)
p_load(knitr)
p_load(rlang)
p_load(optparse)

p_load(seqinr)
p_load(ProteomeRiver)
p_load(janitor)
p_load(tictoc)

tic()
```

## Parameters
```{r}
group_pattern <- "RPE" 

## Directories management 
base_dir <- here::here()
data_dir <- file.path( base_dir, "Data", "Abundance_Data", "P90")
results_dir <- file.path(base_dir, "Results",  paste0(group_pattern, "90"),  "Proteins", "DE_Analysis")
source_dir <- file.path(base_dir, "Source")

# Minimum number of samples per experimental group for the protein to be accepted for analysis 

fasta_file <- file.path( data_dir,   "ALPK1-set1MusMusculus20201226CanIso.fasta")
raw_counts_file <- file.path(data_dir, "ALPK1-set1proteinGroups.txt")
column_pattern_input <- "Reporter intensity corrected"
output_counts_file <- "counts_table_cleaned.tab"  
accession_record_file <- "cleaned_accession_to_protein_group.tab"
fasta_meta_file <- "aa_seq_tbl.RDS"

razor_unique_peptides_group_thresh <- 0
unique_peptides_group_thresh <- 1

command_line_options <- commandArgs(trailingOnly = TRUE)
if( length(command_line_options ) > 0 ) {
  parser <- OptionParser(add_help_option =TRUE)
  
  parser <- add_option(parser, c( "--output-dir"), type="character", default="", dest = "results_dir",
                       help="Directory path for all results files.",
                       metavar="string")   
  
  parser <- add_option(parser, c( "--ids"), type="character", default="", dest = "accession_record_file",
                       help="File to link the cleaned list of accessions to the original list of protein groups from MaxQuant file. Also, contains the MaxQuant output ID.",
                       metavar="string")   
  
  parser <- add_option(parser, c("--fasta"), type="character", default="", dest = "fasta_file",
                       help="Input protein sequence FASTA file with UniProt FASTA header format.",
                       metavar="string")   
  
  parser <- add_option(parser, c("--raw-counts"), type="character", default="", dest = "raw_counts_file",
                       help="Input file with the protein abundance data.",
                       metavar="string")   
  
  parser <- add_option(parser, c( "--output-counts"), type="character", default="", dest = "output_counts_file",
                       help="String representing the name of the output counts data file which will be saved in the directory specified with the --output-dir flag.",
                       metavar="string")   
  
  parser <- add_option(parser, c("--column-pattern"), type="character", default="Reporter intensity corrected", dest = "column_pattern_input",
                       help="String pattern, together with the experimental group pattern, that matches the abundance data columns.",
                       metavar="string")   

  parser <- add_option(parser, c( "--group-pattern"), type="character", default="", dest = "group_pattern",
                       help="Regular expression pattern to identify columns with abundance values belonging to the experiment. [default %default]",
                       metavar="string")    
  
  parser <- add_option(parser, c("--r-u-count"), type="integer", default=0, dest = "razor_unique_peptides_group_thresh",
                       help="Number of razor and unique peptides for the specified experiemtal group needs to be higher than this threshold for the protein to be included for the analysis.",
                       metavar="integer")   
  
  parser <- add_option(parser, c("--u-count"), type="integer", default=1, dest = "unique_peptides_group_thresh",
                       help="Number of unique peptides for the specified experiemtal group needs to be higher than this threshold for the protein to be included for the analysis. The file will be saved in the results directory",
                       metavar="integer")   
  
  
  parser <- add_option(parser, c("--fasta-save"), type="character", default="aa_seq_tbl.RDS", dest = "fasta_meta_file",
                       help="R object storage file that records all the sequence and header information in the FASTA file.",
                       metavar="string")   
    
    

  print(command_line_options)
  
  cmd_arguments <- parse_args(parser)
  
  print(cmd_arguments)  

  results_dir <- cmd_arguments$results_dir
  fasta_file <- cmd_arguments$fasta_file
  raw_counts_file <- cmd_arguments$raw_counts_file
  output_counts_file <- cmd_arguments$output_counts_file
  column_pattern_input <- cmd_arguments$column_pattern_input
  group_pattern <- cmd_arguments$group_pattern
  razor_unique_peptides_group_thresh <- cmd_arguments$razor_unique_peptides_group_thresh
  unique_peptides_group_thresh <- cmd_arguments$unique_peptides_group_thresh
  fasta_meta_file <- cmd_arguments$fasta_meta_file
  accession_record_file  <- cmd_arguments$accession_record_file
  
  if( results_dir == "" ) { stop("No value for --output-dir was supplied.") }
  if( fasta_file == "" ) { stop("No value for --fasta was supplied.") }
  if( raw_counts_file == "" ) { stop("No value for -raw-counts was supplied.") }
  if( output_counts_file == "" ) { stop("No value for -output-counts was supplied.") }
  if( razor_unique_peptides_group_thresh == "" ) { stop("No value for --r-u-count was supplied.") }
  if( unique_peptides_group_thresh == "" ) { stop("No value for --u-count was supplied.") }
  if( fasta_meta_file == "" ) { stop("No value for --fasta-save was supplied.") }
  if( group_pattern== "") { stop("No value for --group-pattern was supplied.") }
  if( accession_record_file == "") { stop("No value for --ids was supplied.") }

}


```

## Command to convert the R markdown file to a R file for use in command line. 
knitr::purl( file.path(  source_dir, "Shared",  "clean_proteins_cmd.Rmd"), 
             output=file.path( source_dir,  "Shared",   "clean_proteins_cmd.R") ) 
     

## Create directories
```{r}
print( "Step 1: Create results directory.")

create_dir_if_not_exists( results_dir)

```

## Read input file 
```{r}
print( "Step 2: Reading the counts file.")

dat_tbl <- vroom::vroom( raw_counts_file )

```
## Clean the column names 
```{r}
print("Step 3: Clean counts table header name.")
dat_cln <-  janitor::clean_names(dat_tbl)

colnames(dat_cln) <- str_replace(colnames(dat_cln), "_i_ds", "_ids" )

```

## Clean up the regular expression pattern to extract relevant abundance data columns
```{r}
print("Step 4: Prepare regular expressions to select counts data columns.")

pattern_suffix <- "_\\d+"
if( group_pattern !="") {
   pattern_suffix <- paste("_\\d+", tolower(group_pattern), sep="_")
}

extract_patt_suffix <- "_(\\d+)"
if( group_pattern !="") {
   extract_patt_suffix <- paste0("_(\\d+)_(", tolower(group_pattern), ")")
}

column_pattern <- paste0(make_clean_names(column_pattern_input), pattern_suffix ) 

extract_replicate_group <- paste0(make_clean_names(column_pattern_input), extract_patt_suffix ) 

```
## Clean column names asociated with razor and unique peptide counts filter specific to the experimental group
```{r}
print("Step 5: Prepare column names associated with peptide counts.")

razor_unique_peptides_group_col <-  paste0( "razor_unique_peptides_", tolower(group_pattern)  ) 
unique_peptides_group_col <-  paste0("unique_peptides_", tolower(group_pattern)  )

```
    

```{r}
## Read fasta file 
# colnames(aa_seq_tbl)
print( "Step 6: Reading the FASTA file and saving the meta-data file.")

if(!file.exists( file.path( results_dir, fasta_meta_file))) {
  aa_seq_tbl <- parse_fasta_file( fasta_file)
  saveRDS( aa_seq_tbl,  file.path( results_dir, fasta_meta_file))
} else {
  aa_seq_tbl <- readRDS( file.path( results_dir, fasta_meta_file))
}

## Add the row id column and create a column containing the cleaned  peptide
print("Step 7: Get row ID and get cleaned peptide sequence.")
evidence_tbl<- dat_cln %>% 
  mutate( evidence_id = id) 
```    



    
## Filtering by number of peptides available, remove decoy peptides and common protein contaminants
```{r}
print( "Step 8: Filtering counts table by number of peptides available, remove decoy proteins and protein contaminants.")

peptides_count_helper <- evidence_tbl %>%
  dplyr::select( evidence_id, 
                 protein_ids, 
                 !!rlang::sym(razor_unique_peptides_group_col),
                 !!rlang::sym(unique_peptides_group_col), 
                 reverse,
                 potential_contaminant, 
                 matches(column_pattern)) %>%
  dplyr::filter(  is.na(reverse)	&  
                  is.na(potential_contaminant)) %>%
  dplyr::filter( !str_detect( protein_ids, "CON__") &
                 !str_detect( protein_ids, "REV__")   ) %>%
  dplyr::mutate( protein_ids = str_split( protein_ids, ";" ) ) %>%
  dplyr::mutate( !!rlang::sym(razor_unique_peptides_group_col) := str_split( !!rlang::sym(razor_unique_peptides_group_col), ";" ) )%>%
  dplyr::mutate( !!rlang::sym(unique_peptides_group_col) := str_split( !!rlang::sym(unique_peptides_group_col) , ";" ) ) %>%
  unnest(cols = c( protein_ids, 
                   !!rlang::sym(razor_unique_peptides_group_col), 
                   !!rlang::sym(unique_peptides_group_col)))

evidence_tbl_cleaned <- NA

evidence_tbl_cleaned <- peptides_count_helper %>%
    dplyr::filter( !!rlang::sym(razor_unique_peptides_group_col)  >= razor_unique_peptides_group_thresh &
                   !!rlang::sym(unique_peptides_group_col)  >= unique_peptides_group_thresh  ) 

```
  
  
## Get best accession per entry
```{r}  
print( "Step 9: Identify best UniProt accession per entry, extract sample number and simplify column header")


  accession_gene_name_tbl <- choose_best_protein_accession( input_tbl = evidence_tbl_cleaned, 
                                                    acc_detail_tab = aa_seq_tbl, 
                                                    accessions_column = protein_ids , 
                                                    row_id_column= uniprot_acc,
                                                    group_id = evidence_id)
  
  
 accession_gene_name_tbl_record <-  accession_gene_name_tbl %>%
    left_join( evidence_tbl %>% dplyr::select( evidence_id, protein_ids), by=c("evidence_id"))
  

evidence_tbl_filt <- evidence_tbl_cleaned %>%
  inner_join( accession_gene_name_tbl %>% 
                dplyr::select(evidence_id, uniprot_acc), by="evidence_id") %>%
  dplyr::select(uniprot_acc, contains(group_pattern), -contains(c("razor", "unique"))) %>%
  distinct

colnames( evidence_tbl_filt) <- str_replace_all(colnames(evidence_tbl_filt), extract_replicate_group, "\\1_\\2")  %>% 
  toupper( ) %>%
  str_replace_all( "UNIPROT_ACC", "uniprot_acc") 
  
```

## Write out the clean counts file 
```{r}
print("Step 10: Save the cleaned counts data into a file.")
vroom::vroom_write( evidence_tbl_filt, file.path(results_dir, output_counts_file) ) 
```

## Write out table that links the cleaned counts data to the original maxquant input
```{r}
vroom::vroom_write( accession_gene_name_tbl_record, file.path(results_dir, accession_record_file))
```



```{r}
toc()
sessionInfo()
```

