---
title: "R Notebook"
output: html_notebook
---
```{r}
#! /usr/bin/env Rscript

# Author(s): Ignatius Pang
# Email: ipang@cmri.org.au
# Children’s Medical Research Institute, finding cures for childhood genetic diseases
```

## Packages installation and loading
```{r}

#Test if BioManager is installed 
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
   BiocManager::install(version = "3.12")
}

# load pacman package manager
if(!require(pacman)){
    install.packages("pacman")
    library(pacman)
}

p_load(tidyverse)
p_load(plotly)
p_load(vroom)
p_load(ggplot2)
p_load(ggpubr)
p_load(tictoc)


p_load(qvalue)
p_load(knitr)

p_load(magrittr)
p_load(optparse)

```
```{r}
tic()
```



## Directories management 
```{r}

group_pattern <- "RPE"
## Directories management 
base_dir <- here::here()
data_dir <- file.path( base_dir, "Data")
phos_results_dir <- file.path(base_dir, "Results",  paste0(group_pattern, "90"),  "Phosphopeptides", "DE_Analysis")
source_dir <- file.path(base_dir, "Source")
prot_results_dir <- file.path(base_dir, "Results",  paste0(group_pattern, "90"),  "Proteins", "DE_Analysis")
results_dir <- phos_results_dir

```

## Command to convert the R markdown file to a R file for use in command line. 
knitr::purl( file.path(  source_dir, "Shared",  "norm_phos_by_prot_abundance.Rmd"), 
             output=file.path( source_dir,  "Shared",   "norm_phos_by_prot_abundance.R") ) 
    
```{r}

proteins_file <- file.path( prot_results_dir, "de_proteins_long_annot.tsv" ) 
phospho_file  <- file.path(   phos_results_dir, "de_phos_long_annot.tsv" ) 

command_line_options <- commandArgs(trailingOnly = TRUE)
if( length(command_line_options ) > 0 ) {
  parser <- OptionParser(add_help_option =TRUE)

  parser <- add_option(parser, c( "--output-dir"), type="character", default="", dest = "results_dir",
                       help="Directory path for all results files.",
                       metavar="string")   
  
  parser <- add_option(parser, c("--protein"), type="character", default="", dest = "proteins_file",
                       help="File with table of differentially expressed proteins.",
                       metavar="string")   
  
  parser <- add_option(parser, c("--phospho"), type="character", default="", dest = "phospho_file",
                       help="File with table of differentially abundant phosphosites.",
                       metavar="string") 

  print(command_line_options)
  
  cmd_arguments <- parse_args(parser)
  
  print(cmd_arguments)  

  proteins_file <- cmd_arguments$proteins_file
  phospho_file <- cmd_arguments$phospho_file
  results_dir <- cmd_arguments$results_dir

  if( results_dir == "" ) { stop("No value for --output-dir was supplied.") }
  if( proteins_file == "" ) { stop("No value for --protein was supplied. ") }
  if( phospho_file == "" ) { stop("No value for --phospho was supplied. ") }

}

```

Meaning of separators
* ; semi-colon - multiple phosphorylation sites on the same peptide
* | pipe - the same phosphopeptide maps to multiple locations on the protein
* : colon - the same phosphopeptide maps to multiple genes 
* ! exclamation mark - sites_id column, field separator for different types of data (e.g. uniprot accession, phosphosite position, phosphosite centered 15-mer sequence)

```{r}

phospho_tbl_orig <- vroom::vroom(phospho_file) 
  
phospho_tbl <- phospho_tbl_orig %>%  
  dplyr::select(sites_id, q.mod, p.mod, log2FC, comparison, maxquant_row_ids) %>%
  mutate( phosphosites_id_copy = sites_id ) %>%
  separate(phosphosites_id_copy, sep="!", into=c("uniprot_acc", "gene_symbol", "positions", "peptide")) %>%
  dplyr::rename( phos_maxquant_row_ids = "maxquant_row_ids" )

phospho_cln <- phospho_tbl %>%
  separate_rows(uniprot_acc, sep="[\\:;]") 


phospho_cln

```



```{r}


proteins_tbl_orig <-  vroom::vroom( proteins_file) 


proteins_tbl <- proteins_tbl_orig %>%
  dplyr::select( uniprot_acc, q.mod, p.mod, log2FC, comparison, maxquant_row_id) %>%
  dplyr::rename( prot_maxquant_row_ids = "maxquant_row_id")

proteins_cln <- proteins_tbl  %>%
  mutate( uniprot_acc_copy = uniprot_acc) %>%
  separate_rows(uniprot_acc_copy, sep="[\\:;]") 
```

## RPE
proteins 3021
phosphoproteins 4242
intersect proteins and phosphoproteins 631
```{r}

proteins_uniprot_list <- proteins_cln %>% distinct(uniprot_acc_copy) %>% pull(uniprot_acc_copy) 
phospho_uniprot_list <- phospho_cln %>% distinct(uniprot_acc) %>% pull(uniprot_acc)
prot_phos_uniprot_list <- intersect( proteins_uniprot_list, 
           phospho_uniprot_list ) 

proteins_uniprot_list %>% length()
phospho_uniprot_list %>% length()
prot_phos_uniprot_list %>% length()

```

## Normalisation of the phosphoproteome with the proteome 

We did all the usual processing, global normalization, imputation, etc, taking the data pretty much to the end and then he SUBTRACTED the proteome log2FC data from the phosphoproteome log2FC data.

We just left alone the phosphoproteome data that didn’t have underlying proteome data. i.e. kept it.  


```{r}

basic_data_shared <- proteins_cln %>%
  inner_join( phospho_cln, by=c("uniprot_acc_copy" = "uniprot_acc",  
                                "comparison" = "comparison"), suffix=c(".phos", ".prot") ) %>%
  mutate( norm_phos_logFC = log2FC.phos - log2FC.prot) %>%
  mutate( adj_qmod.prot  = ifelse(sign(q.mod.phos)  ==  sign(q.mod.prot), 1-q.mod.prot,  q.mod.prot  )) %>%
  mutate( combined_q_mod = 1-pchisq(-2*( log(q.mod.phos) + log(adj_qmod.prot)   ), 2*2 )  ) %>%
  dplyr::mutate( status  = "Phos_and_Prot") %>%
  dplyr::select(comparison,  norm_phos_logFC, combined_q_mod, sites_id, uniprot_acc,  log2FC.phos, q.mod.phos,  log2FC.prot, q.mod.prot, status, phos_maxquant_row_ids, prot_maxquant_row_ids )   %>%
  distinct() %>%
  as.data.frame

basic_data_phospho_only <- phospho_cln %>%
  anti_join( proteins_cln, by=c( "uniprot_acc" = "uniprot_acc_copy" ,  
                                "comparison" = "comparison") )  %>%
  dplyr::mutate(norm_phos_logFC =  log2FC, combined_q_mod = q.mod) %>%
  dplyr::rename( log2FC.phos= log2FC, q.mod.phos = q.mod ) %>%
  dplyr::mutate( status  = "Phos_Only")%>%
  dplyr::select(comparison,  norm_phos_logFC, combined_q_mod, sites_id, uniprot_acc, log2FC.phos, q.mod.phos, status, phos_maxquant_row_ids)    %>%
  as.data.frame 

basic_data <- basic_data_shared %>%
  bind_rows(basic_data_phospho_only)

nrow(basic_data) == nrow(phospho_cln)


vroom::vroom_write( basic_data, file.path( results_dir,  "norm_phosphosite_lfc_minus_protein_lfc_basic.tsv"))

```


```{r}

annotation_from_phospho_tbl <- phospho_tbl_orig %>%
  dplyr::select(-q.mod, -p.mod, -log2FC, -uniprot_acc)

annotated_phos_tbl <- basic_data %>%
  left_join( annotation_from_phospho_tbl, by=c("sites_id" = "sites_id",
                                               "comparison" = "comparison",
                                               "phos_maxquant_row_ids" = "maxquant_row_ids"))


vroom::vroom_write( annotated_phos_tbl, file.path( results_dir,  "norm_phosphosite_lfc_minus_protein_lfc_annotated.tsv"))


```


## Compare before and after normalization with protein abundance 
```{r}

before_prot_norm <- phospho_cln %>%
  dplyr::filter( q.mod < 0.05) %>%
  dplyr::distinct(comparison, sites_id) %>%
  dplyr::mutate(  Normalization =  "Before")


after_prot_norm <- basic_data %>%
  dplyr::filter( combined_q_mod < 0.05) %>%
  dplyr::distinct(comparison, sites_id) %>%
  dplyr::mutate(  Normalization =  "After")


compare_before_and_after <- before_prot_norm %>%
  full_join( after_prot_norm, by=c("comparison", "sites_id"), suffix=c(".before", ".after")) %>%
  mutate( Normalization = case_when (  is.na(Normalization.before) & !is.na(Normalization.after)  ~ "After Only",
          !is.na(Normalization.before) & !is.na(Normalization.after)   ~ "Before and After",
          !is.na(Normalization.before) & is.na(Normalization.after)   ~ "Before Only",
          TRUE ~ NA_character_ ) ) %>%
  group_by( comparison,  Normalization) %>%
  summarise( Counts = n()) %>%
  ungroup()

 compare_before_and_after

 vroom::vroom_write( compare_before_and_after, file.path(results_dir, "compare_before_and_after_norm_by_prot_abundance.tsv"))
 
 
 cmp_before_after_plot <- compare_before_and_after %>%
   ggplot( aes( Normalization, Counts)) +
   geom_col() +
   facet_grid( . ~ comparison ) +
     geom_text(stat='identity', aes(label= Counts), vjust=-0.5) 

 
 cmp_before_after_plot
 
 ggsave( plot=cmp_before_after_plot, file.path(results_dir, "compare_before_and_after_norm_by_prot_abundance.pdf"), width = 14, height=7)
 ggsave( plot=cmp_before_after_plot, file.path(results_dir, "compare_before_and_after_norm_by_prot_abundance.png"), width = 14, height=7 )

 
```

## Volcano Plot
```{r}
qm.threshold <- 0.05
logFC.threshold <- 1

basic_data_volcano_plot_data <- basic_data %>%
  left_join( annotated_phos_tbl %>%
               dplyr::distinct( uniprot_acc, gene_name), 
             by=c("uniprot_acc" = "uniprot_acc")) %>%
  dplyr::mutate( colour= case_when ( abs(norm_phos_logFC) >= logFC.threshold & combined_q_mod >= qm.threshold ~ "orange",
                                     abs(norm_phos_logFC) >= logFC.threshold & combined_q_mod < qm.threshold ~ "purple",
                                     abs(norm_phos_logFC) < logFC.threshold & combined_q_mod  < qm.threshold ~ "blue",
                                     TRUE ~ "black" ))  %>%
  dplyr::mutate( colour = factor( colour, levels=c("orange", "purple", "blue" ,"black")))

basic_data_volcano_plot <- basic_data_volcano_plot_data %>%
  ggplot( aes( norm_phos_logFC, -log10(combined_q_mod), col=colour, key=gene_name )) +
  geom_point() +
  facet_grid( . ~ comparison)  +
            scale_colour_manual(values = c("orange", "purple", "blue" ,"black"),
                                labels=c(paste0("Not significant, logFC > ", 
                                  logFC.threshold), 
                                  paste0("Significant, logFC >= ", 
                                         logFC.threshold), 
                                  paste0("Significant, logFC <", 
                                         logFC.threshold), 
                                  "Not Significant"))


ggsave(  filename=file.path( results_dir, "volplot_gg_phos_vs_prot.all.png"), plot=basic_data_volcano_plot, width=15, height=6   ) 
ggsave(  filename=file.path( results_dir, "volplot_gg_phos_vs_prot.all.svg"), plot=basic_data_volcano_plot, width=15, height=6   ) 


ggplotly(basic_data_volcano_plot)

```
```{r}
toc()
sessionInfo()
```




